// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/instructions/emergencySwap.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.emergencySwap = emergencySwap;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.bool("aToB"),
    borsh.u64("targetLimitBps"),
]);
function emergencySwap(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        {
            address: accounts.adminAuthority.address,
            role: 3,
            signer: accounts.adminAuthority,
        },
        { address: accounts.strategy, role: 1 },
        { address: accounts.globalConfig, role: 0 },
        { address: accounts.tokenAMint, role: 0 },
        { address: accounts.tokenBMint, role: 0 },
        { address: accounts.tokenAVault, role: 1 },
        { address: accounts.tokenBVault, role: 1 },
        { address: accounts.baseVaultAuthority, role: 1 },
        { address: accounts.pool, role: 1 },
        { address: accounts.position, role: 1 },
        { address: accounts.poolTokenVaultA, role: 1 },
        { address: accounts.poolTokenVaultB, role: 1 },
        { address: accounts.tickArray0, role: 1 },
        { address: accounts.tickArray1, role: 1 },
        { address: accounts.tickArray2, role: 1 },
        { address: accounts.oracle, role: 1 },
        { address: accounts.poolProgram, role: 0 },
        { address: accounts.scopePrices, role: 0 },
        { address: accounts.tokenInfos, role: 0 },
        { address: accounts.tokenATokenProgram, role: 0 },
        { address: accounts.tokenBTokenProgram, role: 0 },
        { address: accounts.memoProgram, role: 0 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([73, 226, 248, 215, 5, 197, 211, 229]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        aToB: args.aToB,
        targetLimitBps: args.targetLimitBps,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
