// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/instructions/openLiquidityPosition.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.openLiquidityPosition = openLiquidityPosition;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.i64("tickLowerIndex"),
    borsh.i64("tickUpperIndex"),
    borsh.u8("bump"),
]);
function openLiquidityPosition(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        {
            address: accounts.adminAuthority.address,
            role: 3,
            signer: accounts.adminAuthority,
        },
        { address: accounts.strategy, role: 1 },
        { address: accounts.globalConfig, role: 0 },
        { address: accounts.pool, role: 1 },
        { address: accounts.tickArrayLower, role: 1 },
        { address: accounts.tickArrayUpper, role: 1 },
        { address: accounts.baseVaultAuthority, role: 1 },
        { address: accounts.position, role: 1 },
        { address: accounts.positionMint, role: 1 },
        { address: accounts.positionMetadataAccount, role: 1 },
        { address: accounts.positionTokenAccount, role: 1 },
        { address: accounts.rent, role: 0 },
        { address: accounts.system, role: 0 },
        { address: accounts.tokenProgram, role: 0 },
        { address: accounts.tokenProgram2022, role: 0 },
        { address: accounts.tokenATokenProgram, role: 0 },
        { address: accounts.tokenBTokenProgram, role: 0 },
        { address: accounts.memoProgram, role: 0 },
        { address: accounts.associatedTokenProgram, role: 0 },
        { address: accounts.poolProgram, role: 0 },
        { address: accounts.oldTickArrayLowerOrBaseVaultAuthority, role: 1 },
        { address: accounts.oldTickArrayUpperOrBaseVaultAuthority, role: 1 },
        { address: accounts.oldPositionOrBaseVaultAuthority, role: 1 },
        { address: accounts.oldPositionMintOrBaseVaultAuthority, role: 1 },
        { address: accounts.oldPositionTokenAccountOrBaseVaultAuthority, role: 1 },
        { address: accounts.tokenAVault, role: 1 },
        { address: accounts.tokenBVault, role: 1 },
        { address: accounts.tokenAMint, role: 0 },
        { address: accounts.tokenBMint, role: 0 },
        { address: accounts.poolTokenVaultA, role: 1 },
        { address: accounts.poolTokenVaultB, role: 1 },
        { address: accounts.scopePrices, role: 0 },
        { address: accounts.tokenInfos, role: 0 },
        (0, kit_1.isSome)(accounts.eventAuthority)
            ? { address: accounts.eventAuthority.value, role: 0 }
            : { address: programAddress, role: 0 },
        { address: accounts.consensusAccount, role: 0 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([204, 234, 204, 219, 6, 91, 96, 241]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        tickLowerIndex: args.tickLowerIndex,
        tickUpperIndex: args.tickUpperIndex,
        bump: args.bump,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
