// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/instructions/openPositionWithMetadata.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.openPositionWithMetadata = openPositionWithMetadata;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    types.OpenPositionWithMetadataBumps.layout("bumps"),
    borsh.i32("tickLowerIndex"),
    borsh.i32("tickUpperIndex"),
]);
function openPositionWithMetadata(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.funder.address, role: 3, signer: accounts.funder },
        { address: accounts.owner, role: 0 },
        { address: accounts.position, role: 1 },
        {
            address: accounts.positionMint.address,
            role: 3,
            signer: accounts.positionMint,
        },
        { address: accounts.positionMetadataAccount, role: 1 },
        { address: accounts.positionTokenAccount, role: 1 },
        { address: accounts.whirlpool, role: 0 },
        { address: accounts.tokenProgram, role: 0 },
        { address: accounts.systemProgram, role: 0 },
        { address: accounts.rent, role: 0 },
        { address: accounts.associatedTokenProgram, role: 0 },
        { address: accounts.metadataProgram, role: 0 },
        { address: accounts.metadataUpdateAuth, role: 0 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([242, 29, 134, 48, 58, 110, 14, 60]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        bumps: types.OpenPositionWithMetadataBumps.toEncodable(args.bumps),
        tickLowerIndex: args.tickLowerIndex,
        tickUpperIndex: args.tickUpperIndex,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
