// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/instructions/swapV2.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.swapV2 = swapV2;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.u64("amount"),
    borsh.u64("otherAmountThreshold"),
    borsh.u128("sqrtPriceLimit"),
    borsh.bool("amountSpecifiedIsInput"),
    borsh.bool("aToB"),
    borsh.option(types.RemainingAccountsInfo.layout(), "remainingAccountsInfo"),
]);
function swapV2(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.tokenProgramA, role: 0 },
        { address: accounts.tokenProgramB, role: 0 },
        { address: accounts.memoProgram, role: 0 },
        {
            address: accounts.tokenAuthority.address,
            role: 2,
            signer: accounts.tokenAuthority,
        },
        { address: accounts.whirlpool, role: 1 },
        { address: accounts.tokenMintA, role: 0 },
        { address: accounts.tokenMintB, role: 0 },
        { address: accounts.tokenOwnerAccountA, role: 1 },
        { address: accounts.tokenVaultA, role: 1 },
        { address: accounts.tokenOwnerAccountB, role: 1 },
        { address: accounts.tokenVaultB, role: 1 },
        { address: accounts.tickArray0, role: 1 },
        { address: accounts.tickArray1, role: 1 },
        { address: accounts.tickArray2, role: 1 },
        { address: accounts.oracle, role: 1 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([43, 4, 237, 11, 26, 201, 30, 98]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        amount: args.amount,
        otherAmountThreshold: args.otherAmountThreshold,
        sqrtPriceLimit: args.sqrtPriceLimit,
        amountSpecifiedIsInput: args.amountSpecifiedIsInput,
        aToB: args.aToB,
        remainingAccountsInfo: (args.remainingAccountsInfo &&
            types.RemainingAccountsInfo.toEncodable(args.remainingAccountsInfo)) ||
            null,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
