// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/instructions/twoHopSwapV2.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.twoHopSwapV2 = twoHopSwapV2;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.u64("amount"),
    borsh.u64("otherAmountThreshold"),
    borsh.bool("amountSpecifiedIsInput"),
    borsh.bool("aToBOne"),
    borsh.bool("aToBTwo"),
    borsh.u128("sqrtPriceLimitOne"),
    borsh.u128("sqrtPriceLimitTwo"),
    borsh.option(types.RemainingAccountsInfo.layout(), "remainingAccountsInfo"),
]);
function twoHopSwapV2(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.whirlpoolOne, role: 1 },
        { address: accounts.whirlpoolTwo, role: 1 },
        { address: accounts.tokenMintInput, role: 0 },
        { address: accounts.tokenMintIntermediate, role: 0 },
        { address: accounts.tokenMintOutput, role: 0 },
        { address: accounts.tokenProgramInput, role: 0 },
        { address: accounts.tokenProgramIntermediate, role: 0 },
        { address: accounts.tokenProgramOutput, role: 0 },
        { address: accounts.tokenOwnerAccountInput, role: 1 },
        { address: accounts.tokenVaultOneInput, role: 1 },
        { address: accounts.tokenVaultOneIntermediate, role: 1 },
        { address: accounts.tokenVaultTwoIntermediate, role: 1 },
        { address: accounts.tokenVaultTwoOutput, role: 1 },
        { address: accounts.tokenOwnerAccountOutput, role: 1 },
        {
            address: accounts.tokenAuthority.address,
            role: 2,
            signer: accounts.tokenAuthority,
        },
        { address: accounts.tickArrayOne0, role: 1 },
        { address: accounts.tickArrayOne1, role: 1 },
        { address: accounts.tickArrayOne2, role: 1 },
        { address: accounts.tickArrayTwo0, role: 1 },
        { address: accounts.tickArrayTwo1, role: 1 },
        { address: accounts.tickArrayTwo2, role: 1 },
        { address: accounts.oracleOne, role: 1 },
        { address: accounts.oracleTwo, role: 1 },
        { address: accounts.memoProgram, role: 0 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([186, 143, 209, 29, 254, 2, 194, 117]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        amount: args.amount,
        otherAmountThreshold: args.otherAmountThreshold,
        amountSpecifiedIsInput: args.amountSpecifiedIsInput,
        aToBOne: args.aToBOne,
        aToBTwo: args.aToBTwo,
        sqrtPriceLimitOne: args.sqrtPriceLimitOne,
        sqrtPriceLimitTwo: args.sqrtPriceLimitTwo,
        remainingAccountsInfo: (args.remainingAccountsInfo &&
            types.RemainingAccountsInfo.toEncodable(args.remainingAccountsInfo)) ||
            null,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
