// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/instructions/twoHopSwap.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.twoHopSwap = twoHopSwap;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.u64("amount"),
    borsh.u64("otherAmountThreshold"),
    borsh.bool("amountSpecifiedIsInput"),
    borsh.bool("aToBOne"),
    borsh.bool("aToBTwo"),
    borsh.u128("sqrtPriceLimitOne"),
    borsh.u128("sqrtPriceLimitTwo"),
]);
function twoHopSwap(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.tokenProgram, role: 0 },
        {
            address: accounts.tokenAuthority.address,
            role: 2,
            signer: accounts.tokenAuthority,
        },
        { address: accounts.whirlpoolOne, role: 1 },
        { address: accounts.whirlpoolTwo, role: 1 },
        { address: accounts.tokenOwnerAccountOneA, role: 1 },
        { address: accounts.tokenVaultOneA, role: 1 },
        { address: accounts.tokenOwnerAccountOneB, role: 1 },
        { address: accounts.tokenVaultOneB, role: 1 },
        { address: accounts.tokenOwnerAccountTwoA, role: 1 },
        { address: accounts.tokenVaultTwoA, role: 1 },
        { address: accounts.tokenOwnerAccountTwoB, role: 1 },
        { address: accounts.tokenVaultTwoB, role: 1 },
        { address: accounts.tickArrayOne0, role: 1 },
        { address: accounts.tickArrayOne1, role: 1 },
        { address: accounts.tickArrayOne2, role: 1 },
        { address: accounts.tickArrayTwo0, role: 1 },
        { address: accounts.tickArrayTwo1, role: 1 },
        { address: accounts.tickArrayTwo2, role: 1 },
        { address: accounts.oracleOne, role: 0 },
        { address: accounts.oracleTwo, role: 0 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([195, 96, 237, 108, 68, 162, 219, 230]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        amount: args.amount,
        otherAmountThreshold: args.otherAmountThreshold,
        amountSpecifiedIsInput: args.amountSpecifiedIsInput,
        aToBOne: args.aToBOne,
        aToBTwo: args.aToBTwo,
        sqrtPriceLimitOne: args.sqrtPriceLimitOne,
        sqrtPriceLimitTwo: args.sqrtPriceLimitTwo,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
