// @kamino-finance/scope-sdk/dist/@codegen/scope/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopeChainError = exports.EmaType = exports.OracleType = exports.TokenTypes = exports.UpdateTokenMetadataMode = exports.MintToScopeChain = exports.TokenMetadata = exports.EmaTwap = exports.DatedPrice = exports.Price = exports.Hash = exports.AggregatorRound = exports.AggregatorAccountData = exports.SwitchboardDecimal = exports.PythLazerData = exports.State = exports.ValidatorSystem = exports.StakeSystem = exports.List = exports.LiqPool = exports.Fee = exports.MostRecentOfData = exports.DiscountToMaturityData = void 0;
const UpdateTokenMetadataMode = __importStar(require("./UpdateTokenMetadataMode"));
exports.UpdateTokenMetadataMode = UpdateTokenMetadataMode;
const TokenTypes = __importStar(require("./TokenTypes"));
exports.TokenTypes = TokenTypes;
const OracleType = __importStar(require("./OracleType"));
exports.OracleType = OracleType;
const EmaType = __importStar(require("./EmaType"));
exports.EmaType = EmaType;
const ScopeChainError = __importStar(require("./ScopeChainError"));
exports.ScopeChainError = ScopeChainError;
var DiscountToMaturityData_1 = require("./DiscountToMaturityData");
Object.defineProperty(exports, "DiscountToMaturityData", { enumerable: true, get: function () { return DiscountToMaturityData_1.DiscountToMaturityData; } });
var MostRecentOfData_1 = require("./MostRecentOfData");
Object.defineProperty(exports, "MostRecentOfData", { enumerable: true, get: function () { return MostRecentOfData_1.MostRecentOfData; } });
var Fee_1 = require("./Fee");
Object.defineProperty(exports, "Fee", { enumerable: true, get: function () { return Fee_1.Fee; } });
var LiqPool_1 = require("./LiqPool");
Object.defineProperty(exports, "LiqPool", { enumerable: true, get: function () { return LiqPool_1.LiqPool; } });
var List_1 = require("./List");
Object.defineProperty(exports, "List", { enumerable: true, get: function () { return List_1.List; } });
var StakeSystem_1 = require("./StakeSystem");
Object.defineProperty(exports, "StakeSystem", { enumerable: true, get: function () { return StakeSystem_1.StakeSystem; } });
var ValidatorSystem_1 = require("./ValidatorSystem");
Object.defineProperty(exports, "ValidatorSystem", { enumerable: true, get: function () { return ValidatorSystem_1.ValidatorSystem; } });
var State_1 = require("./State");
Object.defineProperty(exports, "State", { enumerable: true, get: function () { return State_1.State; } });
var PythLazerData_1 = require("./PythLazerData");
Object.defineProperty(exports, "PythLazerData", { enumerable: true, get: function () { return PythLazerData_1.PythLazerData; } });
var SwitchboardDecimal_1 = require("./SwitchboardDecimal");
Object.defineProperty(exports, "SwitchboardDecimal", { enumerable: true, get: function () { return SwitchboardDecimal_1.SwitchboardDecimal; } });
var AggregatorAccountData_1 = require("./AggregatorAccountData");
Object.defineProperty(exports, "AggregatorAccountData", { enumerable: true, get: function () { return AggregatorAccountData_1.AggregatorAccountData; } });
var AggregatorRound_1 = require("./AggregatorRound");
Object.defineProperty(exports, "AggregatorRound", { enumerable: true, get: function () { return AggregatorRound_1.AggregatorRound; } });
var Hash_1 = require("./Hash");
Object.defineProperty(exports, "Hash", { enumerable: true, get: function () { return Hash_1.Hash; } });
var Price_1 = require("./Price");
Object.defineProperty(exports, "Price", { enumerable: true, get: function () { return Price_1.Price; } });
var DatedPrice_1 = require("./DatedPrice");
Object.defineProperty(exports, "DatedPrice", { enumerable: true, get: function () { return DatedPrice_1.DatedPrice; } });
var EmaTwap_1 = require("./EmaTwap");
Object.defineProperty(exports, "EmaTwap", { enumerable: true, get: function () { return EmaTwap_1.EmaTwap; } });
var TokenMetadata_1 = require("./TokenMetadata");
Object.defineProperty(exports, "TokenMetadata", { enumerable: true, get: function () { return TokenMetadata_1.TokenMetadata; } });
var MintToScopeChain_1 = require("./MintToScopeChain");
Object.defineProperty(exports, "MintToScopeChain", { enumerable: true, get: function () { return MintToScopeChain_1.MintToScopeChain; } });
