// @kamino-finance/scope-sdk/dist/@codegen/scope/types/PythLazerData.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PythLazerData = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class PythLazerData {
    feedId;
    exponent;
    confidenceFactor;
    constructor(fields) {
        this.feedId = fields.feedId;
        this.exponent = fields.exponent;
        this.confidenceFactor = fields.confidenceFactor;
    }
    static layout(property) {
        return borsh.struct([
            borsh.u16("feedId"),
            borsh.u8("exponent"),
            borsh.u32("confidenceFactor"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new PythLazerData({
            feedId: obj.feedId,
            exponent: obj.exponent,
            confidenceFactor: obj.confidenceFactor,
        });
    }
    static toEncodable(fields) {
        return {
            feedId: fields.feedId,
            exponent: fields.exponent,
            confidenceFactor: fields.confidenceFactor,
        };
    }
    toJSON() {
        return {
            feedId: this.feedId,
            exponent: this.exponent,
            confidenceFactor: this.confidenceFactor,
        };
    }
    static fromJSON(obj) {
        return new PythLazerData({
            feedId: obj.feedId,
            exponent: obj.exponent,
            confidenceFactor: obj.confidenceFactor,
        });
    }
    toEncodable() {
        return PythLazerData.toEncodable(this);
    }
}
exports.PythLazerData = PythLazerData;
