// @kamino-finance/scope-sdk/dist/@codegen/scope/types/ValidatorSystem.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorSystem = void 0;
const kit_1 = require("@solana/kit"); // eslint-disable-line @typescript-eslint/no-unused-vars
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
const utils_1 = require("../utils");
class ValidatorSystem {
    validatorList;
    managerAuthority;
    totalValidatorScore;
    /** sum of all active lamports staked */
    totalActiveBalance;
    /** allow & auto-add validator when a user deposits a stake-account of a non-listed validator */
    autoAddValidatorEnabled;
    constructor(fields) {
        this.validatorList = new types.List({ ...fields.validatorList });
        this.managerAuthority = fields.managerAuthority;
        this.totalValidatorScore = fields.totalValidatorScore;
        this.totalActiveBalance = fields.totalActiveBalance;
        this.autoAddValidatorEnabled = fields.autoAddValidatorEnabled;
    }
    static layout(property) {
        return borsh.struct([
            types.List.layout("validatorList"),
            (0, utils_1.borshAddress)("managerAuthority"),
            borsh.u32("totalValidatorScore"),
            borsh.u64("totalActiveBalance"),
            borsh.u8("autoAddValidatorEnabled"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ValidatorSystem({
            validatorList: types.List.fromDecoded(obj.validatorList),
            managerAuthority: obj.managerAuthority,
            totalValidatorScore: obj.totalValidatorScore,
            totalActiveBalance: obj.totalActiveBalance,
            autoAddValidatorEnabled: obj.autoAddValidatorEnabled,
        });
    }
    static toEncodable(fields) {
        return {
            validatorList: types.List.toEncodable(fields.validatorList),
            managerAuthority: fields.managerAuthority,
            totalValidatorScore: fields.totalValidatorScore,
            totalActiveBalance: fields.totalActiveBalance,
            autoAddValidatorEnabled: fields.autoAddValidatorEnabled,
        };
    }
    toJSON() {
        return {
            validatorList: this.validatorList.toJSON(),
            managerAuthority: this.managerAuthority,
            totalValidatorScore: this.totalValidatorScore,
            totalActiveBalance: this.totalActiveBalance.toString(),
            autoAddValidatorEnabled: this.autoAddValidatorEnabled,
        };
    }
    static fromJSON(obj) {
        return new ValidatorSystem({
            validatorList: types.List.fromJSON(obj.validatorList),
            managerAuthority: (0, kit_1.address)(obj.managerAuthority),
            totalValidatorScore: obj.totalValidatorScore,
            totalActiveBalance: new bn_js_1.default(obj.totalActiveBalance),
            autoAddValidatorEnabled: obj.autoAddValidatorEnabled,
        });
    }
    toEncodable() {
        return ValidatorSystem.toEncodable(this);
    }
}
exports.ValidatorSystem = ValidatorSystem;
