// @mrgnlabs/marginfi-client-v2/dist/services/bank/types/bank.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmodeImpactStatus = exports.EmodeTag = exports.EmodeFlags = exports.EmodeEntryFlags = exports.BankConfigFlag = exports.AssetTag = exports.OracleSetup = exports.OperationalState = exports.RiskTier = void 0;
var RiskTier;
(function (RiskTier) {
    RiskTier["Collateral"] = "Collateral";
    RiskTier["Isolated"] = "Isolated";
})(RiskTier || (exports.RiskTier = RiskTier = {}));
var OperationalState;
(function (OperationalState) {
    OperationalState["Paused"] = "Paused";
    OperationalState["Operational"] = "Operational";
    OperationalState["ReduceOnly"] = "ReduceOnly";
})(OperationalState || (exports.OperationalState = OperationalState = {}));
var OracleSetup;
(function (OracleSetup) {
    OracleSetup["None"] = "None";
    OracleSetup["PythLegacy"] = "PythLegacy";
    OracleSetup["SwitchboardV2"] = "SwitchboardV2";
    OracleSetup["PythPushOracle"] = "PythPushOracle";
    OracleSetup["SwitchboardPull"] = "SwitchboardPull";
    OracleSetup["StakedWithPythPush"] = "StakedWithPythPush";
})(OracleSetup || (exports.OracleSetup = OracleSetup = {}));
var AssetTag;
(function (AssetTag) {
    AssetTag[AssetTag["DEFAULT"] = 0] = "DEFAULT";
    AssetTag[AssetTag["SOL"] = 1] = "SOL";
    AssetTag[AssetTag["STAKED"] = 2] = "STAKED";
})(AssetTag || (exports.AssetTag = AssetTag = {}));
var BankConfigFlag;
(function (BankConfigFlag) {
    BankConfigFlag[BankConfigFlag["PYTH_MIGRATED"] = 1] = "PYTH_MIGRATED";
})(BankConfigFlag || (exports.BankConfigFlag = BankConfigFlag = {}));
/**
 * Bitwise flags for EMode entry
 */
var EmodeEntryFlags;
(function (EmodeEntryFlags) {
    /**
     * If set, isolated banks with this tag also benefit.
     * If not set, isolated banks continue to offer zero collateral, even if they use this tag.
     * (NOT YET IMPLEMENTED)
     */
    EmodeEntryFlags[EmodeEntryFlags["APPLIES_TO_ISOLATED"] = 1] = "APPLIES_TO_ISOLATED";
    /** Reserved for future use */
    EmodeEntryFlags[EmodeEntryFlags["RESERVED_1"] = 2] = "RESERVED_1";
    EmodeEntryFlags[EmodeEntryFlags["RESERVED_2"] = 4] = "RESERVED_2";
    EmodeEntryFlags[EmodeEntryFlags["RESERVED_3"] = 8] = "RESERVED_3";
    EmodeEntryFlags[EmodeEntryFlags["RESERVED_4"] = 16] = "RESERVED_4";
    EmodeEntryFlags[EmodeEntryFlags["RESERVED_5"] = 32] = "RESERVED_5";
})(EmodeEntryFlags || (exports.EmodeEntryFlags = EmodeEntryFlags = {}));
/**
 * Bitwise flags for EMode settings
 */
var EmodeFlags;
(function (EmodeFlags) {
    /** If set, at least one entry is configured */
    EmodeFlags[EmodeFlags["EMODE_ON"] = 1] = "EMODE_ON";
    /** Reserved for future use */
    EmodeFlags[EmodeFlags["RESERVED_1"] = 2] = "RESERVED_1";
    EmodeFlags[EmodeFlags["RESERVED_2"] = 4] = "RESERVED_2";
    EmodeFlags[EmodeFlags["RESERVED_3"] = 8] = "RESERVED_3";
})(EmodeFlags || (exports.EmodeFlags = EmodeFlags = {}));
var EmodeTag;
(function (EmodeTag) {
    EmodeTag[EmodeTag["UNSET"] = 0] = "UNSET";
    EmodeTag[EmodeTag["SOL"] = 501] = "SOL";
    EmodeTag[EmodeTag["LST_T1"] = 1571] = "LST_T1";
    EmodeTag[EmodeTag["LST_T2"] = 1572] = "LST_T2";
})(EmodeTag || (exports.EmodeTag = EmodeTag = {}));
var EmodeImpactStatus;
(function (EmodeImpactStatus) {
    EmodeImpactStatus[EmodeImpactStatus["ActivateEmode"] = 0] = "ActivateEmode";
    EmodeImpactStatus[EmodeImpactStatus["ExtendEmode"] = 1] = "ExtendEmode";
    EmodeImpactStatus[EmodeImpactStatus["IncreaseEmode"] = 2] = "IncreaseEmode";
    EmodeImpactStatus[EmodeImpactStatus["ReduceEmode"] = 3] = "ReduceEmode";
    EmodeImpactStatus[EmodeImpactStatus["RemoveEmode"] = 4] = "RemoveEmode";
    EmodeImpactStatus[EmodeImpactStatus["InactiveEmode"] = 5] = "InactiveEmode";
})(EmodeImpactStatus || (exports.EmodeImpactStatus = EmodeImpactStatus = {}));
