// @mrgnlabs/marginfi-client-v2/dist/vendor/pyth_crank/pyth.consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VAA_START = exports.VAA_SPLIT_INDEX = exports.CLOSE_ENCODED_VAA_COMPUTE_BUDGET = exports.WRITE_ENCODED_VAA_COMPUTE_BUDGET = exports.INIT_ENCODED_VAA_COMPUTE_BUDGET = exports.UPDATE_PRICE_FEED_COMPUTE_BUDGET = exports.POST_TWAP_UPDATE_COMPUTE_BUDGET = exports.POST_UPDATE_COMPUTE_BUDGET = exports.POST_UPDATE_ATOMIC_COMPUTE_BUDGET = exports.VERIFY_ENCODED_VAA_COMPUTE_BUDGET = exports.TWAP_MESSAGE_VARIANT = exports.PRICE_FEED_MESSAGE_VARIANT = exports.KECCAK160_HASH_SIZE = exports.MINOR_VERSION = exports.MAJOR_VERSION = exports.ACCUMULATOR_MAGIC = exports.DEFAULT_PUSH_ORACLE_PROGRAM_ID = exports.DEFAULT_WORMHOLE_PROGRAM_ID = exports.DEFAULT_RECEIVER_PROGRAM_ID = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.DEFAULT_RECEIVER_PROGRAM_ID = new web3_js_1.PublicKey("rec5EKMGg6MxZYaMdyBfgwp4d5rB9T1VQH5pJv5LtFJ");
exports.DEFAULT_WORMHOLE_PROGRAM_ID = new web3_js_1.PublicKey("HDwcJBJXjL9FpJ7UBsYBtaDjsBUhuLCUYoz3zr8SWWaQ");
exports.DEFAULT_PUSH_ORACLE_PROGRAM_ID = new web3_js_1.PublicKey("pythWSnswVUd12oZpeFP8e9CVaEqJg25g1Vtc2biRsT");
exports.ACCUMULATOR_MAGIC = "504e4155";
exports.MAJOR_VERSION = 1;
exports.MINOR_VERSION = 0;
exports.KECCAK160_HASH_SIZE = 20;
exports.PRICE_FEED_MESSAGE_VARIANT = 0;
exports.TWAP_MESSAGE_VARIANT = 1;
/**
 * A hard-coded budget for the compute units required for the `verifyEncodedVaa` instruction in the Wormhole program.
 */
exports.VERIFY_ENCODED_VAA_COMPUTE_BUDGET = 350000;
/**
 * A hard-coded budget for the compute units required for the `postUpdateAtomic` instruction in the Pyth Solana Receiver program.
 */
exports.POST_UPDATE_ATOMIC_COMPUTE_BUDGET = 170000;
/**
 * A hard-coded budget for the compute units required for the `postUpdate` instruction in the Pyth Solana Receiver program.
 */
exports.POST_UPDATE_COMPUTE_BUDGET = 35000;
/**
 * A hard-coded budget for the compute units required for the `postTwapUpdate` instruction in the Pyth Solana Receiver program.
 */
exports.POST_TWAP_UPDATE_COMPUTE_BUDGET = 50_000;
/**
 * A hard-coded budget for the compute units required for the `updatePriceFeed` instruction in the Pyth Push Oracle program.
 */
exports.UPDATE_PRICE_FEED_COMPUTE_BUDGET = 55000;
/**
 * A hard-coded budget for the compute units required for the `initEncodedVaa` instruction in the Wormhole program.
 */
exports.INIT_ENCODED_VAA_COMPUTE_BUDGET = 3000;
/**
 * A hard-coded budget for the compute units required for the `writeEncodedVaa` instruction in the Wormhole program.
 */
exports.WRITE_ENCODED_VAA_COMPUTE_BUDGET = 3000;
/**
 * A hard-coded budget for the compute units required for the `closeEncodedVaa` instruction in the Wormhole program.
 */
exports.CLOSE_ENCODED_VAA_COMPUTE_BUDGET = 30000;
exports.VAA_SPLIT_INDEX = 721;
exports.VAA_START = 46;
