// @kamino-finance/scope-sdk/dist/@codegen/scope/instructions/updateMapping.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.updateMapping = updateMapping;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.u16("token"),
    borsh.u8("priceType"),
    borsh.bool("twapEnabled"),
    borsh.u16("twapSource"),
    borsh.u16("refPriceIndex"),
    borsh.str("feedName"),
    borsh.array(borsh.u8(), 20, "genericData"),
]);
function updateMapping(args, accounts, programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.admin.address, role: 2, signer: accounts.admin },
        { address: accounts.configuration, role: 0 },
        { address: accounts.oracleMappings, role: 1 },
        (0, kit_1.isSome)(accounts.priceInfo)
            ? { address: accounts.priceInfo.value, role: 0 }
            : { address: programAddress, role: 0 },
    ];
    const identifier = Buffer.from([56, 102, 90, 236, 243, 21, 185, 105]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        token: args.token,
        priceType: args.priceType,
        twapEnabled: args.twapEnabled,
        twapSource: args.twapSource,
        refPriceIndex: args.refPriceIndex,
        feedName: args.feedName,
        genericData: args.genericData,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
