// @kamino-finance/farms-sdk/dist/@codegen/farms/types/DatedPrice.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatedPrice = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class DatedPrice {
    price;
    lastUpdatedSlot;
    unixTimestamp;
    reserved;
    reserved2;
    index;
    constructor(fields) {
        this.price = new types.Price({ ...fields.price });
        this.lastUpdatedSlot = fields.lastUpdatedSlot;
        this.unixTimestamp = fields.unixTimestamp;
        this.reserved = fields.reserved;
        this.reserved2 = fields.reserved2;
        this.index = fields.index;
    }
    static layout(property) {
        return borsh.struct([
            types.Price.layout("price"),
            borsh.u64("lastUpdatedSlot"),
            borsh.u64("unixTimestamp"),
            borsh.array(borsh.u64(), 2, "reserved"),
            borsh.array(borsh.u16(), 3, "reserved2"),
            borsh.u16("index"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new DatedPrice({
            price: types.Price.fromDecoded(obj.price),
            lastUpdatedSlot: obj.lastUpdatedSlot,
            unixTimestamp: obj.unixTimestamp,
            reserved: obj.reserved,
            reserved2: obj.reserved2,
            index: obj.index,
        });
    }
    static toEncodable(fields) {
        return {
            price: types.Price.toEncodable(fields.price),
            lastUpdatedSlot: fields.lastUpdatedSlot,
            unixTimestamp: fields.unixTimestamp,
            reserved: fields.reserved,
            reserved2: fields.reserved2,
            index: fields.index,
        };
    }
    toJSON() {
        return {
            price: this.price.toJSON(),
            lastUpdatedSlot: this.lastUpdatedSlot.toString(),
            unixTimestamp: this.unixTimestamp.toString(),
            reserved: this.reserved.map((item) => item.toString()),
            reserved2: this.reserved2,
            index: this.index,
        };
    }
    static fromJSON(obj) {
        return new DatedPrice({
            price: types.Price.fromJSON(obj.price),
            lastUpdatedSlot: new bn_js_1.default(obj.lastUpdatedSlot),
            unixTimestamp: new bn_js_1.default(obj.unixTimestamp),
            reserved: obj.reserved.map((item) => new bn_js_1.default(item)),
            reserved2: obj.reserved2,
            index: obj.index,
        });
    }
    toEncodable() {
        return DatedPrice.toEncodable(this);
    }
}
exports.DatedPrice = DatedPrice;
