// @kamino-finance/farms-sdk/dist/@codegen/farms/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Price = exports.DatedPrice = exports.TokenInfo = exports.TimeUnit = exports.RewardType = exports.RewardScheduleCurve = exports.RewardPerTimeUnitPoint = exports.RewardInfo = exports.LockingMode = exports.GlobalConfigOption = exports.FarmConfigOption = void 0;
const FarmConfigOption = __importStar(require("./FarmConfigOption"));
exports.FarmConfigOption = FarmConfigOption;
const GlobalConfigOption = __importStar(require("./GlobalConfigOption"));
exports.GlobalConfigOption = GlobalConfigOption;
const LockingMode = __importStar(require("./LockingMode"));
exports.LockingMode = LockingMode;
const RewardType = __importStar(require("./RewardType"));
exports.RewardType = RewardType;
const TimeUnit = __importStar(require("./TimeUnit"));
exports.TimeUnit = TimeUnit;
var RewardInfo_1 = require("./RewardInfo");
Object.defineProperty(exports, "RewardInfo", { enumerable: true, get: function () { return RewardInfo_1.RewardInfo; } });
var RewardPerTimeUnitPoint_1 = require("./RewardPerTimeUnitPoint");
Object.defineProperty(exports, "RewardPerTimeUnitPoint", { enumerable: true, get: function () { return RewardPerTimeUnitPoint_1.RewardPerTimeUnitPoint; } });
var RewardScheduleCurve_1 = require("./RewardScheduleCurve");
Object.defineProperty(exports, "RewardScheduleCurve", { enumerable: true, get: function () { return RewardScheduleCurve_1.RewardScheduleCurve; } });
var TokenInfo_1 = require("./TokenInfo");
Object.defineProperty(exports, "TokenInfo", { enumerable: true, get: function () { return TokenInfo_1.TokenInfo; } });
var DatedPrice_1 = require("./DatedPrice");
Object.defineProperty(exports, "DatedPrice", { enumerable: true, get: function () { return DatedPrice_1.DatedPrice; } });
var Price_1 = require("./Price");
Object.defineProperty(exports, "Price", { enumerable: true, get: function () { return Price_1.Price; } });
