// @kamino-finance/farms-sdk/dist/@codegen/farms/types/TokenInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenInfo = void 0;
const kit_1 = require("@solana/kit"); // eslint-disable-line @typescript-eslint/no-unused-vars
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
const utils_1 = require("../utils");
class TokenInfo {
    mint;
    decimals;
    tokenProgram;
    padding;
    constructor(fields) {
        this.mint = fields.mint;
        this.decimals = fields.decimals;
        this.tokenProgram = fields.tokenProgram;
        this.padding = fields.padding;
    }
    static layout(property) {
        return borsh.struct([
            (0, utils_1.borshAddress)("mint"),
            borsh.u64("decimals"),
            (0, utils_1.borshAddress)("tokenProgram"),
            borsh.array(borsh.u64(), 6, "padding"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new TokenInfo({
            mint: obj.mint,
            decimals: obj.decimals,
            tokenProgram: obj.tokenProgram,
            padding: obj.padding,
        });
    }
    static toEncodable(fields) {
        return {
            mint: fields.mint,
            decimals: fields.decimals,
            tokenProgram: fields.tokenProgram,
            padding: fields.padding,
        };
    }
    toJSON() {
        return {
            mint: this.mint,
            decimals: this.decimals.toString(),
            tokenProgram: this.tokenProgram,
            padding: this.padding.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new TokenInfo({
            mint: (0, kit_1.address)(obj.mint),
            decimals: new bn_js_1.default(obj.decimals),
            tokenProgram: (0, kit_1.address)(obj.tokenProgram),
            padding: obj.padding.map((item) => new bn_js_1.default(item)),
        });
    }
    toEncodable() {
        return TokenInfo.toEncodable(this);
    }
}
exports.TokenInfo = TokenInfo;
