// @switchboard-xyz/common/dist/esm/networks/ethereum.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SWITCHBOARD_ETHEREUM_CONFIG = exports.SWITCHBOARD_ETHEREUM_TESTNET_CONFIG = exports.SWITCHBOARD_ETHEREUM_MAINNET_CONFIG = void 0;
exports.SWITCHBOARD_ETHEREUM_MAINNET_CONFIG = {
    chain: 'ethereum',
    chainId: 1,
    networkName: 'Mainnet',
    address: '',
    sbPushOracle: '',
    metadata: {
        defaultRpcUrl: 'https://ethereum.publicnode.com',
        defaultExplorer: 'https://etherscan.io',
    },
    queues: [],
    attestationQueues: [
        {
            name: 'SwitchboardLabs Attestation Queue',
            address: '',
        },
    ],
};
exports.SWITCHBOARD_ETHEREUM_TESTNET_CONFIG = {
    chain: 'ethereum',
    chainId: 5,
    networkName: 'Goerli',
    address: '',
    sbPushOracle: '',
    metadata: {
        defaultRpcUrl: 'https://ethereum-goerli.publicnode.com',
        defaultExplorer: 'https://goerli.etherscan.io',
    },
    queues: [],
    attestationQueues: [
        {
            name: 'SwitchboardLabs Attestation Queue',
            address: '',
        },
    ],
};
/**
@defaultValue the default Aurora configuration

```json
{
  "mainnet": {
    "chain": "ethereum",
    "chainId": 1,
    "networkName": "Mainnet",
    "address": "",
    "sbPushOracle": "",
    "metadata": {
      "defaultRpcUrl": "https://ethereum.publicnode.com",
      "defaultExplorer": "https://etherscan.io"
    },
    "queues": [],
    "attestationQueues": [
      {
        "name": "SwitchboardLabs Attestation Queue",
        "address": ""
      }
    ]
  },
  "testnet": {
    "chain": "ethereum",
    "chainId": 5,
    "networkName": "Goerli",
    "address": "",
    "sbPushOracle": "",
    "metadata": {
      "defaultRpcUrl": "https://ethereum-goerli.publicnode.com",
      "defaultExplorer": "https://goerli.etherscan.io"
    },
    "queues": [],
    "attestationQueues": [
      {
        "name": "SwitchboardLabs Attestation Queue",
        "address": ""
      }
    ]
  }
}
```
 */
exports.SWITCHBOARD_ETHEREUM_CONFIG = {
    mainnet: exports.SWITCHBOARD_ETHEREUM_MAINNET_CONFIG,
    testnet: exports.SWITCHBOARD_ETHEREUM_TESTNET_CONFIG,
};
