// @hubbleprotocol/hubble-config/dist/constants/configs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HUBBLE_CONFIGS = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.HUBBLE_CONFIGS = [
    {
        cluster: 'mainnet-beta',
        kamino: {
            mints: [
                { address: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'), scopeToken: 'USDH' },
                { address: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'), scopeToken: 'USDC' },
                { address: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'), scopeToken: 'SOL' },
                { address: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'), scopeToken: 'MSOL' },
                { address: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'), scopeToken: 'STSOL' },
                { address: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'), scopeToken: 'USDT' },
                { address: new web3_js_1.PublicKey('5oVNBeEEQvYi1cX3ir8Dx5n1P7pdxydbGF2X4TxVusJm'), scopeToken: 'scnSOL' },
                { address: new web3_js_1.PublicKey('9iLH8T7zoWhY7sBmj1WK9ENbWdS1nL8n9wAxaeRitTa6'), scopeToken: 'USH' },
                { address: new web3_js_1.PublicKey('7kbnvuGBxxj8AG9qp8Scn56muWGaRaFqxg1FsRp3PaFT'), scopeToken: 'UXD' },
                { address: new web3_js_1.PublicKey('7Q2afV64in6N6SeZsAAB81TJzwDoD6zpqmHkzi9Dcavn'), scopeToken: 'JSOL' },
                { address: new web3_js_1.PublicKey('DUSTawucrTsGU8hcqRdHDCbuYhCPADMLM2VcCb8VnFnQ'), scopeToken: 'DUST' },
                { address: new web3_js_1.PublicKey('USDrbBQwQbQ2oWHUPfA8QBHcyVxKUq1xHyXsSLKdUq2'), scopeToken: 'USDR' },
                { address: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'), scopeToken: 'ETH' },
                { address: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'), scopeToken: 'BTC' },
                { address: new web3_js_1.PublicKey('HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6'), scopeToken: 'HBB' },
                { address: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'), scopeToken: 'JITOSOL' },
                { address: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'), scopeToken: 'RAY' },
                { address: new web3_js_1.PublicKey('DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263'), scopeToken: 'BONK' },
                { address: new web3_js_1.PublicKey('7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU'), scopeToken: 'SAMO' },
                { address: new web3_js_1.PublicKey('LAinEtNLgpmCP9Rvsf5Hn8W6EhNiKLZQti1xfWMLy6X'), scopeToken: 'LaineSOL' },
                { address: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'), scopeToken: 'bSOL' },
                { address: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'), scopeToken: 'HADES' },
                { address: new web3_js_1.PublicKey('orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE'), scopeToken: 'ORCA' },
                { address: new web3_js_1.PublicKey('EjmyN6qEC1Tf1JxiG1ae7UTJhUxSwk1TCWNWqxWV4J6o'), scopeToken: 'DAI' },
                { address: new web3_js_1.PublicKey('5PmpMzWjraf3kSsGEKtqdUsCoLhptg4yriZ17LKKdBBy'), scopeToken: 'HDG' },
                { address: new web3_js_1.PublicKey('HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p'), scopeToken: 'LDO' },
                { address: new web3_js_1.PublicKey('MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey'), scopeToken: 'MNDE' },
                { address: new web3_js_1.PublicKey('ratioMVg27rSZbSvBopUvsdrGUzeALUfFma61mpxc8J'), scopeToken: 'RATIO' },
                { address: new web3_js_1.PublicKey('5oVNBeEEQvYi1cX3ir8Dx5n1P7pdxydbGF2X4TxVusJm'), scopeToken: 'SCNSOL' },
                { address: new web3_js_1.PublicKey('UXPhBoR3qG4UCiGNJfV7MqhHyFqKN68g45GoYvAeL2M'), scopeToken: 'UXP' },
                { address: new web3_js_1.PublicKey('C7RpHABBJjfkmizgnR2MRbpfEF454Rb18XC6RmBC8cGg'), scopeToken: 'kSOLSTSOLOrca' },
                { address: new web3_js_1.PublicKey('5BmZgW7dk1kximGfn7MPvDigp3yRmgT64jS9Skdq4nPY'), scopeToken: 'kUSDHUSDCOrca' },
                { address: new web3_js_1.PublicKey('3NZ9JMVBmGAqocybic2c7LQCJScmgsAZ6vQqTDzcqmJh'), scopeToken: 'wBTC' },
                { address: new web3_js_1.PublicKey('RLBxxFkseAZ4RgJH3Sqn8jXxhmGoz9jWxDNJMh8pL7a'), scopeToken: 'RLB' },
                { address: new web3_js_1.PublicKey('CgnTSoL3DgY9SFHxcLj6CgCgKKoTBr6tp4CPAEWy25DE'), scopeToken: 'CGNTSOL' },
                { address: new web3_js_1.PublicKey('HxhWkVpk5NS4Ltg5nij2G671CKXFRKPK8vy271Ub4uEK'), scopeToken: 'HXRO' },
                { address: new web3_js_1.PublicKey('A9mUU4qviSctJVPJdBJWkb28deg915LYJKrzQ19ji3FM'), scopeToken: 'USDCet' },
                { address: new web3_js_1.PublicKey('hntyVP6YFm1Hg25TN9WGLqM12b8TQmcknKrdu1oxWux'), scopeToken: 'HNT' },
                { address: new web3_js_1.PublicKey('mb1eu7TzEc71KxDpsmsKoucSSuuoGLv1drys1oP2jh6'), scopeToken: 'MOBILE' },
                { address: new web3_js_1.PublicKey('iotEVVZLEywoTn1QdwNPddxPWszn3zFhEot3MfL9fns'), scopeToken: 'IOT' },
            ],
            whirlpoolsToSnapshot: [
                {
                    address: new web3_js_1.PublicKey('5Z66YYYaTmmx1R4mATAGLSc8aV4Vfy5tNdJQzk1GP9RF'),
                    collateralA: 'ORCA',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('HJPjoWUrhoZzkNfRpHuieeFk9WcZWjwy6PBjZ81ngndJ'),
                    collateralA: 'SOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('AiMZS5U3JMvpdvsr1KeaMiS354Z1DeSg5XjA4yYRxtFf'),
                    collateralA: 'MSOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('HQcY5n2zP6rW74fyFEhWeBd3LnJpBcZechkvJpmdb8cx'),
                    collateralA: 'SOL',
                    collateralB: 'MSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                },
                {
                    address: new web3_js_1.PublicKey('7A1R3L7AxcxuZHMJjFgskKGeBR5Rwst3Ai5bv5uAWZFG'),
                    collateralA: 'MSOL',
                    collateralB: 'USDT',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'),
                },
                {
                    address: new web3_js_1.PublicKey('AXtdSZ2mpagmtM5aipN5kV9CyGBA8dxhSBnqMRp7UpdN'),
                    collateralA: 'STSOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('2AEWSvUds1wsufnsDPCXjFsJCMJH5SNNm7fSF4kxys9a'),
                    collateralA: 'SOL',
                    collateralB: 'STSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                },
                {
                    address: new web3_js_1.PublicKey('FAbwB8VgdgSGty5E8dnmNbu5PZnQcvSuLnboJVpw1Rty'),
                    collateralA: 'STSOL',
                    collateralB: 'USDT',
                    mintA: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                    mintB: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'),
                },
                {
                    address: new web3_js_1.PublicKey('E5KuHFnU2VuuZFKeghbTLazgxeni4dhQ7URE4oBtJju2'),
                    collateralA: 'ETH',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('ErSQss3jrqDpQoLEYvo6onzjsi6zm4Sjpoz1pjqz2o6D'),
                    collateralA: 'BTC',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('6jZQFLhSAzTYfo33MSQYvwKvZYwxat8kUa29Mz63oHN9'),
                    collateralA: 'MSOL',
                    collateralB: 'ETH',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                },
                {
                    address: new web3_js_1.PublicKey('H1fREbTWrkhCs2stH3tKANWJepmqeF9hww4nWRYrM7uV'),
                    collateralA: 'MSOL',
                    collateralB: 'BTC',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'),
                },
                {
                    address: new web3_js_1.PublicKey('9vqYJjDUFecLL2xPUC4Rc7hyCtZ6iJ4mDiVZX7aFXoAe'),
                    collateralA: 'SAMO',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('J7qn7AvZ4QK9qT6BikVBKA3hUp89Lg9UkqJmXZQEjRxq'),
                    collateralA: 'STSOL',
                    collateralB: 'BTC',
                    mintA: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                    mintB: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'),
                },
                {
                    address: new web3_js_1.PublicKey('Db4AyCBKyH5pcCxJuvQzWfFsVsSH6rM9sm21HbA4WU5'),
                    collateralA: 'STSOL',
                    collateralB: 'ETH',
                    mintA: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                    mintB: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                },
                {
                    address: new web3_js_1.PublicKey('ApLVWYdXzjoDhBHeRx6SnbFWv4MYjFMih5FijDQUJk5R'),
                    collateralA: 'USH',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('9iLH8T7zoWhY7sBmj1WK9ENbWdS1nL8n9wAxaeRitTa6'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('GpqMSH1YM6oPmJ5xxEE2KfePf7uf5rXFbTW2TnxicRj6'),
                    collateralA: 'USDH',
                    collateralB: 'STSOL',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                },
                {
                    address: new web3_js_1.PublicKey('Fvtf8VCjnkqbETA6KtyHYqHm26ut6w184Jqm4MQjPvv7'),
                    collateralA: 'USDH',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('F7qyox3dAegTNfd8oBQD97LuCHWzQ9hSjbsF7Kv8kTNc'),
                    collateralA: 'ABR',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('8Ej6U2za4nwsCUyEXidzG9aWcd69PPtrA3hnVYWgJUx1'),
                    collateralA: 'SOL',
                    collateralB: 'JSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('7Q2afV64in6N6SeZsAAB81TJzwDoD6zpqmHkzi9Dcavn'),
                },
                {
                    address: new web3_js_1.PublicKey('HPh1hkQFYchT3bfeYQTx4uLRXuj7eRMjdJZJjtRY6gDq'),
                    collateralA: 'CMFI',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('5Wsd311hY8NXQhkt9cWHwTnqafk7BGEbLu8Py3DSnPAr'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('xLeM685ZkAZYXDKwzfB9eyszw1oELAmzCn45amKscmh'),
                    collateralA: 'MEAN',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('MEANeD3XDdUmNMsRGjASkSWdC8prLYsoRJ61pPeHctD'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('2fPqLazJ91cKRoZoH9XHC2YQnRq2wBZZpbcNx4HYoQXY'),
                    collateralA: 'STEP',
                    collateralB: 'USDH',
                    mintA: new web3_js_1.PublicKey('StepAscQoEioFxxWGnh2sLBDFp9d8rvKz2Yp39iDpyT'),
                    mintB: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                },
                {
                    address: new web3_js_1.PublicKey('5AX84BrKDWpUZ87fbQpkm7XsSx8bWwANePRmAx17tQjM'),
                    collateralA: 'SOL',
                    collateralB: 'STEP',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('StepAscQoEioFxxWGnh2sLBDFp9d8rvKz2Yp39iDpyT'),
                },
                {
                    address: new web3_js_1.PublicKey('4RGgiAMx3YED2nPTvTrVuy7u93pw9AtJBQjGJ4gLiTiz'),
                    collateralA: 'KI',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('kiTkNc7nYAu8dLKjQFYPx3BqdzwagZGBUrcb7d4nbN5'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('CpbNcvqxXdQyQ3SRPDPSwLfzd9sgzBm8TjRFsfJL7Pf4'),
                    collateralA: 'DUST',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('DUSTawucrTsGU8hcqRdHDCbuYhCPADMLM2VcCb8VnFnQ'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('CPsTfDvZYeVB5uTqQZcwwTTBJ7KPFvB6JKLGSWsFZEL7'),
                    collateralA: 'SOL',
                    collateralB: 'DUST',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('DUSTawucrTsGU8hcqRdHDCbuYhCPADMLM2VcCb8VnFnQ'),
                },
                {
                    address: new web3_js_1.PublicKey('4eJ1jCPysCrEH53VUAxgNT8BMccXsgHX1nX4FxXAUVWy'),
                    collateralA: 'USDH',
                    collateralB: 'MSOL',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                },
                {
                    address: new web3_js_1.PublicKey('6BFWHpnQA7BTHq8XXzuPFmKiZxwH866udVVXNvMrWPqB'),
                    collateralA: 'USDR',
                    collateralB: 'USDT',
                    mintA: new web3_js_1.PublicKey('USDrbBQwQbQ2oWHUPfA8QBHcyVxKUq1xHyXsSLKdUq2'),
                    mintB: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'),
                },
                {
                    address: new web3_js_1.PublicKey('5kDjnhGz9jjwUE84syEjUJjPyQNo4eA5uxs9fd5zzswS'),
                    collateralA: 'USDR',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('USDrbBQwQbQ2oWHUPfA8QBHcyVxKUq1xHyXsSLKdUq2'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('AiZa55wSymdzwU9VDoWBrjizFjHdzDJFRNks2enP35sw'),
                    collateralA: 'SOL',
                    collateralB: 'USDH',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                },
                {
                    address: new web3_js_1.PublicKey('GpVjnSuCA9rfrWS13nL1wPmzqNzqLrP5ih1vhu7yhiPv'),
                    collateralA: 'USDH',
                    collateralB: 'BTC',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'),
                },
                {
                    address: new web3_js_1.PublicKey('3jLRacqwVaxLC6fSNaZSHiC7samXPkSkJ3j5d6QJUaEL'),
                    collateralA: 'USDH',
                    collateralB: 'ETH',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                },
                {
                    address: new web3_js_1.PublicKey('7GZXejXCyJ3R78d71328wqMWT7Ejwu7pxxWRWvHSeVfb'),
                    collateralA: 'USDH',
                    collateralB: 'HBB',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6'),
                },
                {
                    address: new web3_js_1.PublicKey('AU971DrPyhhrpRnmEBp5pDTWL2ny7nofb5vYBjDJkR2E'),
                    collateralA: 'ETH',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('7qbRF6YsyGuLUVs6Y1q64bdVrfe4ZcUUz1JRdoVNUJnm'),
                    collateralA: 'SOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('Hp53XEtt4S8SvPCXarsLSdGfZBuUr5mMmZmX2DRNXQKp'),
                    collateralA: 'SOL',
                    collateralB: 'JITOSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
                },
                {
                    address: new web3_js_1.PublicKey('GgAqL1Lfs3zYZK6Gea9Bey5jmPbK8c4dmBbe5gPeYQEz'),
                    collateralA: 'USDC',
                    collateralB: 'JITOSOL',
                    mintA: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                    mintB: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
                },
                {
                    address: new web3_js_1.PublicKey('GGRzS8AmNmMmjiMKUKRp4BML4UurpTPh7ZbLyeGqX8LA'),
                    collateralA: 'USDH',
                    collateralB: 'JITOSOL',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
                },
                {
                    address: new web3_js_1.PublicKey('GMx3g7HZCVzLuE6Ckdu9sgExgiZQ7kiF7xfCrPnwVzbr'),
                    collateralA: 'USDH',
                    collateralB: 'BONK',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263'),
                },
                {
                    address: new web3_js_1.PublicKey('3ne4mWqdYuNiYrYZC9TrA3FcfuFdErghH97vNPbjicr1'),
                    collateralA: 'SOL',
                    collateralB: 'BONK',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263'),
                },
                {
                    address: new web3_js_1.PublicKey('8dxebMPEZjYJvE5JfC9iicZt9pkATBLW1PFgRHi5wGGv'),
                    collateralA: 'MSOL',
                    collateralB: 'JITOSOL',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
                },
                {
                    address: new web3_js_1.PublicKey('DpmWxsGtJoe9r5JjYT5XWN5ZDBW54echsfQBt54WEi9m'),
                    collateralA: 'USDH',
                    collateralB: 'SAMO',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU'),
                },
                {
                    address: new web3_js_1.PublicKey('DFNxfUy4YfAafSJBDL3cTeLDb2DPBMJ5wCxePzWujmVA'),
                    collateralA: 'SOL',
                    collateralB: 'SAMO',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU'),
                },
                {
                    address: new web3_js_1.PublicKey('6MSGDnvqB7qnTjXerKFyJwKUPWjNfCc9kyThR5sL8GWZ'),
                    collateralA: 'bSOL',
                    collateralB: 'BONK',
                    mintA: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
                    mintB: new web3_js_1.PublicKey('DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263'),
                },
                {
                    address: new web3_js_1.PublicKey('6MSGDnvqB7qnTjXerKFyJwKUPWjNfCc9kyThR5sL8GWZ'),
                    collateralA: 'LaineSOL',
                    collateralB: 'BONK',
                    mintA: new web3_js_1.PublicKey('LAinEtNLgpmCP9Rvsf5Hn8W6EhNiKLZQti1xfWMLy6X'),
                    mintB: new web3_js_1.PublicKey('DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263'),
                },
                {
                    address: new web3_js_1.PublicKey('6MSGDnvqB7qnTjXerKFyJwKUPWjNfCc9kyThR5sL8GWZ'),
                    collateralA: 'LaineSOL',
                    collateralB: 'JITOSOL',
                    mintA: new web3_js_1.PublicKey('LAinEtNLgpmCP9Rvsf5Hn8W6EhNiKLZQti1xfWMLy6X'),
                    mintB: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
                },
                {
                    address: new web3_js_1.PublicKey('2QK4n92cVZvAW8WjSYz4imyFECJunKHETDDu6X3w29b8'),
                    collateralA: 'HADES',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('H5VULAZyHVkp8E8UM4afihuoebHPjSoUberhYBXeafGU'),
                    collateralA: 'HADES',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('78UabsZZey1U72t8axX1vfDUTHubn7SPCEUraMiU2y8Y'),
                    collateralA: 'HADES',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('Aj96H3zePCQ3Z8jL5tND2BiBYTaCWcfGESkn9Sa4FF9u'),
                    collateralA: 'SOL',
                    collateralB: 'HADES',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                },
                {
                    address: new web3_js_1.PublicKey('DjfEf8RqbBytbq2oyG5oSrdfGYgtXhk24zgwapswNTfZ'),
                    collateralA: 'SOL',
                    collateralB: 'HADES',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                },
                {
                    address: new web3_js_1.PublicKey('i3Ug1L3QnDPeyUg7SzyUwSTo4GLRecVhDd1aeivirDq'),
                    collateralA: 'SOL',
                    collateralB: 'HADES',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                },
                {
                    address: new web3_js_1.PublicKey('5snaYowgJDfuM1LPbTNUYHbgkKHtVVnzHiiLDWUV2hh8'),
                    collateralA: 'bSOL',
                    collateralB: 'JITOSOL',
                    mintA: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
                    mintB: new web3_js_1.PublicKey('J1toso1uCk3RLmjorhTtrVwY9HJ7X8V9yYac6Y7kGCPn'),
                },
                {
                    address: new web3_js_1.PublicKey('5snaYowgJDfuM1LPbTNUYHbgkKHtVVnzHiiLDWUV2hh8'),
                    collateralA: 'SOL',
                    collateralB: 'bSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
                },
                {
                    address: new web3_js_1.PublicKey('ABwq1ro51rdcuknyTs1M5oYqmLxY4E9eaBy19QGvpXym'),
                    collateralA: 'bSOL',
                    collateralB: 'BONK',
                    mintA: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
                    mintB: new web3_js_1.PublicKey('DezXAZ8z7PnrnRJjz3wXBoRgixCa6xjnB7YaB1pPB263'),
                },
                {
                    address: new web3_js_1.PublicKey('EnSJFxjoR9gXKVup6taJxjc6siLGJFmHyqrxawRq9gk5'),
                    collateralA: 'SOL',
                    collateralB: 'bSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
                },
                {
                    address: new web3_js_1.PublicKey('3EptK4R8rUDWuWMvfF7Wf23whHPSgrmfravJQXWHQ2L3'),
                    collateralA: 'USDC',
                    collateralB: 'HADES',
                    mintA: new web3_js_1.PublicKey('BWXrrYFhT7bMHmNBFoQFWdsSgA3yXoAnMhDK6Fn1eSEn'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('8phK65jxmTPEN158xLgSr4oZvssw9SyTErpNZj3g7px4'),
                    collateralA: 'SOL',
                    collateralB: 'bSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('bSo13r4TkiE4KumL71LsHTPpL2euBYLFx6h9HP3piy1'),
                },
                {
                    address: new web3_js_1.PublicKey('AHoX7W4sjiYQnjdbMLqr4NrBMZqTyoA6QoeRdNfgDez'),
                    collateralA: 'RLB',
                    collateralB: 'SOL',
                    mintA: new web3_js_1.PublicKey('RLBxxFkseAZ4RgJH3Sqn8jXxhmGoz9jWxDNJMh8pL7a'),
                    mintB: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                },
                {
                    address: new web3_js_1.PublicKey('HwCngan6JmSxeYdizHxDLXUUi21ggoQgVPVSp9rP2Hk1'),
                    collateralA: 'RLB',
                    collateralB: 'SOL',
                    mintA: new web3_js_1.PublicKey('RLBxxFkseAZ4RgJH3Sqn8jXxhmGoz9jWxDNJMh8pL7a'),
                    mintB: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                },
                {
                    address: new web3_js_1.PublicKey('6bskSanDywHt17rRXRbnx1zoErCeCpst5P4544s6cZka'),
                    collateralA: 'RLB',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('RLBxxFkseAZ4RgJH3Sqn8jXxhmGoz9jWxDNJMh8pL7a'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('DvgSQJyx6JNaPzmhBwzWw6rntGBQCr5fmNnV2AfyEfCg'),
                    collateralA: 'STSOL',
                    collateralB: 'MSOL',
                    mintA: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                    mintB: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                },
                {
                    address: new web3_js_1.PublicKey('max1GTEmP7zAxJSEe9jFkihYSnxZmbEbp9BtBNYw1K6'),
                    collateralA: 'LDO',
                    collateralB: 'STSOL',
                    mintA: new web3_js_1.PublicKey('HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p'),
                    mintB: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                },
                {
                    address: new web3_js_1.PublicKey('BVXNG6BrL2Tn3NmppnMeXHjBHTaQSnSnLE99JKwZSWPg'),
                    collateralA: 'MNDE',
                    collateralB: 'MSOL',
                    mintA: new web3_js_1.PublicKey('MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey'),
                    mintB: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                },
                {
                    address: new web3_js_1.PublicKey('5Z66YYYaTmmx1R4mATAGLSc8aV4Vfy5tNdJQzk1GP9RF'),
                    collateralA: 'ORCA',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('H3bEHDH3ResiGgR8avFUEfU8rfyLzjEQbf4WR5mkXmJ9'),
                    collateralA: 'MSOL',
                    collateralB: 'CGNTSOL',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('CgnTSoL3DgY9SFHxcLj6CgCgKKoTBr6tp4CPAEWy25DE'),
                },
                {
                    address: new web3_js_1.PublicKey('Zko5YLwAZL1F2W13ohkG9DVAZCmYqcCqHrEgmaudmbe'),
                    collateralA: 'CGNTSOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('CgnTSoL3DgY9SFHxcLj6CgCgKKoTBr6tp4CPAEWy25DE'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('GrXkAvh5dohLd8Va432MxW2ckthWhfw9FEGSxJHEsKg'),
                    collateralA: 'USDC',
                    collateralB: 'HXRO',
                    mintA: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                    mintB: new web3_js_1.PublicKey('HxhWkVpk5NS4Ltg5nij2G671CKXFRKPK8vy271Ub4uEK'),
                },
                {
                    address: new web3_js_1.PublicKey('AFCKH5AnW2inQdV3RgcLyk9PgLUhbN6Xy64pceJSwrfs'),
                    collateralA: 'USDC',
                    collateralB: 'HXRO',
                    mintA: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                    mintB: new web3_js_1.PublicKey('HxhWkVpk5NS4Ltg5nij2G671CKXFRKPK8vy271Ub4uEK'),
                },
                {
                    address: new web3_js_1.PublicKey('7A1R3L7AxcxuZHMJjFgskKGeBR5Rwst3Ai5bv5uAWZFG'),
                    collateralA: 'MSOL',
                    collateralB: 'USDT',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB'),
                },
                {
                    address: new web3_js_1.PublicKey('7A1R3L7AxcxuZHMJjFgskKGeBR5Rwst3Ai5bv5uAWZFG'),
                    collateralA: 'USDCet',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('A9mUU4qviSctJVPJdBJWkb28deg915LYJKrzQ19ji3FM'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('45zpzzpZquaVv4BAdXfzNbuba7DAXV32d1sTrUh5wcnW'),
                    collateralA: 'HNT',
                    collateralB: 'MOBILE',
                    mintA: new web3_js_1.PublicKey('hntyVP6YFm1Hg25TN9WGLqM12b8TQmcknKrdu1oxWux'),
                    mintB: new web3_js_1.PublicKey('mb1eu7TzEc71KxDpsmsKoucSSuuoGLv1drys1oP2jh6'),
                },
                {
                    address: new web3_js_1.PublicKey('FmbpZzT61ftb5pR3Rn5kgerLRvFzKicc7aCj4Zaes5E9'),
                    collateralA: 'MOBILE',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('mb1eu7TzEc71KxDpsmsKoucSSuuoGLv1drys1oP2jh6'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('FMFVwK5TXbKb6Q9LVxN445fisLdZHNSL4wfRSCMrBpay'),
                    collateralA: 'MOBILE',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('mb1eu7TzEc71KxDpsmsKoucSSuuoGLv1drys1oP2jh6'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('6x6FPHbG1igaDCJ57UScYxg1sFjevhMNaW9md9PfudFx'),
                    collateralA: 'HNT',
                    collateralB: 'IOT',
                    mintA: new web3_js_1.PublicKey('hntyVP6YFm1Hg25TN9WGLqM12b8TQmcknKrdu1oxWux'),
                    mintB: new web3_js_1.PublicKey('iotEVVZLEywoTn1QdwNPddxPWszn3zFhEot3MfL9fns'),
                },
                {
                    address: new web3_js_1.PublicKey('6CgjHrdiQXDZMXL8BjUFMH6fq4tzWtPBGMB4tTWbDy6s'),
                    collateralA: 'HNT',
                    collateralB: 'IOT',
                    mintA: new web3_js_1.PublicKey('hntyVP6YFm1Hg25TN9WGLqM12b8TQmcknKrdu1oxWux'),
                    mintB: new web3_js_1.PublicKey('iotEVVZLEywoTn1QdwNPddxPWszn3zFhEot3MfL9fns'),
                },
                {
                    address: new web3_js_1.PublicKey('HoSnxZhcLASeHt1rirASizonYoCcBszHHFaF1Z8cvncA'),
                    collateralA: 'USDH',
                    collateralB: 'HNT',
                    mintA: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                    mintB: new web3_js_1.PublicKey('hntyVP6YFm1Hg25TN9WGLqM12b8TQmcknKrdu1oxWux'),
                },
            ],
            raydiumPoolsToSnapshot: [
                {
                    address: new web3_js_1.PublicKey('3tD34VtprDSkYCnATtQLCiVgTkECU3d12KtjupeR6N2X'),
                    collateralA: 'SOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('HZf7wppva3wk4dCnrUe2GE1c8aUEXsUNk5asMFQ5sYch'),
                    collateralA: 'SOL',
                    collateralB: 'MSOL',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                },
                {
                    address: new web3_js_1.PublicKey('CiHghAc3Fat5JgBB4L6vUcY6hud89YwgL1q4nKu1rsNQ'),
                    collateralA: 'MSOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('9cGUdTiSGsP9Du3NXHhkxK7fbNT4pL543yZH4bUatR1v'),
                    collateralA: 'STSOL',
                    collateralB: 'ETH',
                    mintA: new web3_js_1.PublicKey('7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj'),
                    mintB: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                },
                {
                    address: new web3_js_1.PublicKey('4nFbdT7DeXATvaRZfR3WqALGJnogMjqe9vf2H6C1WXBr'),
                    collateralA: 'BTC',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('3NZ9JMVBmGAqocybic2c7LQCJScmgsAZ6vQqTDzcqmJh'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('HCfytQ49w6Dn9UhHCqjNYTZYQ6z5SwqmsyYYqW4EKDdA'),
                    collateralA: 'SOL',
                    collateralB: 'BTC',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('3NZ9JMVBmGAqocybic2c7LQCJScmgsAZ6vQqTDzcqmJh'),
                },
                {
                    address: new web3_js_1.PublicKey('3BScXnPjT4hut1G5yJ5UGQWhUmoYxyBFQf3juLBeMH2S'),
                    collateralA: 'BTC',
                    collateralB: 'ETH',
                    mintA: new web3_js_1.PublicKey('3NZ9JMVBmGAqocybic2c7LQCJScmgsAZ6vQqTDzcqmJh'),
                    mintB: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                },
                {
                    address: new web3_js_1.PublicKey('61R1ndXxvsWXXkWSyNkCxnzwd3zUNB8Q2ibmkiLPC8ht'),
                    collateralA: 'RAY',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('Enfoa5Xdtirwa46xxa5LUVcQWe7EUb2pGzTjfvU7EBS1'),
                    collateralA: 'SOL',
                    collateralB: 'RAY',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'),
                },
                {
                    address: new web3_js_1.PublicKey('Chbi4BsySNmnNB4i821eMH6TzWA9DiUx2J1oHVVTk9Dm'),
                    collateralA: 'MSOL',
                    collateralB: 'HBB',
                    mintA: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    mintB: new web3_js_1.PublicKey('HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6'),
                },
                {
                    address: new web3_js_1.PublicKey('2QdhepnKRTLjjSqPL1PtKNwqrUkoLee5Gqs8bvZhRdMv'),
                    collateralA: 'SOL',
                    collateralB: 'USDC',
                    mintA: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    mintB: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                },
                {
                    address: new web3_js_1.PublicKey('EhJGjF8FSqAoJPGyR3ixzFjcUM1y64fbMZEZLEUc56XN'),
                    collateralA: 'RLB',
                    collateralB: 'SOL',
                    mintA: new web3_js_1.PublicKey('RLBxxFkseAZ4RgJH3Sqn8jXxhmGoz9jWxDNJMh8pL7a'),
                    mintB: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                },
            ],
            strategies: [
                new web3_js_1.PublicKey('Cfuy5T6osdazUeLego5LFycBQebm9PP3H7VNdCndXXEN'),
                new web3_js_1.PublicKey('DJ4tLS5q566UXJRdEREBKAg9URuNoyKZWByaUu62npHC'),
                new web3_js_1.PublicKey('ByXB4xCxVhmUEmQj3Ut7byZ1Hbva1zhKjaVcv3jBMN7E'),
                new web3_js_1.PublicKey('D6LyYoruQkw11oDqjdNifLLEd3aWkkCgTdhMvn4pvqRv'),
                new web3_js_1.PublicKey('98kNMp1aqWoYAaUU8m5REBAYVwhFb4aX9yoSpgq8kUFu'),
                new web3_js_1.PublicKey('2VQaDuSqqxeX2h9dS9WgpvN6ShaBxd8JjaaWEvbmTDY1'),
                new web3_js_1.PublicKey('BCSKhjooRRtieLs4NX92bP6q3GNyqPWC3a5wbYWfxK6Y'),
                new web3_js_1.PublicKey('7T4LUZLjZpaRrXYrWjn4E4x2gackFnuidz6j83xDNvXW'),
                new web3_js_1.PublicKey('FshxiQWH1kTya19d186VyCSLT8PAVKn6wpVU4wP26S2M'),
                new web3_js_1.PublicKey('BfyQYYr2T9eJfMfq5gPXcq3SUkJSh2ahtk7ZNUCzkx9e'),
                new web3_js_1.PublicKey('6taYeUSiNzuzQ8ys5d4TErzKRTPcqAC8Mx78RfKUoD9w'),
                new web3_js_1.PublicKey('VF45TSF5WPAay9qy2zr1hPYgieBv7r17vYLRK6v1RmB'),
                new web3_js_1.PublicKey('2J7zeWbFC6PxtwnwEHb8zpF62w9Xanc18ThsF3UHMtVf'),
                new web3_js_1.PublicKey('GFaca7ca27vXofRJ3y1FtWgKwKWjn3LdCexttCsRLcvc'),
                new web3_js_1.PublicKey('A4ufgHTe3jLzxbR6sDdrZhLxNdR1Lw2ija1uEdDFLPbX'),
                new web3_js_1.PublicKey('CofEPsAoV6bdn7guCPHhmb5nJ1xmnhZ4Ha2zZVx14Ppb'),
                new web3_js_1.PublicKey('BKipQNa4zGAxS2kHjvgRMyFymBXJg5wnYjiud5zdqvAS'),
                new web3_js_1.PublicKey('HVhXUenYdGnBNwP9ZrEN43AFv2Faa9GPMgZsHALPWYoT'),
                new web3_js_1.PublicKey('DFXL6ake3YJ7iqyM3iJEQwjPG3w55C37QS8k6SAAyXH8'),
                new web3_js_1.PublicKey('C53jcmBve7bDeQFMShCZafNX5ok7XapJLyJscQxvsjPy'),
                new web3_js_1.PublicKey('G2kaQU5ECRaxGVoFRB1qMYE7QCtmZLupo7ShTYyymv14'),
                new web3_js_1.PublicKey('32QXT5qS7YmPvGu2sngFkazPvp8j3choYDwW6DT8eFnD'),
                new web3_js_1.PublicKey('CYM2ouv2Kih73k6E76Vs15qttsz7XFqtzjo33YyvKfKL'),
                new web3_js_1.PublicKey('5i2qShi1FdGHDPnKJVPCetTveyve5GfMngg6yb22XJ6A'),
                new web3_js_1.PublicKey('FJBDScgmTchvBpY7DnrPwn3mQ8jGQ1SMdrGMWVesDVhB'),
                new web3_js_1.PublicKey('4dk4RuUvpXJKTrjYqLpagQ5JPKwb5oZ43KZDE42hqVsN'),
                new web3_js_1.PublicKey('9jn566VXtkf6SJfpwJhJyc8ycmWriTXBaGz4Y2XKQQph'),
                new web3_js_1.PublicKey('9FMsTHZKuW1YfCqzy2wEGt58o6vbJh7G5adWokQ5ejGf'),
                new web3_js_1.PublicKey('CBYt9Nd63Br7E1vaeiLHeCHkV8Ut56ALmxSmhuqFfo7z'),
                new web3_js_1.PublicKey('73WsXLMrCZywN1sa7VdSW7UvxVjjrQsJygGJzKhTnGH'),
                new web3_js_1.PublicKey('EsepQgrgTHiKJHQVnJUZ636yDDt4uqWuQweStNw8Bgch'),
                new web3_js_1.PublicKey('8kE6dbhS4pCie8LNaeVZx7jhEcpKKewho8UYiardNaxX'),
                new web3_js_1.PublicKey('5VwUt1yNJh3HWJvydwkDMMgdMTMw99BCqnhLvtAanfgZ'),
                new web3_js_1.PublicKey('DM6f3WvxmW6decwhMywz89VEW65FcxWV1nxvxveH8iWC'),
                new web3_js_1.PublicKey('6YL2hyzt639WcAZywnS9ApMBnWwQgTWyiDfb7AJDFqCJ'),
                new web3_js_1.PublicKey('6zgPtnJkWa9JcXL9tnsrx2iaWRyjzAidigDz3mMZZ7ko'),
                new web3_js_1.PublicKey('9d7nQMn8sVjYeM8PooVU5cQgLTgT4cX7oZX6GUUZFrEf'),
                new web3_js_1.PublicKey('3R8wxZYtyicN8KrQtXnkgvgG4NuYW1Ra5NkepQz3Kovj'),
                new web3_js_1.PublicKey('EvcCVFV36YvgwJzeyp4jiEu3TDfCPQP2Xu3Zwfsu4DeG'),
                new web3_js_1.PublicKey('DcqhRqfgzgkxiamwE4uG7XGYisBt2XEn62skD3vGfKxk'),
                new web3_js_1.PublicKey('FjZWSxXjLMQb6P64rHqCrrumJSkEWj7XvhgMaFNb4fbE'),
                new web3_js_1.PublicKey('B3Wh2AfiYL9VDfz8nYErgKyXwBUA32YPtxfEafLhBDZE'),
                new web3_js_1.PublicKey('4Le8QimmygEhuAyifX9ngTMXegFzsWtKJNsfv3A2qcce'),
                new web3_js_1.PublicKey('HperM5vN34nKVwGuMmyu9SoKAVBytccNDbF9ekoQFXkw'),
                new web3_js_1.PublicKey('4sEAHHyLytB4pszzob6rJhaNj8WXj6zsyCesMGkLvbKd'),
                new web3_js_1.PublicKey('Dnvk55Ya5VqqigtovM5Wxk9AFrHJhn72iF5CUXBd9f3Q'),
                new web3_js_1.PublicKey('BmkdmGUHnB2DGZduxVLV3Eo2BKupLipapCKVQTjk6Jhh'),
                new web3_js_1.PublicKey('BviEYunJoaMkumtEX9FyWiW7coaeWALDaaXiFme5zJiW'),
                new web3_js_1.PublicKey('4Ypy6JwmgvcZ9bNzwTmfXK51y8uj1XUHJYSKAvJVNc5g'),
                new web3_js_1.PublicKey('DEjvZw9ZMkh1wKvUG4S1jmiqojC46ePP13kWvxAmg6hR'),
                new web3_js_1.PublicKey('3pxDd6AJ6EwbVa4P3mGCX3qWiVmdWma4EZCP2JQ2uB62'),
                new web3_js_1.PublicKey('HnjNekxuafTkA9xZiP73QzBQH8LmKifcWuP4ncZ7zMQV'),
                new web3_js_1.PublicKey('8tx2ASp98JitHqQk5iUqtHtd4tAkPAmFJqXwyFtEEuUB'),
                new web3_js_1.PublicKey('C1a6BcMVhx3uGCpdeMHN7iuLAwzmQiWv1BzBfyFgq9oj'),
                new web3_js_1.PublicKey('233yH7imV7EXNKPgUU37tSDFYtRrDXXLUbE2uLqkZqHt'),
                new web3_js_1.PublicKey('6zNRgnvCStERUhFAz9SWPv3PzDgT6k4vwVzmcnCQRDM4'),
                new web3_js_1.PublicKey('A6XLLT4NREqCnzBtdMY31Z58HqMSCUMUCx2LoMmFv8wg'),
                new web3_js_1.PublicKey('6mZtb1h9zvkcDxTzS9481pbjJedtTHbMoQicg8NZReqt'),
                new web3_js_1.PublicKey('8TGRn1PNJVCAiqAosnAZy9i5TrwMZQskDGVhTLj3hJeL'),
                new web3_js_1.PublicKey('CbLWjxwx6BT5PyXnYdmCX6rr2JhCtzQQwjLTN2D4m57U'),
                new web3_js_1.PublicKey('AmRCw9xKs8jgmvcr2ETrwAAZXWC829QiaWcbgJgpBuvh'),
                new web3_js_1.PublicKey('7Bdn78UV2Eu6kwEsu7ar4oWx6w58cyTZGSfQwXVbwQWk'),
                new web3_js_1.PublicKey('E3kjGC76RG7KBkeYqUVLXo91muvgMmt2dTCqvA4RqFPp'),
                new web3_js_1.PublicKey('9RC8QNFFfnQ7c6JrZNBpBDW2X2nytJL6h7gRQS8ENuQL'),
                new web3_js_1.PublicKey('zKLfwe3XuKj3HpFn67Q5k7nMz1ZtyKr3ZKG7NoqTVW1'),
                new web3_js_1.PublicKey('GcErgVRWQDLRmMMkNQjxZeSCYte2hFkVbLeiw6HcwZnY'),
                new web3_js_1.PublicKey('8t2qfUx38BoasGP8V8gST1dBZNuGhV66qUHpgFNBHC8o'),
                new web3_js_1.PublicKey('FPjFsUAvupj4FPUY8HpWj7dEqFcJUWyn193fwvYxYkTW'),
                new web3_js_1.PublicKey('Ehic3YqbFJPioH4WEVi15syxd6PduF6G846CZJ9KrnUJ'),
                new web3_js_1.PublicKey('Hipe568Vrn6VPBzaPZkS9iKAkLMFzmeWCS7G7w1jHZd5'),
                new web3_js_1.PublicKey('HvdhFUvjTTm28BH3LirANFa4hoafyQkpw12Dj7D64sNb'),
                new web3_js_1.PublicKey('JP64K4wp3pDdNUp4uP5dARaUF5stg1jpNgfFXC5ciDK'),
                new web3_js_1.PublicKey('KKugMgpzSHKtoT7aJdYksNbB1VpcgRU8NghkjXCB4ab'),
                new web3_js_1.PublicKey('5po2XSZvwynEiwMxApgBcx7SDqeV5oy91icjC1uPaudi'),
                new web3_js_1.PublicKey('EEqvgDEhfWJW2sMWr1cRYddb4ixGiGnYPoCSkHEXd1KT'),
                new web3_js_1.PublicKey('55dwiG4E7vsiZUzzffneaXZh4rsHamMTbzaEBtQQ2kLk'),
                new web3_js_1.PublicKey('3bx7xzmxgyWaW9ptn1x7Zt1LdLXhQhiwPCVuToDVqTAq'),
                new web3_js_1.PublicKey('Gug8wWqnDbZ21xV5qAv8aHnuyKBuJvnuyPhSuAtmPfXT'),
                new web3_js_1.PublicKey('9sHCCWmhqLtPa4yqgDke7CNtaE9KGpUWWpni8MC4WJUU'),
                new web3_js_1.PublicKey('6iyPtRHPdRPNfkkSWQA3jahtkgjgthfM3dwAk71nPJTJ'),
                new web3_js_1.PublicKey('Ht3kPVGEED4sDAUALLirqU4NodNc4zbSi8WVdNAKoRKh'),
                new web3_js_1.PublicKey('3YQnXFCTZwtL2H3MadkCBPA1NRXJkXn416psWunnZYAr'),
                new web3_js_1.PublicKey('B8uttDQzGBz8deXh7LRNFDG6wUdxdBSQ9y2pCMPyzjBH'),
                new web3_js_1.PublicKey('G7jTpLp3F53DJwLaqYtVsnjeG2p5d5AosBFMEK4NMcy9'),
                new web3_js_1.PublicKey('A7e4aYqohqLr9oFFwAEbUZsBpYvKsutWxDtW3MXVhLB4'),
                new web3_js_1.PublicKey('HXXdvkZ4RYymVWaA47Hs75X7E9B1JmsJFNRyAjvL66Hz'),
                new web3_js_1.PublicKey('2FzPrKLsGAU9QHQG1ohjb2xgZJYQzyTzrLusqHAN2Ekg'),
                new web3_js_1.PublicKey('EuoZcSLnXho591Uxt3mvQrtREzdnsuvNNz6WyearTNHm'),
                new web3_js_1.PublicKey('FBp7wr3qbB7WYRX8BppqbqNy7jj2mcVFnCy9s8XeTJ9f'),
                new web3_js_1.PublicKey('AuGGSC8opfii5X5uipnUhYpvHQCUfDUkd5Z2PWGqJqwL'),
                new web3_js_1.PublicKey('DAqZ5XiVNwgg7AhR6kmgnz6THjGb6W72DqYSjY2tz3Du'),
                new web3_js_1.PublicKey('4hsTKUftS75am74p1qYeJLBSbed2Tqdcb2DLtrQNqoAp'),
                new web3_js_1.PublicKey('HZj6V8DBca1tbZJwXEJDFHoK9sG7udgC82Wu788uZupz'),
                new web3_js_1.PublicKey('4wHQqd14L4126Em11TzdT8DLX4dHyB8drbTEdi15xk4Z'),
                new web3_js_1.PublicKey('59h7Hpobueq5trqmH9G4A5fh4vZVY5Kw8TryMfdfWK4'),
                new web3_js_1.PublicKey('8u5892nUCULGGmZkyimtBvVHnoGUaBKRBiqyVZtSapDV'),
                new web3_js_1.PublicKey('88ThXKrCmt8QGX6FFmCRGvxwmfTWsrtwMeNgjEsggE7W'),
                new web3_js_1.PublicKey('kkA4gzNEn52u4LrDSWS4jPBrooV2qziM6g2iG5rs7kq'),
                new web3_js_1.PublicKey('BjdSuiqJBeuJeBRZtidZUMJJbBMNpEptGaC6HRM4SuwN'),
                new web3_js_1.PublicKey('8uSeb2xNPBtennNiLP5ELvU4djfyw6n2zDSgD7ZtyC9y'),
                new web3_js_1.PublicKey('HyPDBuHnXaFkrhE1da3YULrC4fFxUN6TEa9B7xhvFpmN'),
                new web3_js_1.PublicKey('8eoYZJccX2cju4oWnLD46V3n5tDmWpEbPZm75aiF33SV'),
                new web3_js_1.PublicKey('BjHxRgqCB6KJ6eoYBeVDtDFJrwrg5hfgCeEHHaP5Raba'),
                new web3_js_1.PublicKey('3qwa9ApkT9E7DtCZkYTCURE3fg1g8jwaQpuhuNwzAAfx'),
                new web3_js_1.PublicKey('7XZQQL11KsMPafyqmyKvUuzSQVd4Bvkf2iPJn8hGGKjU'),
                new web3_js_1.PublicKey('89LNNaDTwq65CXhmtGEYe4erWi7f1qMQ87xhpiCBazU2'),
                new web3_js_1.PublicKey('CYLt3Bs51QT3WeFhjtYnPGZNDzdd6SY5vfUMa86gT2D8'),
                new web3_js_1.PublicKey('HoqXSSbQGZS8o6fTqtfn9d14FVCm9WiuRB13zbHNKs7d'),
                new web3_js_1.PublicKey('7KqB3vRJQDdGwK7ewiDpAxXpTeMmgGicdDdbftQH41XC'),
                new web3_js_1.PublicKey('9zBNQtnenpQY6mCoRqbPpeePeSy17h34DZP82oegt1fL'),
                new web3_js_1.PublicKey('3w1MiUh6Nn4YJLue8Ut8uwonKvFupLKve9nifDaicBf2'),
                new web3_js_1.PublicKey('G28LWfC5F8fCtqukagwmdAt9myJXuXCvDBXZtCy2J4r8'),
                new web3_js_1.PublicKey('8wk3Y4jAy7S56zHhC2F16nrFhJmv2QmwPyfPdWVbedbK'),
                new web3_js_1.PublicKey('Agz8ExDrjd7UqjACotmw6XhKgXsjhnufza9Nfn6C7Ktx'),
                new web3_js_1.PublicKey('4upkN3v9KnwxAD4aj64nTPynAS45ccSBnKns6qo6yeRu'),
                new web3_js_1.PublicKey('F3v6sBb5gXL98kaMkaKm5GfEoBNUaSd3ZGErbjqgzTho'),
                new web3_js_1.PublicKey('5QKrn6Eua5EgmGdSTn8nEUa3dcZx2tHEYBv6AkPetojE'),
                new web3_js_1.PublicKey('TZ3LFwCT9mTj1h31tvGcjszS8xAoSZYTx81XTSpBaEL'),
                new web3_js_1.PublicKey('81h4VfGfbZE8xFB2JpWeLKpAw9YKyJjwCrBC7WQ4yvw6'),
                new web3_js_1.PublicKey('4WRpV94hyhzRBcCqVyky7UdXCCRWVoUCQUGmoThduvfL'),
                new web3_js_1.PublicKey('5EfeGn1h7m6Rx9mGEmamDoxMtdhRmUh2N9fYRiDQqteS'),
                new web3_js_1.PublicKey('AmjRKyaLCFJFa7U9axraV9bPW5prcWYe84j6LkvfMKtS'),
                new web3_js_1.PublicKey('8NP2J7q6swBkVoLDZAqkejKPQrWkRizZHaVVM897CKpA'),
                new web3_js_1.PublicKey('HCntzqDU5wXSWjwgLQP5hqh3kLHRYizKtPErvSCyggXd'),
                new web3_js_1.PublicKey('GwfjL2jQugK7sJdhMDj7tcGuqxi5H1qAZHubfFL8amNJ'),
                new web3_js_1.PublicKey('F8dVvsYwQciqe9CRuiUDUgEtqjiZm1DT93fnE67f1KsM'),
                new web3_js_1.PublicKey('8jenQV2vi7EVTLW11rnUaoeYnr8KabEBZmxAAeoZMPSQ'),
                new web3_js_1.PublicKey('FtRQ5X7vAwzZPaMzMW22tSUKrWXjqsMY1KZzhW4qfMF3'),
                new web3_js_1.PublicKey('Hv8QW2UxK4DLtg8dJixQwWHR4nhkimMMDuqJmjJcim1m'),
                new web3_js_1.PublicKey('8vbhUqcft4m2PzBmEntBzyWEo9dH7uxqzbDBbTJfe1rr'),
                new web3_js_1.PublicKey('75kopyQAEro5GdUDDmthdUK6oMKJ7z8oeezXduVvJu9D'),
                new web3_js_1.PublicKey('2U18NpPb8W6m4fgxgzXx6FLueQ3SGn3LGh2QnzvCXcHs'),
                new web3_js_1.PublicKey('4Hm233wZuhkPEJcEvZfcmVQV1nrhP7RmmpRZgvMCTmNM'),
                new web3_js_1.PublicKey('FxBe7AxKQQNCKDzkMzbpCmtnhDPmnTavfRqHeii8EmvW'),
                new web3_js_1.PublicKey('E7K2S9qhypk9S1EaGKWaXyN9rP2RmPPmwHKWGQuVZHNm'),
                new web3_js_1.PublicKey('GGmL7gXP1EKwwagMY6c4nBbypiUkVVvRrAt51FJafynQ'),
                new web3_js_1.PublicKey('BXQEJ1xWEAxSiixRVMCdyAM9s2NypCkBuN88WQMHxcCE'),
                new web3_js_1.PublicKey('8dahx59u8Xq79VZQkujJQHsQ5JzHTTjAJGZqKPiqsFbA'),
                new web3_js_1.PublicKey('8YMXFERNrQ1gJbzhZKWjnebnKPpUPGQoJDKbWKymU1SL'),
                new web3_js_1.PublicKey('6dWMu9iNNuBLpm4bnPdB5gtAPHUnhrcWconsoRoP6Ttt'),
                new web3_js_1.PublicKey('E52WMjRwax2B2ZXABjZUt6xZquCMQgXAuErSs7j6RGS4'),
                new web3_js_1.PublicKey('87tynGCmaM9FUzE8QT1ENSXPhPNJFv5mTkrPuGvgc9x7'),
                new web3_js_1.PublicKey('GQvawguCeqAGVww8e2rVn9iJw31fG5AY1bVh2ftiic9J'),
                new web3_js_1.PublicKey('8KV2ZPhiD4AgHBNow4bkgk2i7gh8WJrZBAUaXwtSA1De'),
                new web3_js_1.PublicKey('8YqBPENiMne3kEeKSJfPsrusvjDveiuawd91Afj4MAQS'),
                new web3_js_1.PublicKey('BaEVFsux7SerrsUxzb9y4pm14BMoE5eQNc1SCSPPeAq7'),
                new web3_js_1.PublicKey('EQZHYCWwiaCHCPfSKohfWfmmHtJ7RW3mk5ZogbDmZuBW'),
                new web3_js_1.PublicKey('GNBK4kfVPfjLecXReJD3rmAjeSgUnKNvAiiBKWxdMSc6'),
                new web3_js_1.PublicKey('DchrvcBg9kqWAqC2ziQRMcxsqSoHY7fJygRTMeUz9a5P'),
                new web3_js_1.PublicKey('ArE62nhfci2x1eHnTDAd76Z4ok7t9ywy8P9BhwoJU3jb'),
                new web3_js_1.PublicKey('BTeVjJCb1nYLEytsqiiY9nkwqaSsLzVoWrkDu5h994no'),
                new web3_js_1.PublicKey('3CaSWZVdWWwaYdef3GqRYYhs6Pbz1Ye8MqcLUqteQcSo'),
                new web3_js_1.PublicKey('5RD9AXADJfwkcj39WTePLs8GYyx7A874nvXSShXxSKgK'),
                new web3_js_1.PublicKey('FvxU7ftZu4VgtLi5nP5JbzuyqevaxR6jvgv7UPbHNHbn'),
                new web3_js_1.PublicKey('AKtWUmeVHL8KLVTyikPfqAWpwnUV89BJLti9v8nbTBzg'),
                new web3_js_1.PublicKey('4h6EPQ2DmpPqjVPJWCDza42UZC1CxjVSymRghZfwWieT'),
                new web3_js_1.PublicKey('493Bxf1L3TjNpSoRw2pcakhRyXeTYha9K6CGBwHZr7zC'),
                new web3_js_1.PublicKey('DSv8ey82ahkqjqMMykbNkGXfsCVvPch6X2JeazsHhYsV'),
                new web3_js_1.PublicKey('2fGnnDF3BwkRFhQLQzpUWMVHVcdmH9cEQihWZ29NUpNW'),
                new web3_js_1.PublicKey('DktbtdJHuDnXkZgMuV7Kai87CNqQBmAfjij7WA7MDcR9'),
                new web3_js_1.PublicKey('9oTss5uEEjYnvNYEFjSiMvqrH5TGmLtZjdwpjiVWQEyv'),
                new web3_js_1.PublicKey('6M8z6aQdeLtrwiSK4AkMhabyRpq3KrGkzwXhuBo5Zey6'),
                new web3_js_1.PublicKey('BLcPbyo9yzhVHxPtjHcgN5Nm1ksG2W7915P6FWW8LZ3b'),
                new web3_js_1.PublicKey('H8PBwav5aUYBz1Ne5T7nDBKsouBfaoHdfRc5UJwT5Ag5'),
                new web3_js_1.PublicKey('7JvdArAkW8RBWTu9hxA9CLb4gHbMgNpT6uzd1EqKPLKT'),
                new web3_js_1.PublicKey('BXa1S3U2nY8dWXJYXWHgD8zU186HvKQkpTFhUBRqVVWf'),
                new web3_js_1.PublicKey('9Ai4GBV4S6GhGe7ckFrfgo7Px8dDHtUEgXs7Nqa55RdD'),
                new web3_js_1.PublicKey('6nj5N2Nrmt1kXNa1rcj2cCER8weu1a3g3qPskuCm8gfW'),
                new web3_js_1.PublicKey('7FoZBdnghZgpYLp3EWDFzte6pQiPaQLMgw2XbBHuZbeM'),
                new web3_js_1.PublicKey('4wbXr6BhEVGa7edNipTJq3GxHg4zLmVgcAUsfBmTxTHr'),
                new web3_js_1.PublicKey('2m2BzYQYFRUP2WitXAC4wa8H9zN4qsV197gcdno59Qfe'),
                new web3_js_1.PublicKey('4qzbc1TH2VCDyjk1PK6WdrW1CDRiXyBoy2xfgt1Ff5kT'),
                new web3_js_1.PublicKey('61EjRY8cNK81btuYw5VsnCzD63bmFb4d6tWGZbWVNmPr'),
                new web3_js_1.PublicKey('12iZna9cRnhSY85cDTvas3mav36bYE9WeDuL9uFzH2Zw'),
                new web3_js_1.PublicKey('Mpxtgfv19XStXoZYmaGLziQqw2LxAY8txABU6A8cRNE'),
                new web3_js_1.PublicKey('DXESfdikAgtzbq8JQoQyiGRTGT5nPL7aXBSoYL3T8cjo'),
                new web3_js_1.PublicKey('UCdb2euhoW672rDsPgRN1es3nSn7BGvKfvFVrLc3xUc'),
                new web3_js_1.PublicKey('AMxoX3u5R6Ni1pZEvXJ9qbswhAdPKCo7gF29G5Vis5ft'),
                new web3_js_1.PublicKey('jwCXPTY4reNNvbKmTE1YAVZ4ZUvy3ozhLMt8fbPikvY'),
                new web3_js_1.PublicKey('Cbgo7WrxWW46cy49TzxhzpoJVcCzEd8r6pVWJFiiPkTL'),
                new web3_js_1.PublicKey('H7gU74WroNC6DLHBywHTWEnfEsamxLiUn66uRpYueEPA'),
                new web3_js_1.PublicKey('FC6yGPRKoeJz8iVPgvUjyzuie5sBa3FLhuP2tfdNfBpa'),
                new web3_js_1.PublicKey('HWg7yB3C1BnmTKFMU3KGD7E96xx2rUhv4gxrwbZLXHBt'),
                new web3_js_1.PublicKey('8XgX1EkSHC43mwdaUCZeXL2JVFz15JJynFrcxrQa3jss'),
                new web3_js_1.PublicKey('CKGQS4pncGsBZUcp5SwTJBch3PbNchkzo74B8rT1ftaw'),
                new web3_js_1.PublicKey('DEJkEMwVuy8c1y6xJ3TgiCEwMKox8uP3CQPtnnothmNr'),
                new web3_js_1.PublicKey('8yDtTni74G9yTanRTgZmgmJQqigpSczpXeyTvJJauf1N'),
                new web3_js_1.PublicKey('H8vjB9c818uxCYytMcQbabHaWKdRXcrKcKXhvPJGNKB3'),
                new web3_js_1.PublicKey('FrXU5FSxdABjXugKNhmcaVbsc2gi5K8Xxdu262wsfo7r'),
                new web3_js_1.PublicKey('4pSBscwxa6qmeiy3UMfLgW9t1akj18eY4hQC9wTkeb7C'),
                new web3_js_1.PublicKey('684Nbg5uSCy1RsdomNEHH1JZNvdxgyfzjWhe2TRyXrUa'),
                new web3_js_1.PublicKey('6dVR2NQ1M97X5onE1Senjyy3aRZZ5PSxUvbWWSmLNQsU'),
                new web3_js_1.PublicKey('9LtLV3PreTYAXz7MPcg4ZgSu923zFdJXkD7TA5TqhFKy'),
                new web3_js_1.PublicKey('Hqaim4KaeXSrC9rUQC7on4eBX8UguLY4dYSb7VmtFARB'),
                new web3_js_1.PublicKey('3BSZDisLFynSVtLTbRnL8ozGJAHr7tyvH4kaJRQYunWc'),
                new web3_js_1.PublicKey('EWzFK2LZKHvnWMYaGa6YvhHkAW6CptPQ3CCYkVgmAaXK'),
                new web3_js_1.PublicKey('7ceDYDQvEoyq2i9m3haFjWAtfgcUg6qWpihCXoTsDDXu'),
                new web3_js_1.PublicKey('AjBqzzHKQ7QXukKYXxuzRCDsbRGcSJyCxjVmREXgkBMN'),
                new web3_js_1.PublicKey('GscnFgC7WLpgcg8Mop3bX6g7jRvwaQ2BUa564pdimwUW'),
                new web3_js_1.PublicKey('C13WaZzdiJbgyWcea77qzBfKKvoyVwTCXYPMJpHYJPHK'),
                new web3_js_1.PublicKey('66LhxPt4AKJcY7HveisJcPEhtC8SKwS3JeDYfSswV2hJ'),
                new web3_js_1.PublicKey('ExhJyCcuyztviV1azRwNntknGXEN67JFWYXxYdNAM8f6'),
                new web3_js_1.PublicKey('HYVCvwo9ziMCxB2zkAhfwPDTj6mrx2tVHX47iEFsho8M'),
                new web3_js_1.PublicKey('46HF3cV4JBkbPbiT6HsXMkBh9qLsmifLzt8F8iNcPdnv'),
                new web3_js_1.PublicKey('BX3pVJpJqLmwaLGhboDm1uLKnvk2m3Xfd8z1doT93jmb'),
                new web3_js_1.PublicKey('frAcfwyZKKVemA3pAS7E97UiX82AqMkCn56TnVuBWjL'),
                new web3_js_1.PublicKey('AKxzCUr6DVKqWUbcs3CkmBFXA2Zb1WQypEd9P7CmeJ9R'),
                new web3_js_1.PublicKey('DA7BoZkoW4ipPv2Eyo5X9C5N1sZbZCXtsxjZf5oiXeyn'),
                new web3_js_1.PublicKey('GxtrhNXMhujdpL89upXKJoWTzYNup2wmvVtcfW5w5vCP'),
                new web3_js_1.PublicKey('24kpdg4PUZtfgiX6QNd9fJ5N4SHKQ7thHCeQH4wVBqMA'),
                new web3_js_1.PublicKey('AuGcMJudAZJToBw6J1gMjjwwnbgcns88RNEoHUKBmSjY'),
                new web3_js_1.PublicKey('3naDtkh65JJuduEEf2jtUaSkbn5T3XRM9kKFWDDVKZUe'),
                new web3_js_1.PublicKey('DJk2hmmSTVouCooFEzzVzJr1DG9cXtaxz6T5rVbKABZe'),
                new web3_js_1.PublicKey('3cYLhMVrfJWuf6kD8sF9PcrKT9duXUNcnNV6Z1LCJcxH'),
                new web3_js_1.PublicKey('9DBkyNa3vRx8dPqjEEohmCFP3SuFiGYmeNiWMvBpfMCU'),
                new web3_js_1.PublicKey('8bPXpgGxdaiys9gnJboDva8Baujn4K8pjE4jZ1Bo3QCS'),
                new web3_js_1.PublicKey('7ypH9hpQ6fLRXCVdK9Zb6zSgUvzFp44EN7PWfWdUBDb5'),
                new web3_js_1.PublicKey('5FRS6gW5Fb63AiBxF3rtYrn52Sv8GLrkp2R9cH9ppZQm'),
                new web3_js_1.PublicKey('BDKaTfCrmeSihXEHjLFGvJyLGiodP7n3g7zFrHKTnE6x'),
                new web3_js_1.PublicKey('6K4jM79yijUEFxdFhCFZSjav1nZji1gsxUWQE6XrC8YD'),
                new web3_js_1.PublicKey('5aLP9N7W4sWjeUzUCBiH1mxGNcKxfMghDrG6Eaik9mWd'),
                new web3_js_1.PublicKey('2Kdtutrfzfg6LpAUYQbnkAf9H2GdA9KyHDmwPt9UqcyF'),
                new web3_js_1.PublicKey('3m4Qb7YYpf4rZJkba9TAVDL2MnzfraXKkQgCvhJEKRJs'),
                new web3_js_1.PublicKey('2dczcMRpxWHZTcsiEjPT4YBcSseTaUmWFzw24HxYMFod'),
                new web3_js_1.PublicKey('ANPPfLuytJ6c9rGeiARd6XhMRWvsLpAkUT3spocBv2fe'),
                new web3_js_1.PublicKey('EG5DpbnuQQ988m3m1ZDpa6aPA6MnZzkikWZ6cWwjweia'),
                new web3_js_1.PublicKey('4yD8SKwoA8B7k21dvsxRaTLr79CbFHq29PBFnwcCuc2s'),
                new web3_js_1.PublicKey('9ojKRNtj8W5GvLFxpMrb7iLqGq1RkbzcHGBbGWmcpYii'),
                new web3_js_1.PublicKey('B6CAW9Q8nzgDxT81YjaSdf3kQwrLghVbdiAJr7nN47px'),
                new web3_js_1.PublicKey('4dtWi97ZfCaPyNMa89cCnGrYps4zu6qXBYNK4PTDEZjV'),
                new web3_js_1.PublicKey('2KDqcdqBbeqg4u94BTkUfwRunB6ja3JPwcQDrYTdWGYV'),
                new web3_js_1.PublicKey('Cr2K3qyqPMekk2QTwcw2a7x3gr8Mqpqa7vyTfyE6RYLJ'),
                new web3_js_1.PublicKey('FAVSpnZsNWKTnPmn4qPttZjT6MiWyCQjUVzDuf6pLcTB'),
                new web3_js_1.PublicKey('Ce8cvm76czMD9vjiqRP6ZkFUQfiTrustpS9XyuWtddeQ'),
                new web3_js_1.PublicKey('H7ynJmJMo2GiA1jFK6kNaY7s5zrQtAZP8E57UokQXaEn'),
                new web3_js_1.PublicKey('2YS1k4i2xrThW2eZ8cs6m1q7yDTRzRsWKs6KnBBLR31P'),
                new web3_js_1.PublicKey('HWNE4fSF1RqB6CJ4iSC353AEn8bf9fm7XYydsbCt6qJm'),
                new web3_js_1.PublicKey('BwX83UT4tgX8cHWzQSoEwBFueYZsDU69kV4YQNVSkJ2C'),
                new web3_js_1.PublicKey('GymZwgYmamLpUTL3eBKA4XXbCAVmZkSxGcny1AYwxGgx'),
                new web3_js_1.PublicKey('7khjhQbJcV1BiqZpYNUXiVn6CtvMDQJwPZ8VbZ1z6wxX'),
                new web3_js_1.PublicKey('5ZQ3CwWS6JEJJocuWwPeH6qbH7T2N4yCjgSaVzznEWCA'),
                new web3_js_1.PublicKey('AcsMjKfaafpuTo9vrx7PqFSxJi4E8rrPYEDqrCgZojve'),
                new web3_js_1.PublicKey('Dz9j5w8yDNKoeKyExqUWJfe84JjL8dPiRNk6FrGmS2Lc'),
                new web3_js_1.PublicKey('H656JiomQDU8Wxybmap22SmFnXSYGz8iV42fXz2HMGNE'),
                new web3_js_1.PublicKey('6dCdRpwQB2GVXshcpVyDFVZA5FkfUNjmZCVKA4m1RiBh'),
                new web3_js_1.PublicKey('CFqiPtUpiKh18o1KYtncLDsNpkTNVwvceGN8z1mGmvD1'),
                new web3_js_1.PublicKey('GgJ3dBamB7vgWhq3Zmc7RXYbDKD6Ax4AQXx3MSeK95DV'),
                new web3_js_1.PublicKey('9Fx9C2mmEre5hwXNBxzZb6Uscp773GSMaKTzjRMQQCkt'),
                new web3_js_1.PublicKey('4AKAPqvQmmnJfkfohW59MsAYBYRu9eNJB4Xf4nJ2Yw2D'),
                new web3_js_1.PublicKey('5Q6wwYjZPMWkHKrYohSMNxrqz3m6F42hEcRnSDy94rD4'),
                new web3_js_1.PublicKey('3WL7EiRVG3Soct8Cfv9xnbgA2nhSULzyEUR668oZHctq'),
                new web3_js_1.PublicKey('Hfx4dVpn632h3XX3GWc11y8b7ouZRjyQjjZ8thPCjcqg'),
                new web3_js_1.PublicKey('ES4MhbocF5D98AeG2Fk7A564J99V876L5ZA77nSUnNVY'),
                new web3_js_1.PublicKey('2M1Qf5C8zJAxkqKm2AK3H2xzYxWuMbhA2rzzPig9j247'),
                new web3_js_1.PublicKey('9Qk5HxNpnXxbpaD8y9pfBvMWQYqABpHG8mrR1fjJtNAW'),
                new web3_js_1.PublicKey('FMeUbyzofVutHpeqUco9peqAyDoPLZCjvBZuV2cduwxA'),
                new web3_js_1.PublicKey('Mn6FeUkDHkrxNYh9iCtty85peHf4YtnKTqjaqQkE4R3'),
                new web3_js_1.PublicKey('HfiqWFbXWQM9ZTnWQLBvv4Qsvg58iLiPsfGvEiXeNx4R'),
                new web3_js_1.PublicKey('7pwXHZ4MSAT4CWJBB8uYi3teCxBXwgi8F6DLm39EwA3X'),
                new web3_js_1.PublicKey('6VqFtL9WWsuG97mSY7vD8sgGBwAgjoiRXUxaZFnZsZ8E'),
                new web3_js_1.PublicKey('Cgb4iehuTNAgaafXF9Y9e8N3wFpcrqbHC2vvCoBdBJXY'),
                new web3_js_1.PublicKey('GXzow9dJwKk2QQLW779znijWrZBEFL5YC14CdMhASUU8'),
                new web3_js_1.PublicKey('EDECTcgfdjdpwVPAgac4oAQwwW1oET4ZAqfRSjLaCT4v'),
                new web3_js_1.PublicKey('8RsjvJ9VoLNJb5veXzbyc7DKqvPG296oY2BsPnuxPTQ2'),
                new web3_js_1.PublicKey('CY8s6rQztkDeZsaDLaApvweCLjKNDvfdjyKvA4rGjqUc'),
                new web3_js_1.PublicKey('7uGCZJSepSkWqujUw3Cz2sjXg4GW7rzJNXiKe13L7Am9'),
                new web3_js_1.PublicKey('BxwyGBEdfmqZwvko3ZY8XaNhW7u6gFpBbDnu2XDxW9d9'),
                new web3_js_1.PublicKey('GnzcyPiBJUiLWCWwVvnKATdKiwzUMJZPg4c3YoUAQvB5'),
                new web3_js_1.PublicKey('5BUR22MZR9YivcYpsx9xp2aDJGKBcF9HWCYynBPw6r8'),
                new web3_js_1.PublicKey('8GFK2j6x6fEi3o7db4ounurD8PDYafSpeJf8off3qK8p'),
                new web3_js_1.PublicKey('J1Rm2S4v1ojsqtqVoh6NwuzQkv9v71hhoRyoBbdvH5fa'),
                new web3_js_1.PublicKey('CEz5keL9hBCUbtVbmcwenthRMwmZLupxJ6YtYAgzp4ex'),
                new web3_js_1.PublicKey('DKbpb5qxHhc2d4U7CmTBt4BF1eVvfkBsqxKLM4BCTcdR'),
                new web3_js_1.PublicKey('8ewNqtPY3WtGmq6opBDBJMvGu5B3zshEFG67dJRsNATz'),
                new web3_js_1.PublicKey('AepjvYK4QfGhV3UjSRkZviR2AJAkLGtqdyKJFCf9kpz9'),
                new web3_js_1.PublicKey('5BAAVFfKb7xNF4Kscgtc7kqLtEjjZ2hM8NkE43J2nte8'),
                new web3_js_1.PublicKey('GfpUVgQqk9SNcjP11XUVM8hvtAAD5mMwhY7fRWSZD8H2'),
                new web3_js_1.PublicKey('9xXg8XD9NQvpK2XzK1ARSJSaFEpUNx2L63FrnYFVieQN'),
                new web3_js_1.PublicKey('3xUYbqcgTm4t5cbs7va6441evoCoqZq3HofdgUFNfTMS'),
                new web3_js_1.PublicKey('DTzvmLytpKEwJCKocWZeFYJiKCiCWB7vHTJo5PJDZGR5'),
                new web3_js_1.PublicKey('EoXt4R1gyuPY8PZmXL95rgecszERmCqVfnxue6QpwpyX'),
                new web3_js_1.PublicKey('CVt8bRum8gdXRLAnuxt66e2vtC9PdDexuhKKyDk87eno'),
                new web3_js_1.PublicKey('BDFQqTfLVpSxrQYgZiesrEg8WxrTjcm3W8SA9MtABujh'),
                new web3_js_1.PublicKey('55vEQ9LEwo6s98q4MRWPpgYwiBtkhDz2ppEAkNbjWT1L'),
                new web3_js_1.PublicKey('2e5jv5eyVgFnyC2BzmL3r8GsqvJ5VMVecXffRsjaUxi6'),
                new web3_js_1.PublicKey('H62RwxzuHiKkSTbrieJc9Xukj6ZZHfoMGZnSGnpEsWNo'),
                new web3_js_1.PublicKey('uNn4DKMqXm8yJJRcZRY7MZVMnJxqy62mUgC2e1tHFQV'),
                new web3_js_1.PublicKey('BzJU7gd1Pzo9XY8LfXESsZBsnmKt7Zc5xdbWbY3KJHSH'),
                new web3_js_1.PublicKey('B9xEHj9AHhoaWZsCsKGNnaa7tojJRrd21ak7i4CevGhD'),
                new web3_js_1.PublicKey('HEsY45jM3tkrxHxpYyaxhvDfJEDub27unKXYZ6zFnjpt'),
                new web3_js_1.PublicKey('DgwAHUnarWGyE7fxEnqoJGBgp6wy5e41FmiwXXVxGcBv'),
                new web3_js_1.PublicKey('CJhyKExm1sqFkcRnemEQj847TgqEsPEgE5GbiHFqntkH'),
                new web3_js_1.PublicKey('CJQSJhT1NWaksQgD4yGN439MP46raZ4qDCo9aQ3vfWhH'),
                new web3_js_1.PublicKey('ANXieKhnr6Bnz7S8CtCvFuVHdP6KKFJw7XSq9aP7LNbR'),
            ],
            programId: new web3_js_1.PublicKey('6LtLpnUFNByNXLyCoK9wA2MykKAmQNZKBdY8s47dehDc'),
            globalConfig: new web3_js_1.PublicKey('GKnHiWh3RRrE1zsNzWxRkomymHc374TvJPSTv2wPeYdB'),
            collateralInfos: new web3_js_1.PublicKey('3v6ootgJJZbSWEDfZMA1scfh7wcsVVfeocExRxPqCyWH'),
        },
        scope: {
            oracleMappings: new web3_js_1.PublicKey('Chpu5ZgfWX5ZzVpUx9Xvv4WPM75Xd7zPJNDPsFnCpLpk'),
            oraclePrices: new web3_js_1.PublicKey('3NJYftD5sjVfxSnUdZ1wVML8f3aC6mp1CXCL6L7TnU8C'),
            programId: new web3_js_1.PublicKey('HFn8GnPADiny6XqUoWE8uRPPxb29ikn4yTuPa9MF2fWJ'),
            configurationAccount: new web3_js_1.PublicKey('AdTiP7QyjUyv6crF4H8z7fxJKU7Z5eCAGvJN1Y55cXxb'),
        },
        borrowing: {
            programId: new web3_js_1.PublicKey('HubbLeXBb7qyLHt3x7gvYaRrxQmmgExb7fCJgDqFuB6T'),
            accounts: {
                stablecoinMint: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                stablecoinMintAuthority: new web3_js_1.PublicKey('AkzcmXAWWmGnVmyEfW5eeaHdUUrE9yxGaS2GmtV5UEMq'),
                hbbMintAuthority: new web3_js_1.PublicKey('CNDdAftLoTNZUyb3Tb4WHFZaKDuGYUtZ23gUhYdzKZS'),
                mint: {
                    ETH: new web3_js_1.PublicKey('7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs'),
                    BTC: new web3_js_1.PublicKey('9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E'),
                    SRM: new web3_js_1.PublicKey('SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt'),
                    RAY: new web3_js_1.PublicKey('4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R'),
                    FTT: new web3_js_1.PublicKey('AGFEad2et2ZJif9jaGpdMixQqvW5i81aBdvKe7PHNfz3'),
                    HBB: new web3_js_1.PublicKey('HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6'),
                    WSOL: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    MSOL: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    LDO: new web3_js_1.PublicKey('HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p'),
                    ORCA: new web3_js_1.PublicKey('orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE'),
                },
                collateralVaultsAuthority: new web3_js_1.PublicKey('HZYHFagpyCqXuQjrSCN2jWrMHTVHPf9VWP79UGyvo95L'),
                collateralVault: {
                    SOL: new web3_js_1.PublicKey('68yQyYj4YymoiQ644LVyRdvWLpw9xsjp67kmZhLMGnQp'),
                    ETH: new web3_js_1.PublicKey('HWTdcMDUF2UJqmD39dix2ATbKTJYQVhpPyoHgJTthpEz'),
                    BTC: new web3_js_1.PublicKey('4EhioLhAgbX18VGmX5Qf7qBGB7HEDQELcBxEGTk6c38G'),
                    SRM: new web3_js_1.PublicKey('5p8atdH8vX1mKgA38DPCC6tWGuGCiMRmZqFg1KrsECTW'),
                    RAY: new web3_js_1.PublicKey('BcmAzZsn6uZL8Hsv4ZWidXFshPS5oSyBXetf2ZD2i5i7'),
                    FTT: new web3_js_1.PublicKey('5EpCf866wXEgPFyTstgmJZJK1TPZJ3d1Jcvi2mT6EXFj'),
                    MSOL: new web3_js_1.PublicKey('9FWk5tni1jv6muvuUeh2JJdcfGcuzShkh2Ashzr6QWch'),
                },
                redemptionsQueue: new web3_js_1.PublicKey('7PE2Ac3BM7pj7eoYRqBDzbKehwcfWQf8RCeFUy6AmVT5'),
                redemptionCandidatesQueue: new web3_js_1.PublicKey('9MDc3EZR9cZvhB2sprPh2Tz6n1p9NNJMXiP7ZWPNSbpN'),
                liquidationsQueue: new web3_js_1.PublicKey('3nBiYphUBFv8RGHNZYsTEFQthYnHBDoFJrSXfL59dNu7'),
                borrowingMarketState: new web3_js_1.PublicKey('YSp9bJpQom2HA7VThxYZX9pNwrcBH88NBcBNWs5yskR'),
                borrowingMarketStates: [
                    new web3_js_1.PublicKey('YSp9bJpQom2HA7VThxYZX9pNwrcBH88NBcBNWs5yskR'),
                    new web3_js_1.PublicKey('B6gckY3KmWR2u1gwruE55Hxoj5uqiTTtuXLbE6Bu7bZq'),
                    new web3_js_1.PublicKey('2catbAPzCLX6YZA1axobPEYwbNo1XQaqhvWrZfR4Yxms'),
                ],
                borrowingVaultsV2: new web3_js_1.PublicKey('EUEukBTF2X4gtFzkAJaqBcxUVbrSLn2yxLDBuKJj5JoZ'),
                stabilityPoolState: new web3_js_1.PublicKey('245U3MMJ57YDGUSYRQxsJurqUhnXK4mjjvavbrvSn3uh'),
                epochToScaleToSum: new web3_js_1.PublicKey('HDZR8Knvh9ETmCFGig3ocmzxijVgSrtPYSNCFoj2muRf'),
                stablecoinStabilityPoolVault: new web3_js_1.PublicKey('GE6PqKhREVLRTGZs7rV5mJ2L3FAcj2fqohWCzTjhCTdQ'),
                stabilityVaults: new web3_js_1.PublicKey('DytfDS9DqvCe86TGcFusmGpQgWa2889dcn34qZ1M5UYQ'),
                stablecoinStabilityPoolVaultAuthority: new web3_js_1.PublicKey('CpNGqip6qTSmYEa3ANtPdHUzgD6HSWKeDE8ovNUtQLQy'),
                liquidationRewardsVaultSol: new web3_js_1.PublicKey('gbbqSXucKBYcfvaPR2DevzExFpstvE6Ypu1Nfs3HGB1'),
                liquidationRewardsVaultSrm: new web3_js_1.PublicKey('2hwuFn23CqZdsQ2hCbjRHLuNeHra1zfk6qAurTjEL133'),
                liquidationRewardsVaultEth: new web3_js_1.PublicKey('CyfPwr1uV3pkhQtCdCpLvzzVvAngXYqq4KSAvkVxqhga'),
                liquidationRewardsVaultBtc: new web3_js_1.PublicKey('EQxPELJWp87ziPFm6cnWEqqbGuzwbfbE4ejXgoBKfdJg'),
                liquidationRewardsVaultRay: new web3_js_1.PublicKey('3aqwFikJ1sLwebd3GDdPxFPnzTaPVEQp9J7R3stNfPHL'),
                liquidationRewardsVaultFtt: new web3_js_1.PublicKey('77PfF5Nqx8r3ZwLccxhyW9jmEYNLKrYTa4ruB8uAXQSA'),
                liquidationRewardsVaultMsol: new web3_js_1.PublicKey('5xvgjTUyR5yBEW6oKYTeTd6BaQobGt7ECLPUearAxjBA'),
                liquidationRewardsVaultAuthority: new web3_js_1.PublicKey('2LhXfDs64j5jVXNKa1qapMCMAZ9LXvKwHxpkp55wFh54'),
                borrowingVaults: new web3_js_1.PublicKey('GNSn4pfNqxq1NeBoM3DSLrTp2oaQqptRy7MuaNpzA5fA'),
                globalConfig: new web3_js_1.PublicKey('8HuhMF7ppoYdPD9fAdn9d9ZU1rL4ipty6CJEcZ5F17o1'),
                oracleMappings: new web3_js_1.PublicKey('GKZhrh5reKiRkxkvQiFPT6ACtnVDB16DWbF1vfva5iod'),
                stakingPoolState: new web3_js_1.PublicKey('5YPVVnBGYhribwt4UHCRWNKRs28DdgUgXQj8VrWikBP7'),
                borrowingFeesAccount: new web3_js_1.PublicKey('8PJGWbSjY9C45yq9Jyr2GjMANQ3CuVajMGZwkgnNSBRF'),
                borrowingFeesVaultAuthority: new web3_js_1.PublicKey('EDmzrjSjeS36dcYRzHuzrPxRnjZgfQoudfGGUA9mXyZr'),
                stakingVault: new web3_js_1.PublicKey('9ixberQazk45T5XN3fZPakK2dWBdgJ6dyJeykxw3FN4K'),
                treasuryVault: new web3_js_1.PublicKey('G1apbPBr9GdBWaXY4ejMu8b3NnmByvKT3BTvzEdKCgq6'),
                burningVault: new web3_js_1.PublicKey('Ec7EqbEzbY5EW18E3scKoqYMYw6EGxySrRDpmCrs95e3'),
                burningVaultAuthority: new web3_js_1.PublicKey('9wyZzZ77wg9JR25yiJi4333Fb2t3jiEVUGHoDZS6cdrE'),
                USDC: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
                pyth: {
                    solProductInfo: new web3_js_1.PublicKey('ALP8SdU9oARYVLgLR7LrqMNCYBnhtnQz1cj6bwgwQmgj'),
                    solPriceInfo: new web3_js_1.PublicKey('H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG'),
                    ethProductInfo: new web3_js_1.PublicKey('EMkxjGC1CQ7JLiutDbfYb7UKb3zm9SJcUmr1YicBsdpZ'),
                    ethPriceInfo: new web3_js_1.PublicKey('JBu1AL4obBcCMqKBBxhpWCNUt136ijcuMZLFvTP7iWdB'),
                    btcProductInfo: new web3_js_1.PublicKey('4aDoSXJ5o3AuvL7QFeR6h44jALQfTmUUCTVGDD6aoJTM'),
                    btcPriceInfo: new web3_js_1.PublicKey('GVXRSBjFk6e6J3NbVPXohDJetcTjaeeuykUpbQF8UoMU'),
                    srmProductInfo: new web3_js_1.PublicKey('6T1eJbKWyhZXEjmuBej9gPk3SRTagzrEQovCSwAWD57P'),
                    srmPriceInfo: new web3_js_1.PublicKey('3NBReDRTLKMQEKiLD5tGcx4kXbTf88b7f2xLS9UuGjym'),
                    rayProductInfo: new web3_js_1.PublicKey('EjtaES3pbMCaNrVFtZ5sT36WPXAnARZxCFsoEr2hsUKJ'),
                    rayPriceInfo: new web3_js_1.PublicKey('AnLf8tVYCM816gmBjiy8n53eXKKEDydT5piYjjQDPgTB'),
                    fttProductInfo: new web3_js_1.PublicKey('E7ymZgACaSd1fYLf2V7WkAmFkpRiWfAjm49R7bceny9K'),
                    fttPriceInfo: new web3_js_1.PublicKey('8JPJJkmDScpcNmBRKGZuPuG2GYAveQgP3t5gFuMymwvF'),
                    msolProductInfo: new web3_js_1.PublicKey('BS2iAqT67j8hA9Jji4B8UpL3Nfw9kwPfU5s4qeaf1e7r'),
                    msolPriceInfo: new web3_js_1.PublicKey('E4v1BBgoso9s64TQvmyownAVJbhbEPGyzA3qn4n46qj9'),
                },
                serumMarkets: {
                    SOL: {
                        marketAddress: new web3_js_1.PublicKey('9wFFyRfZBsuAha4YcuxcXLKwMxJR43S7fPfQLusDBzvT'),
                        requestQueue: new web3_js_1.PublicKey('AZG3tFCFtiCqEwyardENBQNpHqxgzbMw8uKeZEw2nRG5'),
                        eventQueue: new web3_js_1.PublicKey('5KKsLVU6TcbVDK4BS6K1DGDxnh4Q9xjYJ8XaDCG5t8ht'),
                        baseVault: new web3_js_1.PublicKey('36c6YqAwyGKQG66XEp2dJc5JqjaBNv7sVghEtJv4c7u6'),
                        quoteVault: new web3_js_1.PublicKey('8CFo8bL8mZQK8abbFyypFMwEDd8tVJjHTTojMLgQTUSZ'),
                        vaultOwner: new web3_js_1.PublicKey('F8Vyqk3unwxkXukZFQeYyGmFfTG3CAX4v24iyrjEYBJV'),
                        bidsAddress: new web3_js_1.PublicKey('14ivtgssEBoBjuZJtSAPKYgpUK7DmnSwuPMqJoVTSgKJ'),
                        asksAddress: new web3_js_1.PublicKey('CEQdAFKdycHugujQg9k2wbmxjcpdYZyVLfV9WerTnafJ'),
                    },
                    ETH: {
                        marketAddress: new web3_js_1.PublicKey('8Gmi2HhZmwQPVdCwzS7CM66MGstMXPcTVHA7jF19cLZz'),
                        requestQueue: new web3_js_1.PublicKey('3ZSxZjD8o8JjPX1HVmQ59ED89R3uKNviRDCEmxCgv9dp'),
                        eventQueue: new web3_js_1.PublicKey('3z4QQPFdgNSxazqEAzmZD5C5tJWepczimVqWak2ZPY8v'),
                        baseVault: new web3_js_1.PublicKey('8cCoWNtgCL7pMapGZ6XQ6NSyD1KC9cosUEs4QgeVq49d'),
                        quoteVault: new web3_js_1.PublicKey('C7KrymKrLWhCsSjFaUquXU3SYRmgYLRmMjQ4dyQeFiGE'),
                        vaultOwner: new web3_js_1.PublicKey('FG3z1H2BBsf5ekEAxSc1K6DERuAuiXpSdUGkYecQrP5v'),
                        bidsAddress: new web3_js_1.PublicKey('3nXzH1gYKM1FKdSLHM7GCRG76mhKwyDjwinJxAg8jjx6'),
                        asksAddress: new web3_js_1.PublicKey('b3L5dvehk48X4mDoKzZUZKA4nXGpPAMFkYxHZmsZ98n'),
                    },
                    BTC: {
                        marketAddress: new web3_js_1.PublicKey('A8YFbxQYFVqKZaoYJLLUVcQiWP7G2MeEgW5wsAQgMvFw'),
                        requestQueue: new web3_js_1.PublicKey('H6UaUrNVELJgTqao1CNL4252kShLKSfwoboT8tF7HNtB'),
                        eventQueue: new web3_js_1.PublicKey('6NQqaa48SnBBJZt9HyVPngcZFW81JfDv9EjRX2M4WkbP'),
                        baseVault: new web3_js_1.PublicKey('GZ1YSupuUq9kB28kX9t1j9qCpN67AMMwn4Q72BzeSpfR'),
                        quoteVault: new web3_js_1.PublicKey('7sP9fug8rqZFLbXoEj8DETF81KasaRA1fr6jQb6ScKc5'),
                        vaultOwner: new web3_js_1.PublicKey('GBWgHXLf1fX4J1p5fAkQoEbnjpgjxUtr4mrVgtj9wW8a'),
                        bidsAddress: new web3_js_1.PublicKey('6wLt7CX1zZdFpa6uGJJpZfzWvG6W9rxXjquJDYiFwf9K'),
                        asksAddress: new web3_js_1.PublicKey('6EyVXMMA58Nf6MScqeLpw1jS12RCpry23u9VMfy8b65Y'),
                    },
                    SRM: {
                        marketAddress: new web3_js_1.PublicKey('ByRys5tuUWDgL73G8JBAEfkdFf8JWBzPBDHsBVQ5vbQA'),
                        requestQueue: new web3_js_1.PublicKey('Hr8Z93aWe4hhJbC5i7YTsPaSToziVh3vyMfv9GRqKFCh'),
                        eventQueue: new web3_js_1.PublicKey('6o44a9xdzKKDNY7Ff2Qb129mktWbsCT4vKJcg2uk41uy'),
                        baseVault: new web3_js_1.PublicKey('Ecfy8et9Mft9Dkavnuh4mzHMa2KWYUbBTA5oDZNoWu84'),
                        quoteVault: new web3_js_1.PublicKey('hUgoKy5wjeFbZrXDW4ecr42T4F5Z1Tos31g68s5EHbP'),
                        vaultOwner: new web3_js_1.PublicKey('GVV4ZT9pccwy9d17STafFDuiSqFbXuRTdvKQ1zJX6ttX'),
                        bidsAddress: new web3_js_1.PublicKey('AuL9JzRJ55MdqzubK4EutJgAumtkuFcRVuPUvTX39pN8'),
                        asksAddress: new web3_js_1.PublicKey('8Lx9U9wdE3afdqih1mCAXy3unJDfzSaXFqAvoLMjhwoD'),
                    },
                    RAY: {
                        marketAddress: new web3_js_1.PublicKey('2xiv8A5xrJ7RnGdxXB42uFEkYHJjszEhaJyKKt4WaLep'),
                        requestQueue: new web3_js_1.PublicKey('39mE6bYktM1XAKKmB6WN971X3Sa1yGkHxtCTWMkVrwN2'),
                        eventQueue: new web3_js_1.PublicKey('H9dZt8kvz1Fe5FyRisb77KcYTaN8LEbuVAfJSnAaEABz'),
                        baseVault: new web3_js_1.PublicKey('GGcdamvNDYFhAXr93DWyJ8QmwawUHLCyRqWL3KngtLRa'),
                        quoteVault: new web3_js_1.PublicKey('22jHt5WmosAykp3LPGSAKgY45p7VGh4DFWSwp21SWBVe'),
                        vaultOwner: new web3_js_1.PublicKey('FmhXe9uG6zun49p222xt3nG1rBAkWvzVz7dxERQ6ouGw'),
                        bidsAddress: new web3_js_1.PublicKey('Hf84mYadE1VqSvVWAvCWc9wqLXak4RwXiPb4A91EAUn5'),
                        asksAddress: new web3_js_1.PublicKey('DC1HsWWRCXVg3wk2NndS5LTbce3axwUwUZH1RgnV4oDN'),
                    },
                    FTT: {
                        marketAddress: new web3_js_1.PublicKey('2Pbh1CvRVku1TgewMfycemghf6sU9EyuFDcNXqvRmSxc'),
                        requestQueue: new web3_js_1.PublicKey('2ncWnY9QuCdNAX4gLutkNoKr9R1owDY1bhLLV1yH8Ukf'),
                        eventQueue: new web3_js_1.PublicKey('2XHxua6ZaPKpCGUNvSvTwc9teJBmexp8iMWCLu4mtzGb'),
                        baseVault: new web3_js_1.PublicKey('4LXjM6rptNvhBZTcWk4AL49oF4oA8AH7D4CV6z7tmpX3'),
                        quoteVault: new web3_js_1.PublicKey('2ycZAqQ3YNPfBZnKTbz2FqPiV7fmTQpzF95vjMUekP5z'),
                        vaultOwner: new web3_js_1.PublicKey('B5b9ddFHrjndUieLAKkyzB1xmq8sNqGGZPmbyYWPzCyu'),
                        bidsAddress: new web3_js_1.PublicKey('9HTDV2r7cQBUKL3fgcJZCUfmJsKA9qCP7nZAXyoyaQou'),
                        asksAddress: new web3_js_1.PublicKey('EpnUJCMCQNZi45nCBoNs6Bugy67Kj3bCSTLYPfz6jkYH'),
                    },
                    MSOL: {
                        marketAddress: new web3_js_1.PublicKey('6oGsL2puUgySccKzn9XA9afqF217LfxP5ocq4B3LWsjy'),
                        requestQueue: new web3_js_1.PublicKey('EHUoDPVVKR5Udp4EZPb4bsHZte5EEHc1PPTPXBgJEPEK'),
                        eventQueue: new web3_js_1.PublicKey('BC8Tdzz7rwvuYkJWKnPnyguva27PQP5DTxosHVQrEzg9'),
                        baseVault: new web3_js_1.PublicKey('2y3BtF5oRBpLwdoaGjLkfmT3FY3YbZCKPbA9zvvx8Pz7'),
                        quoteVault: new web3_js_1.PublicKey('6w5hF2hceQRZbaxjPJutiWSPAFWDkp3YbY2Aq3RpCSKe'),
                        vaultOwner: new web3_js_1.PublicKey('9dEVMESKXcMQNndoPc5ji9iTeDJ9GfToboy8prkZeT96'),
                        bidsAddress: new web3_js_1.PublicKey('8qyWhEcpuvEsdCmY1kvEnkTfgGeWHmi73Mta5jgWDTuT'),
                        asksAddress: new web3_js_1.PublicKey('PPnJy6No31U45SVSjWTr45R8Q73X6bNHfxdFqr2vMq3'),
                    },
                },
                saber: {
                    stablecoinSwap: new web3_js_1.PublicKey('MARpDPs5A7XiyCWPNH8GsMWPLxmwNn9SBmKvPa9LzgA'),
                },
            },
        },
        kaminoLending: [
            {
                lendingMarket: new web3_js_1.PublicKey('ABPeWVeRuvii6HuzTYcp1iH5a3ELfeVEviCtm3GfNnLV'),
                lendingMarketAuthority: new web3_js_1.PublicKey('9gndVt4KDpv3Nd5Y1USq8tozArzwtUySJzzCwEdJZteB'),
                isPrimary: true,
                reserves: [
                    {
                        liquidityToken: {
                            decimals: 9,
                            symbol: 'SOL',
                            mint: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                        },
                        address: new web3_js_1.PublicKey('DDJzgWNupppRR7amWVeWzqop84QuyRA8xM7AQd1WGzCz'),
                        collateralMint: new web3_js_1.PublicKey('5NAXaYxGaBHpN89t3QFt9mg1kgBH8VH2VxFvSRXrp3Fx'),
                        collateralSupply: new web3_js_1.PublicKey('HpHv41G1wWhfqdFURq9KiCfeUurZPNHdVG3VyTdmCGXx'),
                        liquiditySupply: new web3_js_1.PublicKey('BfJe1fr88wsZcbcCRpHUXBsEv5ufo9YTvrNhewPcC5e4'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('ErJBqDkonv8qYAKGLG5VXVSSz2hjPvYe2228W6YHxNRt'),
                        pythOracle: new web3_js_1.PublicKey('H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG'),
                        switchboardOracle: new web3_js_1.PublicKey('GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'USDH',
                            mint: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'),
                        },
                        address: new web3_js_1.PublicKey('5gCzanUXSCwHUvCPjRjQDSFaTxD8PgNzpgsaVi3L4nbB'),
                        collateralMint: new web3_js_1.PublicKey('BiXxYFkqmRvDWkSBJBh6fUMJwowPuQGgntQvNbWNR6LZ'),
                        collateralSupply: new web3_js_1.PublicKey('7YZdJrRdfGxn7xSBfbd8c6AMEHE99u2AoxZqBofwWk5Y'),
                        liquiditySupply: new web3_js_1.PublicKey('3uDAX3bjwoAPXocx4S1aqK32Npro2PYuUgqZPYFH897W'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('GRYZLAxLwcdj8xgqrSSNUdeyfmsrHctARLSCMzkthe2L'),
                        pythOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                        switchboardOracle: new web3_js_1.PublicKey('yeHZhfVXTwBZzYcvmnSraVuKg6AP6YvTzfD49eLr92D'),
                    },
                ],
            },
        ],
    },
    {
        cluster: 'devnet',
        kamino: {
            programId: new web3_js_1.PublicKey('E6qbhrt4pFmCotNUSSEh6E5cRQCEJpMcd79Z56EG9KY'),
            globalConfig: new web3_js_1.PublicKey('7of1j58my7FrQrPce81xUcfHARchvVFuSKEoyuB8WWZj'),
            collateralInfos: new web3_js_1.PublicKey('EZTVhq1x8PzBVj7kRSJqL2aVG115mfyAf8wDZdMESyeR'),
            mints: [
                { address: new web3_js_1.PublicKey('AcU8CfJiLEBn1x7Cgx8atX6DdebmdQtNSsJBkQyf6Q4L'), scopeToken: 'USDT' },
                { address: new web3_js_1.PublicKey('kfZGSThBdZbZY6cDSHP7NmCWa6D2hFc5kjBuGHREzZ3'), scopeToken: 'USDH' },
                { address: new web3_js_1.PublicKey('HEzfH7PCNSCU92zdk3iaKG4E9hQTErfgVGvqgywEvPzu'), scopeToken: 'USDC' },
                { address: new web3_js_1.PublicKey('ESba8sT1R7WPiE3weBU1u6UqEmXRuM8Rs1LLvXayEizf'), scopeToken: 'STSOL' },
                { address: new web3_js_1.PublicKey('2wMQTELEgEcK7V3gwZBH58WvGcRZsyaWNpebqyFugGKe'), scopeToken: 'MSOL' },
                { address: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'), scopeToken: 'SOL' },
            ],
            whirlpoolsToSnapshot: [],
            raydiumPoolsToSnapshot: [],
            strategies: [
                new web3_js_1.PublicKey('AvRMXt1MM9bdyE4yrnbVYjzYWWx8sdhkXPFbmxeb3rq6'),
                new web3_js_1.PublicKey('yT2bBMrE4AVG3BM8LHjDSr1yketzabmfp3rYimS36C9'),
                new web3_js_1.PublicKey('2H4xebnp2M9JYgPPfUw58uUQahWF8f1YTNxwwtmdqVYV'),
                new web3_js_1.PublicKey('EbRR6C3uCRwnHjvbJuuiaipvVCVLYb2ZL5B3GH4zm3FH'),
            ],
        },
        scope: {
            oracleMappings: new web3_js_1.PublicKey('4PoKkcsQakSMovMWh127gviWPpA6wGyY8uZWavccXAGE'),
            oraclePrices: new web3_js_1.PublicKey('Aprk2sZNvxZNUpgq2WuS3VNkbxh35aCqJd1GbtBZzkG5'),
            programId: new web3_js_1.PublicKey('3Vw8Ngkh1MVJTPHthmUbmU2XKtFEkjYvJzMqrv2rh9yX'),
            configurationAccount: new web3_js_1.PublicKey('qAN1wkxEzBg7AzeLC4sGT61czYvDiotZTMKojVL2o2E'),
        },
        borrowing: {
            programId: new web3_js_1.PublicKey('m9P5MBXVg33N5cWuHPd4YctGhKy3p6iDdHr7Ro2YRgr'),
            accounts: {
                stablecoinMint: new web3_js_1.PublicKey('HWzkNyXynnC3mAZAL4yKZNvWs3SwS9eBVKdzmKcxYHSH'),
                stablecoinMintAuthority: new web3_js_1.PublicKey('Bv3oYDiEfetCx5J2tBDDTefjRc1RirtjukEfHaRBoX71'),
                hbbMintAuthority: new web3_js_1.PublicKey('FeVZfsKhu4LJQNSDvq1RHwp5dNhFD3PFW1kBAk3PAYhw'),
                mint: {
                    WSOL: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    ETH: new web3_js_1.PublicKey('2DVyKwW17MxDXaLhWpWXtbVa64JBwaHmLJ1txXNJgRGi'),
                    BTC: new web3_js_1.PublicKey('6cA8Y9Jm2jW2rvUcpBHC9kUjctw9dMF3gPfWZd5bcr57'),
                    SRM: new web3_js_1.PublicKey('314VQ6r5m3sxdfB8cWJBS8MrAb5nVY5PvpcR7jN1PnY2'),
                    RAY: new web3_js_1.PublicKey('FDYnqkgmF9t8o4QxA7YYL2wSLfun2kXXgA6auQBzzXLU'),
                    FTT: new web3_js_1.PublicKey('7DmsbJXUcVwNmj84Cqd91KrEzJyLKGPHdKTqKTwHwsHE'),
                    HBB: new web3_js_1.PublicKey('GnzBpWxvN75JXxKndPwJ63ABdKDACMDWTC5iTUqGzSsD'),
                    MSOL: new web3_js_1.PublicKey('mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So'),
                    LDO: new web3_js_1.PublicKey('99nKXBggZF5Xxbc9RPD1kzD1ZP1TPXQyNSCTn1MK2Dbi'),
                    ORCA: new web3_js_1.PublicKey('orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L'),
                },
                collateralVaultsAuthority: new web3_js_1.PublicKey('5t8jZ1T6cR79vEu5Vo1Jc6Q3RaYtLsNJdPTLLu4z9vX3'),
                collateralVault: {
                    SOL: new web3_js_1.PublicKey('ZRJEmigJ8Hjs2NyTmyHkvas2HZmMKrrNAwNnJ4dFgZQ'),
                    ETH: new web3_js_1.PublicKey('DAaiG32meXcHwpxJgcAqTfaMMNhW3yKLNeJdNyaiRt2G'),
                    BTC: new web3_js_1.PublicKey('9A9ZpnXcnDZLHkHt8Fcro3XLgqGNoVJhx7ndo74MKn4B'),
                    SRM: new web3_js_1.PublicKey('84tA2z64nuBDpKN9Vm75HGDbqix51RyftDWrNrY3TfrF'),
                    RAY: new web3_js_1.PublicKey('DY3UAwAtiJXwdHjrToKzDHci5fLc2Spo3cSav5Ncsymz'),
                    FTT: new web3_js_1.PublicKey('6Dw7ofQLGHcdaMPtYBY5d37ABEtxdqk3omy2eWLzyV17'),
                    MSOL: new web3_js_1.PublicKey('9rN5H8SxGMfaabwHLg4x7xCSKpvPH218VGT4tGHmjMdH'),
                },
                liquidationsQueue: new web3_js_1.PublicKey('66QZM6UAKFpUeKWRe5spag5JEfT4ycJ1UtSzeffTiQJK'),
                borrowingMarketState: new web3_js_1.PublicKey('FqkHHpETrpfgcA5SeH7PKKFDLGWM4tM7ZV31HfutTXNV'),
                borrowingMarketStates: [
                    new web3_js_1.PublicKey('FqkHHpETrpfgcA5SeH7PKKFDLGWM4tM7ZV31HfutTXNV'),
                    new web3_js_1.PublicKey('2pjsM2weitsEP3w1Q4N7bLvFYmdNBWCw1H3E9k6rVQTy'),
                    new web3_js_1.PublicKey('5rDSgRsoq2sJQyzwLsmucvqMrNmYQDMr6L3xoXkxtadY'),
                ],
                stabilityPoolState: new web3_js_1.PublicKey('FB1e3ACdWSgKVDaVp7YRWB64JgFrUNPrxm6ukKDjjKg'),
                epochToScaleToSum: new web3_js_1.PublicKey('FPUTB5KsZQhPrJS6QmUzgAErYcfHBcyVwRSQcvCLzeVi'),
                borrowingVaultsV2: new web3_js_1.PublicKey('BHgw9kZcK6LVSadiqUZYBxe8rCzTcDq1dxsJcspLmWPX'),
                stablecoinStabilityPoolVault: new web3_js_1.PublicKey('EdSUZ9WtLj7QFrJ5wNPFZGhrf8JP1nofskpHJ8Jotgt3'),
                stabilityVaults: new web3_js_1.PublicKey('CSjeLjho28HxsWhYqzh4XqbDAoYTo63d2zn34BBRFkm6'),
                stablecoinStabilityPoolVaultAuthority: new web3_js_1.PublicKey('46ephmWbPP6pZQU8NR123NX1unJ6eYJSK3pf8pmYA8o6'),
                liquidationRewardsVaultSol: new web3_js_1.PublicKey('GdfNFLiFAxFfUbns6JaWheChqbHByBgW9EmqykEMisdm'),
                liquidationRewardsVaultSrm: new web3_js_1.PublicKey('9wFvXWJdrSyzGaKkZYhiWPmG3qKeAtemcQs5sgR7KdTV'),
                liquidationRewardsVaultEth: new web3_js_1.PublicKey('H6U9myYfmYHvz1BQBDcARinaFFVSbeHV4r9VJPiUJBxV'),
                liquidationRewardsVaultBtc: new web3_js_1.PublicKey('9Xm3RaKVgon2ViU3QwvgPE5Gcd5mTRwdzVoRQt56Pkbt'),
                liquidationRewardsVaultRay: new web3_js_1.PublicKey('2HSSye9cZnE3VS84M6z9LnobcdeYukUpZm23n3KiBve2'),
                liquidationRewardsVaultFtt: new web3_js_1.PublicKey('BTNN8XfsH3GyyUpBZUy1gKPrYhW4ng1aaiKQy4kkQtS9'),
                liquidationRewardsVaultMsol: new web3_js_1.PublicKey('DjpZNFotV5XgWgt4W7hh7TJpqgES4w4Nyh2x96sAJHYa'),
                liquidationRewardsVaultAuthority: new web3_js_1.PublicKey('83NPhEayosoeucxJGi8EPpar1oDBzu16wa8ttTUc7bJK'),
                borrowingVaults: new web3_js_1.PublicKey('2rVBBq4gbWgNCpaDAw6XScwaKxnXEEeXSXFCz5Dz8dSm'),
                globalConfig: new web3_js_1.PublicKey('4tKUbxbw9rh4FkxdSG5XgmHjGa2ygJeBdRGGnVuBYCj3'),
                oracleMappings: new web3_js_1.PublicKey('5MkiLwSR4qhhLrnfio9tmre9oepjUkf6FXwsD8Fne1nT'),
                stakingPoolState: new web3_js_1.PublicKey('ESmugrx5Toe58tr5mXWeEFVG36ifmV39QXTohekw5Bc5'),
                borrowingFeesAccount: new web3_js_1.PublicKey('CsNHgkVrHzgjbg53YdZpXqg7SzFzMDDomXGTAJJ8Xyo1'),
                borrowingFeesVaultAuthority: new web3_js_1.PublicKey('6TsuFvmMQHNCbBfkp9UBSJmSb16XvFBAp9WwSkTGHXZX'),
                stakingVault: new web3_js_1.PublicKey('4GT3ESGXLhYZGidYm2UCzyhvhhAjrFxs8pSgCakFd72h'),
                treasuryVault: new web3_js_1.PublicKey('BJBW7y4kFNidyzJQM6SAwceobzSVyBjrPi2FGCPR9wvv'),
                burningVault: new web3_js_1.PublicKey('F9ZFxRgTZQPgRRnxG8CsQ8PjoqhrQkFSCojDbCyzjFeA'),
                burningVaultAuthority: new web3_js_1.PublicKey('FtUEyCRySPDMgZZYQTdVnz7gxM7f4LeZyKPVaSNME4gb'),
                USDC: new web3_js_1.PublicKey('KYz9NWoYmvYhKHCHgJ1cT6G5WV8pZwimgbH3Seu8pJw'),
                pyth: {
                    solProductInfo: new web3_js_1.PublicKey('3Mnn2fX6rQyUsyELYms1sBJyChWofzSNRoqYzvgMVz5E'),
                    solPriceInfo: new web3_js_1.PublicKey('J83w4HKfqxwcq3BEMMkPFSppX3gqekLyLJBexebFVkix'),
                    ethProductInfo: new web3_js_1.PublicKey('2ciUuGZiee5macAMeQ7bHGTJtwcYTgnt6jdmQnnKZrfu'),
                    ethPriceInfo: new web3_js_1.PublicKey('EdVCmQ9FSPcVe5YySXDPCRmc8aDQLKJ9xvYBMZPie1Vw'),
                    btcProductInfo: new web3_js_1.PublicKey('3m1y5h2uv7EQL3KaJZehvAJa4yDNvgc5yAdL9KPMKwvk'),
                    btcPriceInfo: new web3_js_1.PublicKey('HovQMDrbAgAYPCmHVSrezcSmkMtXSSUsLDFANExrZh2J'),
                    srmProductInfo: new web3_js_1.PublicKey('6MEwdxe4g1NeAF9u6KDG14anJpFsVEa2cvr5H6iriFZ8'),
                    srmPriceInfo: new web3_js_1.PublicKey('992moaMQKs32GKZ9dxi8keyM2bUmbrwBZpK4p2K6X5Vs'),
                    rayProductInfo: new web3_js_1.PublicKey('3BtxtRxitVDcsd7pPUWUnFm9KvmNDy9usS4gE6pUFhpH'),
                    rayPriceInfo: new web3_js_1.PublicKey('EhgAdTrgxi4ZoVZLQx1n93vULucPpiFi2BQtz9RJr1y6'),
                    fttProductInfo: new web3_js_1.PublicKey('63VWd2FVbukVozZ1okHt8wVMq7enAYFXYnmp2DUQtBBJ'),
                    fttPriceInfo: new web3_js_1.PublicKey('6vivTRs5ZPeeXbjo7dfburfaYDWoXjBtdtuYgQRuGfu'),
                    msolProductInfo: new web3_js_1.PublicKey('os3is9HtWPHW4EXpGAkdr2prdWVs2pS8qKtf2ZYJdBw'),
                    msolPriceInfo: new web3_js_1.PublicKey('9a6RNx3tCu1TSs6TBSfV2XRXEPEZXQ6WB7jRojZRvyeZ'),
                },
                serumMarkets: {
                    SOL: {
                        marketAddress: new web3_js_1.PublicKey('BbNNvJ9Ayh23qdaG7vvn8y4FGkBiDipCcFa4L6sy6qkH'),
                        requestQueue: new web3_js_1.PublicKey('8jWfsdCT1tsJSBxEtVnD695Bvt6j1vTuXkFjFKXWpzxm'),
                        eventQueue: new web3_js_1.PublicKey('82SBwWDx9gcttA68fq56Cid3SPujFX3ZMaTRBW8Nncf4'),
                        baseVault: new web3_js_1.PublicKey('6aJXaJqEEA8WCoXgJgMevR7Mtg4k2kq7Bv7ZFw7UmgG7'),
                        quoteVault: new web3_js_1.PublicKey('89spP5pkVQyEBBjuDQLPyHQc5JCbjHxLmqDht2PgMwj8'),
                        vaultOwner: new web3_js_1.PublicKey('AJL9Zf7Jx6XTfQTzUhL8FATLfbHRHJ799U4ysKt2BL4k'),
                        bidsAddress: new web3_js_1.PublicKey('8RhfhGjmAHYpWa44Y3JY5aCacFpyhCJWGMC7SLzRFQbr'),
                        asksAddress: new web3_js_1.PublicKey('2wMQj33zSng8pcznYZMfGJPwNFFLH7k6DgRBcBHdVdKp'),
                    },
                    ETH: {
                        marketAddress: new web3_js_1.PublicKey('AohVcYC9UBWe6zuf9dV3okbK4YTWwTPgQtVSYgtGNk2f'),
                        requestQueue: new web3_js_1.PublicKey('3x11hXJx6FxMYfvzmmFUFiPx25wVkiEaSAsQ6yjrxEx7'),
                        eventQueue: new web3_js_1.PublicKey('2nA72kf54iPJsL9Ar5SC5Q6FXUQfz7RfXyUVHQp183TM'),
                        baseVault: new web3_js_1.PublicKey('4ZhvPAcEUBbbYM4WVe6oMCEyYE1GEBmMhVMPeKibdSrL'),
                        quoteVault: new web3_js_1.PublicKey('An8Lkd9JjDn1dJHWbijwJEWuf5wFTLkZ6ABqxU3uTS85'),
                        vaultOwner: new web3_js_1.PublicKey('2WBeH92fiSzUxTRpoD9ECLdpFPZZVmyW3SP9uWUtrT9U'),
                        bidsAddress: new web3_js_1.PublicKey('EvxjBDnJrGH6x5urDBHcCP4awEy8xgqBnXFyEgyQfgSa'),
                        asksAddress: new web3_js_1.PublicKey('7iXouzTk7mxDToqPeog5ZiimU7ZTZE2ebC2ojFmfiGuU'),
                    },
                    BTC: {
                        marketAddress: new web3_js_1.PublicKey('H3Kc3iFCjXHxK7igtYL8AYEJWUEcu6CwWrmJLuEaV5ur'),
                        requestQueue: new web3_js_1.PublicKey('EK5LtxC3AuNAHGqBvpCY4hE76v8TxvvV4qJB2tm9zeDK'),
                        eventQueue: new web3_js_1.PublicKey('59aR7wfrHgRMSqYMR1mds7MudJggka688e5ibH14Vo5g'),
                        baseVault: new web3_js_1.PublicKey('MAuEtSCnfXfEovBHTCPBDx1FUak2d8NripHAN2MnieY'),
                        quoteVault: new web3_js_1.PublicKey('2wSSHX1hhcUBPkkQ3kXL6k3N6zd8Us7eNa4E4KSho6qB'),
                        vaultOwner: new web3_js_1.PublicKey('2HP6SYHycgByXSNdTj3Pf1dwHdG9u1mydF8fHtbjx5Mm'),
                        bidsAddress: new web3_js_1.PublicKey('GCnJtgrJq4pouKquUyFGzucPQx5tLC43JW9LQgTR7Fr'),
                        asksAddress: new web3_js_1.PublicKey('7iKSZm61ZFWKSPTCKskZU6QWKbBRH7UrgB4uwWBzSXE8'),
                    },
                    SRM: {
                        marketAddress: new web3_js_1.PublicKey('8g2t7hBTFuv2q1HkhvKsf5hPGSq3uv7ztvRCFvVaG7E2'),
                        requestQueue: new web3_js_1.PublicKey('AvmtFUwV84MRYLkuZWpBQZ4MFLfentkvAJ49nGGejnQ1'),
                        eventQueue: new web3_js_1.PublicKey('FYoHfoGS7HDGsCGfrb9CeZCsMUeHHuBhN9qjUMzPaQQQ'),
                        baseVault: new web3_js_1.PublicKey('DE7gakW5Gb43n9TQSnXLmUakVGowFU1ZhJTBEXr7Cvq5'),
                        quoteVault: new web3_js_1.PublicKey('HNMxbFLg2wrgoUxHu3bTvz1XPj9n4buioou3WaQqzwrH'),
                        vaultOwner: new web3_js_1.PublicKey('HCUkFFzirnknTGb4tHTvMpQCk1TPGaZfkdXYf8JrjUL6'),
                        bidsAddress: new web3_js_1.PublicKey('A5Tq9rLx9kEdqEAXtKiSLDCfwNAfLL8d3U4S3u3sMJZc'),
                        asksAddress: new web3_js_1.PublicKey('95Z8NkfChbDsdUM5LyLwJ5dHyS94Qjth7e9cepEzuTZs'),
                    },
                    RAY: {
                        marketAddress: new web3_js_1.PublicKey('8XtEufHU7TbNNChTvBhShZGZc1adHNkTmEqv4W9Q3RYK'),
                        requestQueue: new web3_js_1.PublicKey('7GEMsQ49aUqwDnum624uvjXrZyiCnEW72zw9WsFquXMq'),
                        eventQueue: new web3_js_1.PublicKey('2nvokrESHbhrfi6QHrxoASs8ziEmcs5kPKotrvJ7xRJ2'),
                        baseVault: new web3_js_1.PublicKey('HZqiavQ6wE4siwFgvDfoxsgtgLrVNbndmULdTk3FQYHe'),
                        quoteVault: new web3_js_1.PublicKey('5KRyGWDuq8KJUQxTSEf7Ku8eFLkwCFtWBXF6WJJ6jGy9'),
                        vaultOwner: new web3_js_1.PublicKey('Hk5VVfKTL1Z4o54XwykvE4Rp9qgdZY7rUCrWfGMhLLzq'),
                        bidsAddress: new web3_js_1.PublicKey('GKagffnBN64rNgu7yFko5uzPRqWCrM5UXfsbC7ZQbFxt'),
                        asksAddress: new web3_js_1.PublicKey('54K3medHRc2Q4P1oMnKS9PoANpbyswbLC9U2jCEiEhUi'),
                    },
                    FTT: {
                        marketAddress: new web3_js_1.PublicKey('J8Y4NY8Rzk8Y2hR7JtergtL1xLXdeguD9G8aD9gmkpyu'),
                        requestQueue: new web3_js_1.PublicKey('94uGcffyTn5SFMqBKJ7ntxeLABLqy2ADf4VpfQcMtt5U'),
                        eventQueue: new web3_js_1.PublicKey('8mMsobfCGFnqnoo3hjTCA449LRCqFv7nM5CobZGnSDij'),
                        baseVault: new web3_js_1.PublicKey('EuoNht3Uc4hcBcu5Reep54HSXgvpXrgsVsNeEeenA5kf'),
                        quoteVault: new web3_js_1.PublicKey('FTwLw8s6kUjKwfworryNSsefCZUkAUnkRuBmDjZxgGwn'),
                        vaultOwner: new web3_js_1.PublicKey('GYUYJBrsCzJUMUbTTkiCs54d2m7Vh4eDMqyKtpbs52d1'),
                        bidsAddress: new web3_js_1.PublicKey('5ZboPX8DupNM8LFTXHZBGd3JV24xcJAfHfrjFntDvLL3'),
                        asksAddress: new web3_js_1.PublicKey('7whNVtz7Seqm46D8W4poYTsjPYsxp2jsPThVHvQ1UtgH'),
                    },
                    MSOL: {
                        marketAddress: new web3_js_1.PublicKey('HH6wc5BUZtDE3dpp6nzpHQzyryQyk9nzeCmGz9eAkpBH'),
                        requestQueue: new web3_js_1.PublicKey('6GF9C52akeqEvvycvRjamMKHwYHc1qnzLonKVRjb8rAx'),
                        eventQueue: new web3_js_1.PublicKey('5HqNWFTGqojWHKAPiDTSyBmcDK8qQeEixXtB22WNo1y'),
                        baseVault: new web3_js_1.PublicKey('GFvzeLV8aqLJPWjSVrx3mRnZjCiiWykmNPUA4Ef2NvVy'),
                        quoteVault: new web3_js_1.PublicKey('FYbLCiUgaTiBRKnLSAPHSaDDc5QXKPHxxmpZhwaeqqon'),
                        vaultOwner: new web3_js_1.PublicKey('FV5e18t2uZv8w2qbu73qQjAGcwoTRpkn7Nf6JySJJ5oZ'),
                        bidsAddress: new web3_js_1.PublicKey('GmBGVj1aEaMeCHq9LqDJpcDQn7nGcpXvB3LCjqsWxGt7'),
                        asksAddress: new web3_js_1.PublicKey('DB89DPTikq6KEsZLsiu7snzjzM9f2ohaAt5t9kJH1THF'),
                    },
                },
                saber: {
                    stablecoinSwap: new web3_js_1.PublicKey('MARpDPs5A7XiyCWPNH8GsMWPLxmwNn9SBmKvPa9LzgA'),
                },
                redemptionsQueue: undefined,
                redemptionCandidatesQueue: undefined,
            },
        },
        kaminoLending: [
            {
                lendingMarket: new web3_js_1.PublicKey('C3JSmPMmwHwJJuz9S4PBcsd1VVVLLvFFK2J36jqGRVDo'),
                lendingMarketAuthority: new web3_js_1.PublicKey('4fMAG8JYALfQE2tmwPcf59CbKVTvBSeveU3s8jTUqrg6'),
                isPrimary: true,
                reserves: [
                    {
                        liquidityToken: {
                            decimals: 9,
                            symbol: 'SOL',
                            mint: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                        },
                        address: new web3_js_1.PublicKey('6oHq4yaRWTtQ9QakJWvrryovdFMd2PY8jktmgSXj43x5'),
                        collateralMint: new web3_js_1.PublicKey('EhTKUYNtcSbPkt2MCmiYXryzxVgjfAroJFzL3XTrE58q'),
                        collateralSupply: new web3_js_1.PublicKey('26zUYzMq1eAB4mCiyrajnqc4AK2Dp1WfmNYLqyJpccXn'),
                        liquiditySupply: new web3_js_1.PublicKey('6Apg9g69rXZAmZKY65BiS8k7hJyhRYrHmiqe4q9BEkEs'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('9PtwCrYHEfRaCaPeZprim4HcgsjrMnDb8KJVwmNHLP2B'),
                        pythOracle: new web3_js_1.PublicKey('J83w4HKfqxwcq3BEMMkPFSppX3gqekLyLJBexebFVkix'),
                        switchboardOracle: new web3_js_1.PublicKey('GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'USDH',
                            mint: new web3_js_1.PublicKey('kfZGSThBdZbZY6cDSHP7NmCWa6D2hFc5kjBuGHREzZ3'),
                        },
                        address: new web3_js_1.PublicKey('7UvdRA9pkxL1f16APLEqvBTK7pa27BbLQuSE4u2GZcQ8'),
                        collateralMint: new web3_js_1.PublicKey('J73zc1xaahfaZZ9kFUfBUMjSBCnFprJCF8N9NY64wdEE'),
                        collateralSupply: new web3_js_1.PublicKey('4hPPiqZRZ8LnyV38jtEdCH7NACSQYHoa4gBThnDzneTk'),
                        liquiditySupply: new web3_js_1.PublicKey('77vPcgAEUyTTGJDgG6dbzzsXvGCrzUmdxnqzyKrvwm8g'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('7ck3jYXgytArTYc8XzxHB5EWJzMfkqEWupvmVrpzcKyk'),
                        pythOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                        switchboardOracle: new web3_js_1.PublicKey('6Hn9n8eKbbzxMCpkkZBYjS9kHktfMDucQ77jhHmjasuq'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'USDC',
                            mint: new web3_js_1.PublicKey('HEzfH7PCNSCU92zdk3iaKG4E9hQTErfgVGvqgywEvPzu'),
                        },
                        address: new web3_js_1.PublicKey('D81J5UpjPFd5k71KLFzVWVX1YVcS81tZnKrXJkDw5SGm'),
                        collateralMint: new web3_js_1.PublicKey('5BoshzFzGjiL1DRkDAV8JaHfbsgvpjJCSHCbgBr2t8sn'),
                        collateralSupply: new web3_js_1.PublicKey('6TF5DVm4ZDL1THaBug1T95f5yENu4myFgwFZq5iYtAVB'),
                        liquiditySupply: new web3_js_1.PublicKey('4ZyyAKpLBjbdjUDkjWL6dTJNgBP7CAb2JMi2bLC9ghUN'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('EvdusCsBnx6snTLeXhc5isAhQ6HuTdBhyGLHnRhGxHiJ'),
                        pythOracle: new web3_js_1.PublicKey('38xoQ4oeJCBrcVvca2cGk7iV1dAfrmTR1kmhSCJQ8Jto'),
                        switchboardOracle: new web3_js_1.PublicKey('BjUgj6YCnFBZ49wF54ddBVA9qu8TeqkFtkbqmZcee8uW'),
                    },
                    {
                        liquidityToken: {
                            decimals: 9,
                            symbol: 'MSOL',
                            mint: new web3_js_1.PublicKey('2wMQTELEgEcK7V3gwZBH58WvGcRZsyaWNpebqyFugGKe'),
                        },
                        address: new web3_js_1.PublicKey('ETi6oEW9ZXtZTjdSL47kWdD3WmfsqfT3DSMDkMhQ5mMh'),
                        collateralMint: new web3_js_1.PublicKey('2jvcBAbrAQaWb6jq6GJ2yTnpK5rcxCx81SessLqninWd'),
                        collateralSupply: new web3_js_1.PublicKey('3sPoXAXmzi7maN6p7e6m3RYBdw7axciUmhYBdr9bVtZr'),
                        liquiditySupply: new web3_js_1.PublicKey('FK4Ni5pUMwXDrcBAQgpk7GDLeTEA3NKtX5wjBv1sGXC4'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('5iaGemkKZ4wgTQb6KBfewE6qBd2va9CxiNkgMtpJKEZh'),
                        pythOracle: new web3_js_1.PublicKey('9a6RNx3tCu1TSs6TBSfV2XRXEPEZXQ6WB7jRojZRvyeZ'),
                        switchboardOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                    },
                    {
                        liquidityToken: {
                            decimals: 9,
                            symbol: 'STSOL',
                            mint: new web3_js_1.PublicKey('ESba8sT1R7WPiE3weBU1u6UqEmXRuM8Rs1LLvXayEizf'),
                        },
                        address: new web3_js_1.PublicKey('3pbzXZVF8SYChzvaERWCFdk4VMqabAQMFd7Qayw5UtKE'),
                        collateralMint: new web3_js_1.PublicKey('71NSRvWeUbZqzqyJM2HEDEPsD6E4kurFbF4pKtWDdmP'),
                        collateralSupply: new web3_js_1.PublicKey('DM8ZLxgeMnyT8H9bZVmG1KHHVAvaZKXzHcCGiYY8bMxz'),
                        liquiditySupply: new web3_js_1.PublicKey('DUXxfwZQhCVcUswQLLjqygVXziZAZoK2kvGKTEvHFwEo'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('4tADr67qkn8SVWM1zaTdSs3bva2gvhowqdXbY8gugUBY'),
                        pythOracle: new web3_js_1.PublicKey('2LwhbcswZekofMNRtDRMukZJNSRUiKYMFbqtBwqjDfke'),
                        switchboardOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                    },
                    {
                        liquidityToken: {
                            decimals: 8,
                            symbol: 'ETH',
                            mint: new web3_js_1.PublicKey('2DVyKwW17MxDXaLhWpWXtbVa64JBwaHmLJ1txXNJgRGi'),
                        },
                        address: new web3_js_1.PublicKey('GKqcS3WZ2uzfG3F8qC4RfSgToqdrzRyAgQXJPCNf6S7r'),
                        collateralMint: new web3_js_1.PublicKey('71NSRvWeUbZqzqyJM2HEDEPsD6E4kurFbF4pKtWDdmP'),
                        collateralSupply: new web3_js_1.PublicKey('DM8ZLxgeMnyT8H9bZVmG1KHHVAvaZKXzHcCGiYY8bMxz'),
                        liquiditySupply: new web3_js_1.PublicKey('DUXxfwZQhCVcUswQLLjqygVXziZAZoK2kvGKTEvHFwEo'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('4tADr67qkn8SVWM1zaTdSs3bva2gvhowqdXbY8gugUBY'),
                        pythOracle: new web3_js_1.PublicKey('EdVCmQ9FSPcVe5YySXDPCRmc8aDQLKJ9xvYBMZPie1Vw'),
                        switchboardOracle: new web3_js_1.PublicKey('HNStfhaLnqwF2ZtJUizaA9uHDAVB976r2AgTUx9LrdEo'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'USDT',
                            mint: new web3_js_1.PublicKey('AcU8CfJiLEBn1x7Cgx8atX6DdebmdQtNSsJBkQyf6Q4L'),
                        },
                        address: new web3_js_1.PublicKey('FHTZSFkp61hzaZZ5LR3rZPA7wVgcdyd39X889PFNagsy'),
                        collateralMint: new web3_js_1.PublicKey('4zsscY7tKobgCkqYt9etBCj5EfetfsT4Fw1htKnTaQTf'),
                        collateralSupply: new web3_js_1.PublicKey('DVs8NKS2kT9irngA69FwAnuCM8zjtgDQYi65X5F1PtCv'),
                        liquiditySupply: new web3_js_1.PublicKey('A2rU482UhbxzX8wuFs5M2EubKXDQEPZZC4tPZ5aeMG6k'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('CrrAU1NYpogx2Lh9smfwx1gVXHApFtm64Bf6MuRdPmMu'),
                        pythOracle: new web3_js_1.PublicKey('38xoQ4oeJCBrcVvca2cGk7iV1dAfrmTR1kmhSCJQ8Jto'),
                        switchboardOracle: new web3_js_1.PublicKey('ETAaeeuQBwsh9mC2gCov9WdhJENZuffRMXY2HgjCcSL9'),
                    },
                ],
            },
        ],
    },
    {
        cluster: 'localnet',
        kamino: {
            programId: new web3_js_1.PublicKey('E6qbhrt4pFmCotNUSSEh6E5cRQCEJpMcd79Z56EG9KY'),
            globalConfig: new web3_js_1.PublicKey('GKnHiWh3RRrE1zsNzWxRkomymHc374TvJPSTv2wPeYdB'),
            collateralInfos: new web3_js_1.PublicKey('3v6ootgJJZbSWEDfZMA1scfh7wcsVVfeocExRxPqCyWH'),
            mints: [
                { address: new web3_js_1.PublicKey('AcU8CfJiLEBn1x7Cgx8atX6DdebmdQtNSsJBkQyf6Q4L'), scopeToken: 'USDT' },
                { address: new web3_js_1.PublicKey('USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX'), scopeToken: 'USDH' },
                { address: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'), scopeToken: 'USDC' },
                { address: new web3_js_1.PublicKey('ESba8sT1R7WPiE3weBU1u6UqEmXRuM8Rs1LLvXayEizf'), scopeToken: 'STSOL' },
                { address: new web3_js_1.PublicKey('2wMQTELEgEcK7V3gwZBH58WvGcRZsyaWNpebqyFugGKe'), scopeToken: 'MSOL' },
                { address: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'), scopeToken: 'SOL' },
            ],
            whirlpoolsToSnapshot: [],
            raydiumPoolsToSnapshot: [],
            strategies: [
                new web3_js_1.PublicKey('AvRMXt1MM9bdyE4yrnbVYjzYWWx8sdhkXPFbmxeb3rq6'),
                new web3_js_1.PublicKey('yT2bBMrE4AVG3BM8LHjDSr1yketzabmfp3rYimS36C9'),
                new web3_js_1.PublicKey('2H4xebnp2M9JYgPPfUw58uUQahWF8f1YTNxwwtmdqVYV'),
                new web3_js_1.PublicKey('EbRR6C3uCRwnHjvbJuuiaipvVCVLYb2ZL5B3GH4zm3FH'),
            ],
        },
        scope: {
            oracleMappings: new web3_js_1.PublicKey('Chpu5ZgfWX5ZzVpUx9Xvv4WPM75Xd7zPJNDPsFnCpLpk'),
            oraclePrices: new web3_js_1.PublicKey('3NJYftD5sjVfxSnUdZ1wVML8f3aC6mp1CXCL6L7TnU8C'),
            programId: new web3_js_1.PublicKey('HFn8GnPADiny6XqUoWE8uRPPxb29ikn4yTuPa9MF2fWJ'),
            configurationAccount: new web3_js_1.PublicKey('AdTiP7QyjUyv6crF4H8z7fxJKU7Z5eCAGvJN1Y55cXxb'),
        },
        borrowing: {
            programId: new web3_js_1.PublicKey('FAZuPmWCvPzVoSXudtMvixiGNGKyEuKvfEuJh3i6NkXu'),
            accounts: {
                stablecoinMint: new web3_js_1.PublicKey('CkUrb2SwrriK4T7VpWjrepM4PWnfT9hwUEfz2NJmM4Hp'),
                stablecoinMintAuthority: new web3_js_1.PublicKey('89Ecdd8J7xTycvueCuK4ML3xUheZEp917qmzKsHfCgdV'),
                hbbMintAuthority: new web3_js_1.PublicKey('5BKFyoeZB1ZArEzq3KLxF8RhUPzBLas8kgEH5sjn16Jb'),
                borrowingVaultsV2: new web3_js_1.PublicKey('BHgw9kZcK6LVSadiqUZYBxe8rCzTcDq1dxsJcspLmWPX'),
                borrowingMarketStates: [
                    new web3_js_1.PublicKey('YSp9bJpQom2HA7VThxYZX9pNwrcBH88NBcBNWs5yskR'),
                    new web3_js_1.PublicKey('B6gckY3KmWR2u1gwruE55Hxoj5uqiTTtuXLbE6Bu7bZq'),
                ],
                mint: {
                    WSOL: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                    ETH: new web3_js_1.PublicKey('B54v6dyccBv66VBTrQHqebKpBR1UNMT6UHMWZcVKJhct'),
                    BTC: new web3_js_1.PublicKey('6Rkde5L39BoXZ9b8cGXnbTwNjLa6SmmEuZQ7vdCVbiZ7'),
                    SRM: new web3_js_1.PublicKey('ai6TJHW8RgnXdfpCgFBx7d2U53ozFeVR86cvfHK2T1w'),
                    RAY: new web3_js_1.PublicKey('6e6aWyz9viDz2yoJqSiyZVwCKp5qZx8UUzesgs3y3Emd'),
                    FTT: new web3_js_1.PublicKey('6AJ9LYjCpLFekhX7Jzq6YDQrZ72oLTeQuLgGCxHw8nq2'),
                    HBB: new web3_js_1.PublicKey('HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6'),
                    MSOL: undefined,
                    LDO: new web3_js_1.PublicKey('HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p'),
                    ORCA: new web3_js_1.PublicKey('orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE'),
                },
                collateralVaultsAuthority: new web3_js_1.PublicKey('AzrwP37zYTDh1dfbWWc3ceK2LVbijNuDvV9hS1qH96dM'),
                collateralVault: {
                    SOL: new web3_js_1.PublicKey('8CR87Meo2XbfugckHmwFbWfRnPrzqyj5dReyKdU9n4PA'),
                    ETH: new web3_js_1.PublicKey('7j92j4m7YWxWGuvjS3HNcHYRdkUp3NpkiRGTS1wdBhWm'),
                    BTC: new web3_js_1.PublicKey('CuAMwJkgvWMrkUfEoxQJWFRw8P8GtVta4HVx7DWTXDwX'),
                    SRM: new web3_js_1.PublicKey('3Z2wmrPiUHF47DrSFRZnmm4CJc9aTs3bize4QtfnGtQE'),
                    RAY: new web3_js_1.PublicKey('B1Njx94rsEAam4Am19d347t76DbxxR8pUmWLBo7MBEE'),
                    FTT: new web3_js_1.PublicKey('3xFadKYXmPUStW5UybXcmEkWrb3dBcAkhvopipKDyXd5'),
                    MSOL: undefined,
                },
                redemptionsQueue: new web3_js_1.PublicKey('DZ9nuLYkKdGmXTFUMz7FZX4EDeqkQmDL71qcivsCtHFo'),
                liquidationsQueue: new web3_js_1.PublicKey('39NPoTCGhZ3WkZcevFX7xYyKZzYVHtvMmjsKEQpe9Ltp'),
                borrowingMarketState: new web3_js_1.PublicKey('FiULFVsJGoi6BVQV4MZ4nRhSt5q7PMCxXhPNHedtcJK8'),
                stabilityPoolState: new web3_js_1.PublicKey('3TahgKbYxpGp4iqtkWKy1mw8uo4FGhMzs42tgqTqcqyD'),
                epochToScaleToSum: new web3_js_1.PublicKey('G9Ysu9DgFiUuWWWnYsRx8SvN6sfrtDY7NCaHVxgz7jeb'),
                stablecoinStabilityPoolVault: new web3_js_1.PublicKey('21dTv2L7P52goCy4YFsM6k737yZJh3fxdL3kKtqcWCzR'),
                stabilityVaults: new web3_js_1.PublicKey('Az4MhfSV7f7pgAnDr7UyvsiemsmrpGM39AWfkgicLawa'),
                stablecoinStabilityPoolVaultAuthority: new web3_js_1.PublicKey('Ecg7PgFoEjX5tJw2GZN4tUKz7sh3PnSsJECcsvVkRDpq'),
                liquidationRewardsVaultSol: new web3_js_1.PublicKey('AUznMW6M1fsGJusotrWBVHBQgNRvvucerJHVtb969Ls'),
                liquidationRewardsVaultSrm: new web3_js_1.PublicKey('7hMX18PtEqW6z29BBCU1pWFJKoYbR3sZVZW31fD64n4y'),
                liquidationRewardsVaultEth: new web3_js_1.PublicKey('9thYqRns8M6xXhcPTs6Ra42jt153QAjymTnU5MtHFR2F'),
                liquidationRewardsVaultBtc: new web3_js_1.PublicKey('BxDHBQW38Wo1FgCJDbrCyYsHqpfw1gZmWCFNJbyDuf49'),
                liquidationRewardsVaultRay: new web3_js_1.PublicKey('DuCiQJsXzXF9QKufsP8YXKqpbttaJxvx4E2LaSY4P7zz'),
                liquidationRewardsVaultFtt: new web3_js_1.PublicKey('97Jq9TdBbKN7yxAnQMGctm7EDik7HYFTtfr4JVnSyEtS'),
                liquidationRewardsVaultAuthority: new web3_js_1.PublicKey('FQbUdSuYtMxn9HoA649gpYVV6vh8xAiEBTHsVHm6pdAb'),
                borrowingVaults: new web3_js_1.PublicKey('4fN86magQ4ghRfP9qfQbMnSogtJ1r6T9R7Yt7LzQSHA4'),
                globalConfig: new web3_js_1.PublicKey('GnJoxBn17Xn5WpmUMZ2mU17TyBzLGkSxLCsczSEiHvH'),
                oracleMappings: new web3_js_1.PublicKey('4oe3cdNdwRxv2qjXA2q4yD9aj9XdE7Yd7nBnUeunt8m7'),
                stakingPoolState: new web3_js_1.PublicKey('8nAvRvVHAydWL743GgXMbTfXwwrPD1RDbQ69XAGG1FZy'),
                borrowingFeesAccount: new web3_js_1.PublicKey('2pLG1oGQwFDpbCt8m8RHTWUTSKFKA56HffwESFEeFFpc'),
                borrowingFeesVaultAuthority: new web3_js_1.PublicKey('9RXCJ66mJDcDWtNpUJknrkeda97yWLWvRnQSQjzkH2iC'),
                stakingVault: new web3_js_1.PublicKey('D6YWbkypjqoNTUxfYZSh7ra5mD4AapHRSHRMJw8eLbn8'),
                treasuryVault: new web3_js_1.PublicKey('4UAr1CWXFensisCGBWnSdU4dTAsa5AWtzDpYDSfbQJmc'),
                burningVault: new web3_js_1.PublicKey('HKpANDh2zs5vChFZKse7nny9Xt5nUxqoXQahc3QrXCZ4'),
                burningVaultAuthority: new web3_js_1.PublicKey('AbUWAzTWd1fkbV3FDzU69HJf2ntvyzUEnN5FjP5HVHhc'),
                pyth: {
                    solProductInfo: new web3_js_1.PublicKey('Bxd5yBjGL29twaLP2B1Lf6YjFFXRcUmqTz7v32BeHsTN'),
                    solPriceInfo: new web3_js_1.PublicKey('FweBx8gsTNGf1YyebZK7jPqn9SYiyptT2UUYV5pK3Scn'),
                    ethProductInfo: new web3_js_1.PublicKey('8TtUqogFv4srq3S56hVjYeyeY1Xxw4Nt9HqjDY2Pfg2Y'),
                    ethPriceInfo: new web3_js_1.PublicKey('CUug7XgabHU4X7ca7ZCesCJWnrHbK1pMCAaSxQ9RZBYo'),
                    btcProductInfo: new web3_js_1.PublicKey('7TaVKytXHEdEnG8dT6d3rf39jzY3mTrq23fYE66h1L2o'),
                    btcPriceInfo: new web3_js_1.PublicKey('ADfQ2g1DkJuTadSNW5Ssq71qd2mvmtsqv2Dn87xZx5Sb'),
                    srmProductInfo: new web3_js_1.PublicKey('Dyswi54xNFLxuTWdzah7d3HsN5dM9WQ2UTXFLfwMoZDa'),
                    srmPriceInfo: new web3_js_1.PublicKey('Covgt34sxCzsz9FS6wXss2zFmzRnsUfRbsGDuSvnNgSb'),
                    rayProductInfo: new web3_js_1.PublicKey('2gX5FKPh3kubLzHu4Pm7YMJFoWB7rx4qAdDBz5fqp1xr'),
                    rayPriceInfo: new web3_js_1.PublicKey('814AnnK8hLC3iriRGjakb1vo2tBgJknmsuG18bxauZr7'),
                    fttProductInfo: new web3_js_1.PublicKey('6U3UDYYdi7ryqgA7fs7xqK5C9coQyx1mxxvu7XzuLF95'),
                    fttPriceInfo: new web3_js_1.PublicKey('G7kqVzuayYqaCtGSZx1dJapdFWkRc18PoGaTg6Gv7z9k'),
                    msolProductInfo: undefined,
                    msolPriceInfo: undefined,
                },
                liquidationRewardsVaultMsol: undefined,
                USDC: undefined,
                serumMarkets: undefined,
                redemptionCandidatesQueue: undefined,
                saber: {
                    stablecoinSwap: new web3_js_1.PublicKey('MARpDPs5A7XiyCWPNH8GsMWPLxmwNn9SBmKvPa9LzgA'),
                },
            },
        },
        kaminoLending: [
            {
                lendingMarket: new web3_js_1.PublicKey('HD6SgAR8hTjZRm54g7XLQy7auVfyFq89LX347vj1RTWv'),
                lendingMarketAuthority: new web3_js_1.PublicKey('DFKUFHiDvHspRkvTfdGv2hCvGixKt9sN3vjHSGZxnNZX'),
                isPrimary: true,
                reserves: [
                    {
                        liquidityToken: {
                            decimals: 9,
                            symbol: 'SOL',
                            mint: new web3_js_1.PublicKey('So11111111111111111111111111111111111111112'),
                        },
                        address: new web3_js_1.PublicKey('CuYZQKS1gWEB3K6NRQHYA8ftTXERQWGyV4c3zfT6TRvd'),
                        collateralMint: new web3_js_1.PublicKey('C658y9dHBiadsycj2mT6knGrya9JjceJfgwEG5qyZWKu'),
                        collateralSupply: new web3_js_1.PublicKey('8fNz4sAduikZYyrf19Y75zbPQxeea4yRPzQLRr3WoSdv'),
                        liquiditySupply: new web3_js_1.PublicKey('4EWNVRjAyBV65LnUuuNxY9CHs2cJ3fYWZnw3yEFTNMC6'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('HEELm1ZsQAaypfS1fQ6UXFFznX9VQ6S3KrYf6uBYkauv'),
                        pythOracle: new web3_js_1.PublicKey('J83w4HKfqxwcq3BEMMkPFSppX3gqekLyLJBexebFVkix'),
                        switchboardOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'USDH',
                            mint: new web3_js_1.PublicKey('HWzkNyXynnC3mAZAL4yKZNvWs3SwS9eBVKdzmKcxYHSH'),
                        },
                        address: new web3_js_1.PublicKey('41AnqvAYzvaSHYv1PCqmfp7f2wqRXKJPF4hHgLicHrDQ'),
                        collateralMint: new web3_js_1.PublicKey('3hrRVdDiqhdspB8KYHRnCMo97HabVHKstg25c1njyWKk'),
                        collateralSupply: new web3_js_1.PublicKey('8eHxnQ5mj9wfn4dzLLmaz5c829WHprFhAMse3Jrrjo1R'),
                        liquiditySupply: new web3_js_1.PublicKey('A9fNepK75rEVPoPk3b4nRpvrbRaVHbCth4D1rEcfZWZc'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('8hyZAriFUqt53Qaz8qKFzjHxMcu2wvQSuXR8aN6Zpuc'),
                        pythOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                        switchboardOracle: new web3_js_1.PublicKey('DahMqw2Gue2EnnKkeyHmqVMbF1vz31CoCsabnBTo3fif'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'USDC',
                            mint: new web3_js_1.PublicKey('8KinRubdHLTivgyS6kv5Rph7kkMv5sdCroTSXNeq1sWq'),
                        },
                        address: new web3_js_1.PublicKey('GqynGfQuVb9x9ji2aGngqnbXoeydS9i3JSrdxzAkLUr4'),
                        collateralMint: new web3_js_1.PublicKey('T24csPyVhNpVYPjQhVhVp5eP6VvAFzLg9LzGAvSiCJu'),
                        collateralSupply: new web3_js_1.PublicKey('RAALczeSmQddb5yTjguXiApUqNCQoDxYQHp6FT5LxeB'),
                        liquiditySupply: new web3_js_1.PublicKey('GaLu9KD3boE2Rzg1oRfrpHnWf7uVjj1yHiaV37xqVWfg'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('Dq6Rgr261Zs3Yj9ewNMq7nJq4BnZLPrevtsVFGdxb5Rk'),
                        pythOracle: new web3_js_1.PublicKey('38xoQ4oeJCBrcVvca2cGk7iV1dAfrmTR1kmhSCJQ8Jto'),
                        switchboardOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                    },
                    {
                        liquidityToken: {
                            decimals: 9,
                            symbol: 'STSOL',
                            mint: new web3_js_1.PublicKey('G89c7G8YkQQGP3hyPHM2Jn3x2xn5zKZRPspCwVMfzsMT'),
                        },
                        address: new web3_js_1.PublicKey('49bVSGTN7tcYMqgNNRFGfcg8wNcF8ssQmrwouS6wsxvc'),
                        collateralMint: new web3_js_1.PublicKey('44qBTX95d1pGGkqndiUUGE3ZTNCKsEA5PtPo3ztNhnvN'),
                        collateralSupply: new web3_js_1.PublicKey('DJ1oMSZRoJzT5g4TNu7SM8YwS96E5h2NFMVXRBxo5yns'),
                        liquiditySupply: new web3_js_1.PublicKey('6KutGfsMK7w6BRJKWBkebpyuN7dEASn85oui79eJY4NY'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('HRmB1VE6KHw1EvtXY7N3yzfXm3GZt489ZdpP5DKmbq4i'),
                        pythOracle: new web3_js_1.PublicKey('2LwhbcswZekofMNRtDRMukZJNSRUiKYMFbqtBwqjDfke'),
                        switchboardOracle: new web3_js_1.PublicKey('9LNYQZLJG5DAyeACCTzBFG6H3sDhehP5xtYLdhrZtQkA'),
                    },
                    {
                        liquidityToken: {
                            decimals: 6,
                            symbol: 'kUSDHUSDCOrca',
                            mint: new web3_js_1.PublicKey('E8KG1XTQE3bf6ZTDG7xa38PAbAxHaDumsAgmGraY58MQ'),
                        },
                        address: new web3_js_1.PublicKey('HW65CmnrphRxMDg2YRsQJwdMc8tSzzMpdDawmdvkwGMt'),
                        collateralMint: new web3_js_1.PublicKey('9hEo8mpuKf2UNjcWvPyTWiS67SbFigNdiJZnQL5VnCFq'),
                        collateralSupply: new web3_js_1.PublicKey('146ssPwPSmMwHSekGo1NefAcDnCMuLbuANDbv1o49ii3'),
                        liquiditySupply: new web3_js_1.PublicKey('55jcYQPxC9Umd8neAvLPyuhQj5kZ3D9GjcYhNju6J1z7'),
                        liquidityFeeReceiverAddress: new web3_js_1.PublicKey('rYi1vnpsgAjJsxvjtzLpUW7xCqgUjsKYj56piibRLo3'),
                        pythOracle: new web3_js_1.PublicKey('nu11111111111111111111111111111111111111111'),
                        switchboardOracle: new web3_js_1.PublicKey('CTLRyAao68yqeid6i2xbiY8w3XXLrNqe23eWH1iUWmxH'),
                    },
                ],
            },
        ],
    },
];
