// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/accounts/AmmConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmmConfig = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
/* eslint-enable @typescript-eslint/no-unused-vars */
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const utils_1 = require("../utils"); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class AmmConfig {
    bump;
    index;
    owner;
    protocolFeeRate;
    tradeFeeRate;
    tickSpacing;
    fundFeeRate;
    paddingU32;
    fundOwner;
    padding;
    static discriminator = Buffer.from([
        218, 244, 33, 104, 203, 203, 43, 111,
    ]);
    static layout = borsh.struct([
        borsh.u8("bump"),
        borsh.u16("index"),
        (0, utils_1.borshAddress)("owner"),
        borsh.u32("protocolFeeRate"),
        borsh.u32("tradeFeeRate"),
        borsh.u16("tickSpacing"),
        borsh.u32("fundFeeRate"),
        borsh.u32("paddingU32"),
        (0, utils_1.borshAddress)("fundOwner"),
        borsh.array(borsh.u64(), 3, "padding"),
    ]);
    constructor(fields) {
        this.bump = fields.bump;
        this.index = fields.index;
        this.owner = fields.owner;
        this.protocolFeeRate = fields.protocolFeeRate;
        this.tradeFeeRate = fields.tradeFeeRate;
        this.tickSpacing = fields.tickSpacing;
        this.fundFeeRate = fields.fundFeeRate;
        this.paddingU32 = fields.paddingU32;
        this.fundOwner = fields.fundOwner;
        this.padding = fields.padding;
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error(`AmmConfigFields account ${address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error(`AmmConfigFields account ${info.address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(AmmConfig.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = AmmConfig.layout.decode(data.slice(8));
        return new AmmConfig({
            bump: dec.bump,
            index: dec.index,
            owner: dec.owner,
            protocolFeeRate: dec.protocolFeeRate,
            tradeFeeRate: dec.tradeFeeRate,
            tickSpacing: dec.tickSpacing,
            fundFeeRate: dec.fundFeeRate,
            paddingU32: dec.paddingU32,
            fundOwner: dec.fundOwner,
            padding: dec.padding,
        });
    }
    toJSON() {
        return {
            bump: this.bump,
            index: this.index,
            owner: this.owner,
            protocolFeeRate: this.protocolFeeRate,
            tradeFeeRate: this.tradeFeeRate,
            tickSpacing: this.tickSpacing,
            fundFeeRate: this.fundFeeRate,
            paddingU32: this.paddingU32,
            fundOwner: this.fundOwner,
            padding: this.padding.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new AmmConfig({
            bump: obj.bump,
            index: obj.index,
            owner: (0, kit_1.address)(obj.owner),
            protocolFeeRate: obj.protocolFeeRate,
            tradeFeeRate: obj.tradeFeeRate,
            tickSpacing: obj.tickSpacing,
            fundFeeRate: obj.fundFeeRate,
            paddingU32: obj.paddingU32,
            fundOwner: (0, kit_1.address)(obj.fundOwner),
            padding: obj.padding.map((item) => new bn_js_1.default(item)),
        });
    }
}
exports.AmmConfig = AmmConfig;
