// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/accounts/OperationState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationState = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const utils_1 = require("../utils"); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class OperationState {
    bump;
    operationOwners;
    whitelistMints;
    static discriminator = Buffer.from([
        19, 236, 58, 237, 81, 222, 183, 252,
    ]);
    static layout = borsh.struct([
        borsh.u8("bump"),
        borsh.array((0, utils_1.borshAddress)(), 10, "operationOwners"),
        borsh.array((0, utils_1.borshAddress)(), 100, "whitelistMints"),
    ]);
    constructor(fields) {
        this.bump = fields.bump;
        this.operationOwners = fields.operationOwners;
        this.whitelistMints = fields.whitelistMints;
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error(`OperationStateFields account ${address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error(`OperationStateFields account ${info.address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(OperationState.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = OperationState.layout.decode(data.slice(8));
        return new OperationState({
            bump: dec.bump,
            operationOwners: dec.operationOwners,
            whitelistMints: dec.whitelistMints,
        });
    }
    toJSON() {
        return {
            bump: this.bump,
            operationOwners: this.operationOwners,
            whitelistMints: this.whitelistMints,
        };
    }
    static fromJSON(obj) {
        return new OperationState({
            bump: obj.bump,
            operationOwners: obj.operationOwners.map((item) => (0, kit_1.address)(item)),
            whitelistMints: obj.whitelistMints.map((item) => (0, kit_1.address)(item)),
        });
    }
}
exports.OperationState = OperationState;
