// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/accounts/TickArrayBitmapExtension.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TickArrayBitmapExtension = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
/* eslint-enable @typescript-eslint/no-unused-vars */
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const utils_1 = require("../utils"); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class TickArrayBitmapExtension {
    poolId;
    positiveTickArrayBitmap;
    negativeTickArrayBitmap;
    static discriminator = Buffer.from([
        60, 150, 36, 219, 97, 128, 139, 153,
    ]);
    static layout = borsh.struct([
        (0, utils_1.borshAddress)("poolId"),
        borsh.array(borsh.array(borsh.u64(), 8), 14, "positiveTickArrayBitmap"),
        borsh.array(borsh.array(borsh.u64(), 8), 14, "negativeTickArrayBitmap"),
    ]);
    constructor(fields) {
        this.poolId = fields.poolId;
        this.positiveTickArrayBitmap = fields.positiveTickArrayBitmap;
        this.negativeTickArrayBitmap = fields.negativeTickArrayBitmap;
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error(`TickArrayBitmapExtensionFields account ${address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error(`TickArrayBitmapExtensionFields account ${info.address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(TickArrayBitmapExtension.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = TickArrayBitmapExtension.layout.decode(data.slice(8));
        return new TickArrayBitmapExtension({
            poolId: dec.poolId,
            positiveTickArrayBitmap: dec.positiveTickArrayBitmap,
            negativeTickArrayBitmap: dec.negativeTickArrayBitmap,
        });
    }
    toJSON() {
        return {
            poolId: this.poolId,
            positiveTickArrayBitmap: this.positiveTickArrayBitmap.map((item) => item.map((item) => item.toString())),
            negativeTickArrayBitmap: this.negativeTickArrayBitmap.map((item) => item.map((item) => item.toString())),
        };
    }
    static fromJSON(obj) {
        return new TickArrayBitmapExtension({
            poolId: (0, kit_1.address)(obj.poolId),
            positiveTickArrayBitmap: obj.positiveTickArrayBitmap.map((item) => item.map((item) => new bn_js_1.default(item))),
            negativeTickArrayBitmap: obj.negativeTickArrayBitmap.map((item) => item.map((item) => new bn_js_1.default(item))),
        });
    }
}
exports.TickArrayBitmapExtension = TickArrayBitmapExtension;
