// @kamino-finance/farms-sdk/dist/@codegen/farms/accounts/GlobalConfig.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalConfig = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
/* eslint-enable @typescript-eslint/no-unused-vars */
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const utils_1 = require("../utils"); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class GlobalConfig {
    globalAdmin;
    treasuryFeeBps;
    treasuryVaultsAuthority;
    treasuryVaultsAuthorityBump;
    pendingGlobalAdmin;
    padding1;
    static discriminator = Buffer.from([
        149, 8, 156, 202, 160, 252, 176, 217,
    ]);
    static layout = borsh.struct([
        (0, utils_1.borshAddress)("globalAdmin"),
        borsh.u64("treasuryFeeBps"),
        (0, utils_1.borshAddress)("treasuryVaultsAuthority"),
        borsh.u64("treasuryVaultsAuthorityBump"),
        (0, utils_1.borshAddress)("pendingGlobalAdmin"),
        borsh.array(borsh.u128(), 126, "padding1"),
    ]);
    constructor(fields) {
        this.globalAdmin = fields.globalAdmin;
        this.treasuryFeeBps = fields.treasuryFeeBps;
        this.treasuryVaultsAuthority = fields.treasuryVaultsAuthority;
        this.treasuryVaultsAuthorityBump = fields.treasuryVaultsAuthorityBump;
        this.pendingGlobalAdmin = fields.pendingGlobalAdmin;
        this.padding1 = fields.padding1;
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error(`GlobalConfigFields account ${address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error(`GlobalConfigFields account ${info.address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(GlobalConfig.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = GlobalConfig.layout.decode(data.slice(8));
        return new GlobalConfig({
            globalAdmin: dec.globalAdmin,
            treasuryFeeBps: dec.treasuryFeeBps,
            treasuryVaultsAuthority: dec.treasuryVaultsAuthority,
            treasuryVaultsAuthorityBump: dec.treasuryVaultsAuthorityBump,
            pendingGlobalAdmin: dec.pendingGlobalAdmin,
            padding1: dec.padding1,
        });
    }
    toJSON() {
        return {
            globalAdmin: this.globalAdmin,
            treasuryFeeBps: this.treasuryFeeBps.toString(),
            treasuryVaultsAuthority: this.treasuryVaultsAuthority,
            treasuryVaultsAuthorityBump: this.treasuryVaultsAuthorityBump.toString(),
            pendingGlobalAdmin: this.pendingGlobalAdmin,
            padding1: this.padding1.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new GlobalConfig({
            globalAdmin: (0, kit_1.address)(obj.globalAdmin),
            treasuryFeeBps: new bn_js_1.default(obj.treasuryFeeBps),
            treasuryVaultsAuthority: (0, kit_1.address)(obj.treasuryVaultsAuthority),
            treasuryVaultsAuthorityBump: new bn_js_1.default(obj.treasuryVaultsAuthorityBump),
            pendingGlobalAdmin: (0, kit_1.address)(obj.pendingGlobalAdmin),
            padding1: obj.padding1.map((item) => new bn_js_1.default(item)),
        });
    }
}
exports.GlobalConfig = GlobalConfig;
