// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardState = exports.PoolStatusBitFlag = exports.PoolStatusBitIndex = exports.TickState = exports.RewardInfo = exports.PositionRewardInfo = exports.Observation = exports.InitializeRewardParam = void 0;
const PoolStatusBitIndex = __importStar(require("./PoolStatusBitIndex"));
exports.PoolStatusBitIndex = PoolStatusBitIndex;
const PoolStatusBitFlag = __importStar(require("./PoolStatusBitFlag"));
exports.PoolStatusBitFlag = PoolStatusBitFlag;
const RewardState = __importStar(require("./RewardState"));
exports.RewardState = RewardState;
var InitializeRewardParam_1 = require("./InitializeRewardParam");
Object.defineProperty(exports, "InitializeRewardParam", { enumerable: true, get: function () { return InitializeRewardParam_1.InitializeRewardParam; } });
var Observation_1 = require("./Observation");
Object.defineProperty(exports, "Observation", { enumerable: true, get: function () { return Observation_1.Observation; } });
var PositionRewardInfo_1 = require("./PositionRewardInfo");
Object.defineProperty(exports, "PositionRewardInfo", { enumerable: true, get: function () { return PositionRewardInfo_1.PositionRewardInfo; } });
var RewardInfo_1 = require("./RewardInfo");
Object.defineProperty(exports, "RewardInfo", { enumerable: true, get: function () { return RewardInfo_1.RewardInfo; } });
var TickState_1 = require("./TickState");
Object.defineProperty(exports, "TickState", { enumerable: true, get: function () { return TickState_1.TickState; } });
