// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/types/InitializeRewardParam.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitializeRewardParam = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class InitializeRewardParam {
    openTime;
    endTime;
    emissionsPerSecondX64;
    constructor(fields) {
        this.openTime = fields.openTime;
        this.endTime = fields.endTime;
        this.emissionsPerSecondX64 = fields.emissionsPerSecondX64;
    }
    static layout(property) {
        return borsh.struct([
            borsh.u64("openTime"),
            borsh.u64("endTime"),
            borsh.u128("emissionsPerSecondX64"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new InitializeRewardParam({
            openTime: obj.openTime,
            endTime: obj.endTime,
            emissionsPerSecondX64: obj.emissionsPerSecondX64,
        });
    }
    static toEncodable(fields) {
        return {
            openTime: fields.openTime,
            endTime: fields.endTime,
            emissionsPerSecondX64: fields.emissionsPerSecondX64,
        };
    }
    toJSON() {
        return {
            openTime: this.openTime.toString(),
            endTime: this.endTime.toString(),
            emissionsPerSecondX64: this.emissionsPerSecondX64.toString(),
        };
    }
    static fromJSON(obj) {
        return new InitializeRewardParam({
            openTime: new bn_js_1.default(obj.openTime),
            endTime: new bn_js_1.default(obj.endTime),
            emissionsPerSecondX64: new bn_js_1.default(obj.emissionsPerSecondX64),
        });
    }
    toEncodable() {
        return InitializeRewardParam.toEncodable(this);
    }
}
exports.InitializeRewardParam = InitializeRewardParam;
