// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/types/PoolStatusBitIndex.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Swap = exports.CollectReward = exports.CollectFee = exports.DecreaseLiquidity = exports.OpenPositionOrIncreaseLiquidity = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class OpenPositionOrIncreaseLiquidity {
    static discriminator = 0;
    static kind = "OpenPositionOrIncreaseLiquidity";
    discriminator = 0;
    kind = "OpenPositionOrIncreaseLiquidity";
    toJSON() {
        return {
            kind: "OpenPositionOrIncreaseLiquidity",
        };
    }
    toEncodable() {
        return {
            OpenPositionOrIncreaseLiquidity: {},
        };
    }
}
exports.OpenPositionOrIncreaseLiquidity = OpenPositionOrIncreaseLiquidity;
class DecreaseLiquidity {
    static discriminator = 1;
    static kind = "DecreaseLiquidity";
    discriminator = 1;
    kind = "DecreaseLiquidity";
    toJSON() {
        return {
            kind: "DecreaseLiquidity",
        };
    }
    toEncodable() {
        return {
            DecreaseLiquidity: {},
        };
    }
}
exports.DecreaseLiquidity = DecreaseLiquidity;
class CollectFee {
    static discriminator = 2;
    static kind = "CollectFee";
    discriminator = 2;
    kind = "CollectFee";
    toJSON() {
        return {
            kind: "CollectFee",
        };
    }
    toEncodable() {
        return {
            CollectFee: {},
        };
    }
}
exports.CollectFee = CollectFee;
class CollectReward {
    static discriminator = 3;
    static kind = "CollectReward";
    discriminator = 3;
    kind = "CollectReward";
    toJSON() {
        return {
            kind: "CollectReward",
        };
    }
    toEncodable() {
        return {
            CollectReward: {},
        };
    }
}
exports.CollectReward = CollectReward;
class Swap {
    static discriminator = 4;
    static kind = "Swap";
    discriminator = 4;
    kind = "Swap";
    toJSON() {
        return {
            kind: "Swap",
        };
    }
    toEncodable() {
        return {
            Swap: {},
        };
    }
}
exports.Swap = Swap;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("OpenPositionOrIncreaseLiquidity" in obj) {
        return new OpenPositionOrIncreaseLiquidity();
    }
    if ("DecreaseLiquidity" in obj) {
        return new DecreaseLiquidity();
    }
    if ("CollectFee" in obj) {
        return new CollectFee();
    }
    if ("CollectReward" in obj) {
        return new CollectReward();
    }
    if ("Swap" in obj) {
        return new Swap();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "OpenPositionOrIncreaseLiquidity": {
            return new OpenPositionOrIncreaseLiquidity();
        }
        case "DecreaseLiquidity": {
            return new DecreaseLiquidity();
        }
        case "CollectFee": {
            return new CollectFee();
        }
        case "CollectReward": {
            return new CollectReward();
        }
        case "Swap": {
            return new Swap();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "OpenPositionOrIncreaseLiquidity"),
        borsh.struct([], "DecreaseLiquidity"),
        borsh.struct([], "CollectFee"),
        borsh.struct([], "CollectReward"),
        borsh.struct([], "Swap"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
