// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/types/RewardState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ended = exports.Opening = exports.Initialized = exports.Uninitialized = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Uninitialized {
    static discriminator = 0;
    static kind = "Uninitialized";
    discriminator = 0;
    kind = "Uninitialized";
    toJSON() {
        return {
            kind: "Uninitialized",
        };
    }
    toEncodable() {
        return {
            Uninitialized: {},
        };
    }
}
exports.Uninitialized = Uninitialized;
class Initialized {
    static discriminator = 1;
    static kind = "Initialized";
    discriminator = 1;
    kind = "Initialized";
    toJSON() {
        return {
            kind: "Initialized",
        };
    }
    toEncodable() {
        return {
            Initialized: {},
        };
    }
}
exports.Initialized = Initialized;
class Opening {
    static discriminator = 2;
    static kind = "Opening";
    discriminator = 2;
    kind = "Opening";
    toJSON() {
        return {
            kind: "Opening",
        };
    }
    toEncodable() {
        return {
            Opening: {},
        };
    }
}
exports.Opening = Opening;
class Ended {
    static discriminator = 3;
    static kind = "Ended";
    discriminator = 3;
    kind = "Ended";
    toJSON() {
        return {
            kind: "Ended",
        };
    }
    toEncodable() {
        return {
            Ended: {},
        };
    }
}
exports.Ended = Ended;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Uninitialized" in obj) {
        return new Uninitialized();
    }
    if ("Initialized" in obj) {
        return new Initialized();
    }
    if ("Opening" in obj) {
        return new Opening();
    }
    if ("Ended" in obj) {
        return new Ended();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Uninitialized": {
            return new Uninitialized();
        }
        case "Initialized": {
            return new Initialized();
        }
        case "Opening": {
            return new Opening();
        }
        case "Ended": {
            return new Ended();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Uninitialized"),
        borsh.struct([], "Initialized"),
        borsh.struct([], "Opening"),
        borsh.struct([], "Ended"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
