// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/types/RewardInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardInfo = void 0;
const kit_1 = require("@solana/kit"); // eslint-disable-line @typescript-eslint/no-unused-vars
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
const utils_1 = require("../utils");
class RewardInfo {
    rewardState;
    openTime;
    endTime;
    lastUpdateTime;
    emissionsPerSecondX64;
    rewardTotalEmissioned;
    rewardClaimed;
    tokenMint;
    tokenVault;
    authority;
    rewardGrowthGlobalX64;
    constructor(fields) {
        this.rewardState = fields.rewardState;
        this.openTime = fields.openTime;
        this.endTime = fields.endTime;
        this.lastUpdateTime = fields.lastUpdateTime;
        this.emissionsPerSecondX64 = fields.emissionsPerSecondX64;
        this.rewardTotalEmissioned = fields.rewardTotalEmissioned;
        this.rewardClaimed = fields.rewardClaimed;
        this.tokenMint = fields.tokenMint;
        this.tokenVault = fields.tokenVault;
        this.authority = fields.authority;
        this.rewardGrowthGlobalX64 = fields.rewardGrowthGlobalX64;
    }
    static layout(property) {
        return borsh.struct([
            borsh.u8("rewardState"),
            borsh.u64("openTime"),
            borsh.u64("endTime"),
            borsh.u64("lastUpdateTime"),
            borsh.u128("emissionsPerSecondX64"),
            borsh.u64("rewardTotalEmissioned"),
            borsh.u64("rewardClaimed"),
            (0, utils_1.borshAddress)("tokenMint"),
            (0, utils_1.borshAddress)("tokenVault"),
            (0, utils_1.borshAddress)("authority"),
            borsh.u128("rewardGrowthGlobalX64"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new RewardInfo({
            rewardState: obj.rewardState,
            openTime: obj.openTime,
            endTime: obj.endTime,
            lastUpdateTime: obj.lastUpdateTime,
            emissionsPerSecondX64: obj.emissionsPerSecondX64,
            rewardTotalEmissioned: obj.rewardTotalEmissioned,
            rewardClaimed: obj.rewardClaimed,
            tokenMint: obj.tokenMint,
            tokenVault: obj.tokenVault,
            authority: obj.authority,
            rewardGrowthGlobalX64: obj.rewardGrowthGlobalX64,
        });
    }
    static toEncodable(fields) {
        return {
            rewardState: fields.rewardState,
            openTime: fields.openTime,
            endTime: fields.endTime,
            lastUpdateTime: fields.lastUpdateTime,
            emissionsPerSecondX64: fields.emissionsPerSecondX64,
            rewardTotalEmissioned: fields.rewardTotalEmissioned,
            rewardClaimed: fields.rewardClaimed,
            tokenMint: fields.tokenMint,
            tokenVault: fields.tokenVault,
            authority: fields.authority,
            rewardGrowthGlobalX64: fields.rewardGrowthGlobalX64,
        };
    }
    toJSON() {
        return {
            rewardState: this.rewardState,
            openTime: this.openTime.toString(),
            endTime: this.endTime.toString(),
            lastUpdateTime: this.lastUpdateTime.toString(),
            emissionsPerSecondX64: this.emissionsPerSecondX64.toString(),
            rewardTotalEmissioned: this.rewardTotalEmissioned.toString(),
            rewardClaimed: this.rewardClaimed.toString(),
            tokenMint: this.tokenMint,
            tokenVault: this.tokenVault,
            authority: this.authority,
            rewardGrowthGlobalX64: this.rewardGrowthGlobalX64.toString(),
        };
    }
    static fromJSON(obj) {
        return new RewardInfo({
            rewardState: obj.rewardState,
            openTime: new bn_js_1.default(obj.openTime),
            endTime: new bn_js_1.default(obj.endTime),
            lastUpdateTime: new bn_js_1.default(obj.lastUpdateTime),
            emissionsPerSecondX64: new bn_js_1.default(obj.emissionsPerSecondX64),
            rewardTotalEmissioned: new bn_js_1.default(obj.rewardTotalEmissioned),
            rewardClaimed: new bn_js_1.default(obj.rewardClaimed),
            tokenMint: (0, kit_1.address)(obj.tokenMint),
            tokenVault: (0, kit_1.address)(obj.tokenVault),
            authority: (0, kit_1.address)(obj.authority),
            rewardGrowthGlobalX64: new bn_js_1.default(obj.rewardGrowthGlobalX64),
        });
    }
    toEncodable() {
        return RewardInfo.toEncodable(this);
    }
}
exports.RewardInfo = RewardInfo;
