// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/types/TickState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TickState = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class TickState {
    tick;
    liquidityNet;
    liquidityGross;
    feeGrowthOutside0X64;
    feeGrowthOutside1X64;
    rewardGrowthsOutsideX64;
    padding;
    constructor(fields) {
        this.tick = fields.tick;
        this.liquidityNet = fields.liquidityNet;
        this.liquidityGross = fields.liquidityGross;
        this.feeGrowthOutside0X64 = fields.feeGrowthOutside0X64;
        this.feeGrowthOutside1X64 = fields.feeGrowthOutside1X64;
        this.rewardGrowthsOutsideX64 = fields.rewardGrowthsOutsideX64;
        this.padding = fields.padding;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i32("tick"),
            borsh.i128("liquidityNet"),
            borsh.u128("liquidityGross"),
            borsh.u128("feeGrowthOutside0X64"),
            borsh.u128("feeGrowthOutside1X64"),
            borsh.array(borsh.u128(), 3, "rewardGrowthsOutsideX64"),
            borsh.array(borsh.u32(), 13, "padding"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new TickState({
            tick: obj.tick,
            liquidityNet: obj.liquidityNet,
            liquidityGross: obj.liquidityGross,
            feeGrowthOutside0X64: obj.feeGrowthOutside0X64,
            feeGrowthOutside1X64: obj.feeGrowthOutside1X64,
            rewardGrowthsOutsideX64: obj.rewardGrowthsOutsideX64,
            padding: obj.padding,
        });
    }
    static toEncodable(fields) {
        return {
            tick: fields.tick,
            liquidityNet: fields.liquidityNet,
            liquidityGross: fields.liquidityGross,
            feeGrowthOutside0X64: fields.feeGrowthOutside0X64,
            feeGrowthOutside1X64: fields.feeGrowthOutside1X64,
            rewardGrowthsOutsideX64: fields.rewardGrowthsOutsideX64,
            padding: fields.padding,
        };
    }
    toJSON() {
        return {
            tick: this.tick,
            liquidityNet: this.liquidityNet.toString(),
            liquidityGross: this.liquidityGross.toString(),
            feeGrowthOutside0X64: this.feeGrowthOutside0X64.toString(),
            feeGrowthOutside1X64: this.feeGrowthOutside1X64.toString(),
            rewardGrowthsOutsideX64: this.rewardGrowthsOutsideX64.map((item) => item.toString()),
            padding: this.padding,
        };
    }
    static fromJSON(obj) {
        return new TickState({
            tick: obj.tick,
            liquidityNet: new bn_js_1.default(obj.liquidityNet),
            liquidityGross: new bn_js_1.default(obj.liquidityGross),
            feeGrowthOutside0X64: new bn_js_1.default(obj.feeGrowthOutside0X64),
            feeGrowthOutside1X64: new bn_js_1.default(obj.feeGrowthOutside1X64),
            rewardGrowthsOutsideX64: obj.rewardGrowthsOutsideX64.map((item) => new bn_js_1.default(item)),
            padding: obj.padding,
        });
    }
    toEncodable() {
        return TickState.toEncodable(this);
    }
}
exports.TickState = TickState;
