// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/BinLiquidityDistribution.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BinLiquidityDistribution = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class BinLiquidityDistribution {
    /** Define the bin ID wish to deposit to. */
    binId;
    /** DistributionX (or distributionY) is the percentages of amountX (or amountY) you want to add to each bin. */
    distributionX;
    /** DistributionX (or distributionY) is the percentages of amountX (or amountY) you want to add to each bin. */
    distributionY;
    constructor(fields) {
        this.binId = fields.binId;
        this.distributionX = fields.distributionX;
        this.distributionY = fields.distributionY;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i32("binId"),
            borsh.u16("distributionX"),
            borsh.u16("distributionY"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new BinLiquidityDistribution({
            binId: obj.binId,
            distributionX: obj.distributionX,
            distributionY: obj.distributionY,
        });
    }
    static toEncodable(fields) {
        return {
            binId: fields.binId,
            distributionX: fields.distributionX,
            distributionY: fields.distributionY,
        };
    }
    toJSON() {
        return {
            binId: this.binId,
            distributionX: this.distributionX,
            distributionY: this.distributionY,
        };
    }
    static fromJSON(obj) {
        return new BinLiquidityDistribution({
            binId: obj.binId,
            distributionX: obj.distributionX,
            distributionY: obj.distributionY,
        });
    }
    toEncodable() {
        return BinLiquidityDistribution.toEncodable(this);
    }
}
exports.BinLiquidityDistribution = BinLiquidityDistribution;
