// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/FeeParameter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeeParameter = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class FeeParameter {
    /** Portion of swap fees retained by the protocol by controlling protocol_share parameter. protocol_swap_fee = protocol_share * total_swap_fee */
    protocolShare;
    /** Base factor for base fee rate */
    baseFactor;
    constructor(fields) {
        this.protocolShare = fields.protocolShare;
        this.baseFactor = fields.baseFactor;
    }
    static layout(property) {
        return borsh.struct([borsh.u16("protocolShare"), borsh.u16("baseFactor")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FeeParameter({
            protocolShare: obj.protocolShare,
            baseFactor: obj.baseFactor,
        });
    }
    static toEncodable(fields) {
        return {
            protocolShare: fields.protocolShare,
            baseFactor: fields.baseFactor,
        };
    }
    toJSON() {
        return {
            protocolShare: this.protocolShare,
            baseFactor: this.baseFactor,
        };
    }
    static fromJSON(obj) {
        return new FeeParameter({
            protocolShare: obj.protocolShare,
            baseFactor: obj.baseFactor,
        });
    }
    toEncodable() {
        return FeeParameter.toEncodable(this);
    }
}
exports.FeeParameter = FeeParameter;
