// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PairStatus = exports.PairType = exports.LayoutVersion = exports.Rounding = exports.StrategyType = exports.UserRewardInfo = exports.FeeInfo = exports.VariableParameters = exports.StaticParameters = exports.Observation = exports.RewardInfo = exports.ProtocolFee = exports.Bin = exports.FeeParameter = exports.BinLiquidityReduction = exports.InitPresetParametersIx = exports.InitPermissionPairIx = exports.LiquidityParameter = exports.BinLiquidityDistribution = exports.CompressedBinDepositAmount = exports.AddLiquiditySingleSidePreciseParameter = exports.LiquidityParameterByWeight = exports.BinLiquidityDistributionByWeight = exports.LiquidityOneSideParameter = exports.StrategyParameters = exports.LiquidityParameterByStrategy = exports.LiquidityParameterByStrategyOneSide = void 0;
const StrategyType = __importStar(require("./StrategyType"));
exports.StrategyType = StrategyType;
const Rounding = __importStar(require("./Rounding"));
exports.Rounding = Rounding;
const LayoutVersion = __importStar(require("./LayoutVersion"));
exports.LayoutVersion = LayoutVersion;
const PairType = __importStar(require("./PairType"));
exports.PairType = PairType;
const PairStatus = __importStar(require("./PairStatus"));
exports.PairStatus = PairStatus;
var LiquidityParameterByStrategyOneSide_1 = require("./LiquidityParameterByStrategyOneSide");
Object.defineProperty(exports, "LiquidityParameterByStrategyOneSide", { enumerable: true, get: function () { return LiquidityParameterByStrategyOneSide_1.LiquidityParameterByStrategyOneSide; } });
var LiquidityParameterByStrategy_1 = require("./LiquidityParameterByStrategy");
Object.defineProperty(exports, "LiquidityParameterByStrategy", { enumerable: true, get: function () { return LiquidityParameterByStrategy_1.LiquidityParameterByStrategy; } });
var StrategyParameters_1 = require("./StrategyParameters");
Object.defineProperty(exports, "StrategyParameters", { enumerable: true, get: function () { return StrategyParameters_1.StrategyParameters; } });
var LiquidityOneSideParameter_1 = require("./LiquidityOneSideParameter");
Object.defineProperty(exports, "LiquidityOneSideParameter", { enumerable: true, get: function () { return LiquidityOneSideParameter_1.LiquidityOneSideParameter; } });
var BinLiquidityDistributionByWeight_1 = require("./BinLiquidityDistributionByWeight");
Object.defineProperty(exports, "BinLiquidityDistributionByWeight", { enumerable: true, get: function () { return BinLiquidityDistributionByWeight_1.BinLiquidityDistributionByWeight; } });
var LiquidityParameterByWeight_1 = require("./LiquidityParameterByWeight");
Object.defineProperty(exports, "LiquidityParameterByWeight", { enumerable: true, get: function () { return LiquidityParameterByWeight_1.LiquidityParameterByWeight; } });
var AddLiquiditySingleSidePreciseParameter_1 = require("./AddLiquiditySingleSidePreciseParameter");
Object.defineProperty(exports, "AddLiquiditySingleSidePreciseParameter", { enumerable: true, get: function () { return AddLiquiditySingleSidePreciseParameter_1.AddLiquiditySingleSidePreciseParameter; } });
var CompressedBinDepositAmount_1 = require("./CompressedBinDepositAmount");
Object.defineProperty(exports, "CompressedBinDepositAmount", { enumerable: true, get: function () { return CompressedBinDepositAmount_1.CompressedBinDepositAmount; } });
var BinLiquidityDistribution_1 = require("./BinLiquidityDistribution");
Object.defineProperty(exports, "BinLiquidityDistribution", { enumerable: true, get: function () { return BinLiquidityDistribution_1.BinLiquidityDistribution; } });
var LiquidityParameter_1 = require("./LiquidityParameter");
Object.defineProperty(exports, "LiquidityParameter", { enumerable: true, get: function () { return LiquidityParameter_1.LiquidityParameter; } });
var InitPermissionPairIx_1 = require("./InitPermissionPairIx");
Object.defineProperty(exports, "InitPermissionPairIx", { enumerable: true, get: function () { return InitPermissionPairIx_1.InitPermissionPairIx; } });
var InitPresetParametersIx_1 = require("./InitPresetParametersIx");
Object.defineProperty(exports, "InitPresetParametersIx", { enumerable: true, get: function () { return InitPresetParametersIx_1.InitPresetParametersIx; } });
var BinLiquidityReduction_1 = require("./BinLiquidityReduction");
Object.defineProperty(exports, "BinLiquidityReduction", { enumerable: true, get: function () { return BinLiquidityReduction_1.BinLiquidityReduction; } });
var FeeParameter_1 = require("./FeeParameter");
Object.defineProperty(exports, "FeeParameter", { enumerable: true, get: function () { return FeeParameter_1.FeeParameter; } });
var Bin_1 = require("./Bin");
Object.defineProperty(exports, "Bin", { enumerable: true, get: function () { return Bin_1.Bin; } });
var ProtocolFee_1 = require("./ProtocolFee");
Object.defineProperty(exports, "ProtocolFee", { enumerable: true, get: function () { return ProtocolFee_1.ProtocolFee; } });
var RewardInfo_1 = require("./RewardInfo");
Object.defineProperty(exports, "RewardInfo", { enumerable: true, get: function () { return RewardInfo_1.RewardInfo; } });
var Observation_1 = require("./Observation");
Object.defineProperty(exports, "Observation", { enumerable: true, get: function () { return Observation_1.Observation; } });
var StaticParameters_1 = require("./StaticParameters");
Object.defineProperty(exports, "StaticParameters", { enumerable: true, get: function () { return StaticParameters_1.StaticParameters; } });
var VariableParameters_1 = require("./VariableParameters");
Object.defineProperty(exports, "VariableParameters", { enumerable: true, get: function () { return VariableParameters_1.VariableParameters; } });
var FeeInfo_1 = require("./FeeInfo");
Object.defineProperty(exports, "FeeInfo", { enumerable: true, get: function () { return FeeInfo_1.FeeInfo; } });
var UserRewardInfo_1 = require("./UserRewardInfo");
Object.defineProperty(exports, "UserRewardInfo", { enumerable: true, get: function () { return UserRewardInfo_1.UserRewardInfo; } });
