// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/InitPermissionPairIx.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitPermissionPairIx = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class InitPermissionPairIx {
    activeId;
    binStep;
    baseFactor;
    minBinId;
    maxBinId;
    lockDurationInSlot;
    constructor(fields) {
        this.activeId = fields.activeId;
        this.binStep = fields.binStep;
        this.baseFactor = fields.baseFactor;
        this.minBinId = fields.minBinId;
        this.maxBinId = fields.maxBinId;
        this.lockDurationInSlot = fields.lockDurationInSlot;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i32("activeId"),
            borsh.u16("binStep"),
            borsh.u16("baseFactor"),
            borsh.i32("minBinId"),
            borsh.i32("maxBinId"),
            borsh.u64("lockDurationInSlot"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new InitPermissionPairIx({
            activeId: obj.activeId,
            binStep: obj.binStep,
            baseFactor: obj.baseFactor,
            minBinId: obj.minBinId,
            maxBinId: obj.maxBinId,
            lockDurationInSlot: obj.lockDurationInSlot,
        });
    }
    static toEncodable(fields) {
        return {
            activeId: fields.activeId,
            binStep: fields.binStep,
            baseFactor: fields.baseFactor,
            minBinId: fields.minBinId,
            maxBinId: fields.maxBinId,
            lockDurationInSlot: fields.lockDurationInSlot,
        };
    }
    toJSON() {
        return {
            activeId: this.activeId,
            binStep: this.binStep,
            baseFactor: this.baseFactor,
            minBinId: this.minBinId,
            maxBinId: this.maxBinId,
            lockDurationInSlot: this.lockDurationInSlot.toString(),
        };
    }
    static fromJSON(obj) {
        return new InitPermissionPairIx({
            activeId: obj.activeId,
            binStep: obj.binStep,
            baseFactor: obj.baseFactor,
            minBinId: obj.minBinId,
            maxBinId: obj.maxBinId,
            lockDurationInSlot: new bn_js_1.default(obj.lockDurationInSlot),
        });
    }
    toEncodable() {
        return InitPermissionPairIx.toEncodable(this);
    }
}
exports.InitPermissionPairIx = InitPermissionPairIx;
