// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/LiquidityParameterByStrategy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidityParameterByStrategy = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class LiquidityParameterByStrategy {
    /** Amount of X token to deposit */
    amountX;
    /** Amount of Y token to deposit */
    amountY;
    /** Active bin that integrator observe off-chain */
    activeId;
    /** max active bin slippage allowed */
    maxActiveBinSlippage;
    /** strategy parameters */
    strategyParameters;
    constructor(fields) {
        this.amountX = fields.amountX;
        this.amountY = fields.amountY;
        this.activeId = fields.activeId;
        this.maxActiveBinSlippage = fields.maxActiveBinSlippage;
        this.strategyParameters = new types.StrategyParameters({
            ...fields.strategyParameters,
        });
    }
    static layout(property) {
        return borsh.struct([
            borsh.u64("amountX"),
            borsh.u64("amountY"),
            borsh.i32("activeId"),
            borsh.i32("maxActiveBinSlippage"),
            types.StrategyParameters.layout("strategyParameters"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new LiquidityParameterByStrategy({
            amountX: obj.amountX,
            amountY: obj.amountY,
            activeId: obj.activeId,
            maxActiveBinSlippage: obj.maxActiveBinSlippage,
            strategyParameters: types.StrategyParameters.fromDecoded(obj.strategyParameters),
        });
    }
    static toEncodable(fields) {
        return {
            amountX: fields.amountX,
            amountY: fields.amountY,
            activeId: fields.activeId,
            maxActiveBinSlippage: fields.maxActiveBinSlippage,
            strategyParameters: types.StrategyParameters.toEncodable(fields.strategyParameters),
        };
    }
    toJSON() {
        return {
            amountX: this.amountX.toString(),
            amountY: this.amountY.toString(),
            activeId: this.activeId,
            maxActiveBinSlippage: this.maxActiveBinSlippage,
            strategyParameters: this.strategyParameters.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new LiquidityParameterByStrategy({
            amountX: new bn_js_1.default(obj.amountX),
            amountY: new bn_js_1.default(obj.amountY),
            activeId: obj.activeId,
            maxActiveBinSlippage: obj.maxActiveBinSlippage,
            strategyParameters: types.StrategyParameters.fromJSON(obj.strategyParameters),
        });
    }
    toEncodable() {
        return LiquidityParameterByStrategy.toEncodable(this);
    }
}
exports.LiquidityParameterByStrategy = LiquidityParameterByStrategy;
