// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/LiquidityOneSideParameter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidityOneSideParameter = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class LiquidityOneSideParameter {
    /** Amount of X token or Y token to deposit */
    amount;
    /** Active bin that integrator observe off-chain */
    activeId;
    /** max active bin slippage allowed */
    maxActiveBinSlippage;
    /** Liquidity distribution to each bins */
    binLiquidityDist;
    constructor(fields) {
        this.amount = fields.amount;
        this.activeId = fields.activeId;
        this.maxActiveBinSlippage = fields.maxActiveBinSlippage;
        this.binLiquidityDist = fields.binLiquidityDist.map((item) => new types.BinLiquidityDistributionByWeight({ ...item }));
    }
    static layout(property) {
        return borsh.struct([
            borsh.u64("amount"),
            borsh.i32("activeId"),
            borsh.i32("maxActiveBinSlippage"),
            borsh.vec(types.BinLiquidityDistributionByWeight.layout(), "binLiquidityDist"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new LiquidityOneSideParameter({
            amount: obj.amount,
            activeId: obj.activeId,
            maxActiveBinSlippage: obj.maxActiveBinSlippage,
            binLiquidityDist: obj.binLiquidityDist.map((item /* eslint-disable-line @typescript-eslint/no-explicit-any */) => types.BinLiquidityDistributionByWeight.fromDecoded(item)),
        });
    }
    static toEncodable(fields) {
        return {
            amount: fields.amount,
            activeId: fields.activeId,
            maxActiveBinSlippage: fields.maxActiveBinSlippage,
            binLiquidityDist: fields.binLiquidityDist.map((item) => types.BinLiquidityDistributionByWeight.toEncodable(item)),
        };
    }
    toJSON() {
        return {
            amount: this.amount.toString(),
            activeId: this.activeId,
            maxActiveBinSlippage: this.maxActiveBinSlippage,
            binLiquidityDist: this.binLiquidityDist.map((item) => item.toJSON()),
        };
    }
    static fromJSON(obj) {
        return new LiquidityOneSideParameter({
            amount: new bn_js_1.default(obj.amount),
            activeId: obj.activeId,
            maxActiveBinSlippage: obj.maxActiveBinSlippage,
            binLiquidityDist: obj.binLiquidityDist.map((item) => types.BinLiquidityDistributionByWeight.fromJSON(item)),
        });
    }
    toEncodable() {
        return LiquidityOneSideParameter.toEncodable(this);
    }
}
exports.LiquidityOneSideParameter = LiquidityOneSideParameter;
