// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/LiquidityParameter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidityParameter = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class LiquidityParameter {
    /** Amount of X token to deposit */
    amountX;
    /** Amount of Y token to deposit */
    amountY;
    /** Liquidity distribution to each bins */
    binLiquidityDist;
    constructor(fields) {
        this.amountX = fields.amountX;
        this.amountY = fields.amountY;
        this.binLiquidityDist = fields.binLiquidityDist.map((item) => new types.BinLiquidityDistribution({ ...item }));
    }
    static layout(property) {
        return borsh.struct([
            borsh.u64("amountX"),
            borsh.u64("amountY"),
            borsh.vec(types.BinLiquidityDistribution.layout(), "binLiquidityDist"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new LiquidityParameter({
            amountX: obj.amountX,
            amountY: obj.amountY,
            binLiquidityDist: obj.binLiquidityDist.map((item /* eslint-disable-line @typescript-eslint/no-explicit-any */) => types.BinLiquidityDistribution.fromDecoded(item)),
        });
    }
    static toEncodable(fields) {
        return {
            amountX: fields.amountX,
            amountY: fields.amountY,
            binLiquidityDist: fields.binLiquidityDist.map((item) => types.BinLiquidityDistribution.toEncodable(item)),
        };
    }
    toJSON() {
        return {
            amountX: this.amountX.toString(),
            amountY: this.amountY.toString(),
            binLiquidityDist: this.binLiquidityDist.map((item) => item.toJSON()),
        };
    }
    static fromJSON(obj) {
        return new LiquidityParameter({
            amountX: new bn_js_1.default(obj.amountX),
            amountY: new bn_js_1.default(obj.amountY),
            binLiquidityDist: obj.binLiquidityDist.map((item) => types.BinLiquidityDistribution.fromJSON(item)),
        });
    }
    toEncodable() {
        return LiquidityParameter.toEncodable(this);
    }
}
exports.LiquidityParameter = LiquidityParameter;
