// @kamino-finance/kliquidity-sdk/dist/@codegen/meteora/types/StrategyParameters.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.StrategyParameters = void 0;
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class StrategyParameters {
    /** min bin id */
    minBinId;
    /** max bin id */
    maxBinId;
    /** strategy type */
    strategyType;
    /** parameters */
    parameteres;
    constructor(fields) {
        this.minBinId = fields.minBinId;
        this.maxBinId = fields.maxBinId;
        this.strategyType = fields.strategyType;
        this.parameteres = fields.parameteres;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i32("minBinId"),
            borsh.i32("maxBinId"),
            types.StrategyType.layout("strategyType"),
            borsh.array(borsh.u8(), 64, "parameteres"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new StrategyParameters({
            minBinId: obj.minBinId,
            maxBinId: obj.maxBinId,
            strategyType: types.StrategyType.fromDecoded(obj.strategyType),
            parameteres: obj.parameteres,
        });
    }
    static toEncodable(fields) {
        return {
            minBinId: fields.minBinId,
            maxBinId: fields.maxBinId,
            strategyType: fields.strategyType.toEncodable(),
            parameteres: fields.parameteres,
        };
    }
    toJSON() {
        return {
            minBinId: this.minBinId,
            maxBinId: this.maxBinId,
            strategyType: this.strategyType.toJSON(),
            parameteres: this.parameteres,
        };
    }
    static fromJSON(obj) {
        return new StrategyParameters({
            minBinId: obj.minBinId,
            maxBinId: obj.maxBinId,
            strategyType: types.StrategyType.fromJSON(obj.strategyType),
            parameteres: obj.parameteres,
        });
    }
    toEncodable() {
        return StrategyParameters.toEncodable(this);
    }
}
exports.StrategyParameters = StrategyParameters;
