// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/types/AccountsType.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferHookOutput = exports.TransferHookIntermediate = exports.TransferHookInput = exports.TransferHookReward = exports.TransferHookB = exports.TransferHookA = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class TransferHookA {
    static discriminator = 0;
    static kind = "TransferHookA";
    discriminator = 0;
    kind = "TransferHookA";
    toJSON() {
        return {
            kind: "TransferHookA",
        };
    }
    toEncodable() {
        return {
            TransferHookA: {},
        };
    }
}
exports.TransferHookA = TransferHookA;
class TransferHookB {
    static discriminator = 1;
    static kind = "TransferHookB";
    discriminator = 1;
    kind = "TransferHookB";
    toJSON() {
        return {
            kind: "TransferHookB",
        };
    }
    toEncodable() {
        return {
            TransferHookB: {},
        };
    }
}
exports.TransferHookB = TransferHookB;
class TransferHookReward {
    static discriminator = 2;
    static kind = "TransferHookReward";
    discriminator = 2;
    kind = "TransferHookReward";
    toJSON() {
        return {
            kind: "TransferHookReward",
        };
    }
    toEncodable() {
        return {
            TransferHookReward: {},
        };
    }
}
exports.TransferHookReward = TransferHookReward;
class TransferHookInput {
    static discriminator = 3;
    static kind = "TransferHookInput";
    discriminator = 3;
    kind = "TransferHookInput";
    toJSON() {
        return {
            kind: "TransferHookInput",
        };
    }
    toEncodable() {
        return {
            TransferHookInput: {},
        };
    }
}
exports.TransferHookInput = TransferHookInput;
class TransferHookIntermediate {
    static discriminator = 4;
    static kind = "TransferHookIntermediate";
    discriminator = 4;
    kind = "TransferHookIntermediate";
    toJSON() {
        return {
            kind: "TransferHookIntermediate",
        };
    }
    toEncodable() {
        return {
            TransferHookIntermediate: {},
        };
    }
}
exports.TransferHookIntermediate = TransferHookIntermediate;
class TransferHookOutput {
    static discriminator = 5;
    static kind = "TransferHookOutput";
    discriminator = 5;
    kind = "TransferHookOutput";
    toJSON() {
        return {
            kind: "TransferHookOutput",
        };
    }
    toEncodable() {
        return {
            TransferHookOutput: {},
        };
    }
}
exports.TransferHookOutput = TransferHookOutput;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("TransferHookA" in obj) {
        return new TransferHookA();
    }
    if ("TransferHookB" in obj) {
        return new TransferHookB();
    }
    if ("TransferHookReward" in obj) {
        return new TransferHookReward();
    }
    if ("TransferHookInput" in obj) {
        return new TransferHookInput();
    }
    if ("TransferHookIntermediate" in obj) {
        return new TransferHookIntermediate();
    }
    if ("TransferHookOutput" in obj) {
        return new TransferHookOutput();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "TransferHookA": {
            return new TransferHookA();
        }
        case "TransferHookB": {
            return new TransferHookB();
        }
        case "TransferHookReward": {
            return new TransferHookReward();
        }
        case "TransferHookInput": {
            return new TransferHookInput();
        }
        case "TransferHookIntermediate": {
            return new TransferHookIntermediate();
        }
        case "TransferHookOutput": {
            return new TransferHookOutput();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "TransferHookA"),
        borsh.struct([], "TransferHookB"),
        borsh.struct([], "TransferHookReward"),
        borsh.struct([], "TransferHookInput"),
        borsh.struct([], "TransferHookIntermediate"),
        borsh.struct([], "TransferHookOutput"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
