// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountsType = exports.Direction = exports.TickLabel = exports.CurrIndex = exports.RemainingAccountsInfo = exports.RemainingAccountsSlice = exports.WhirlpoolBumps = exports.WhirlpoolRewardInfo = exports.Tick = exports.PositionRewardInfo = exports.OpenPositionWithMetadataBumps = exports.OpenPositionBumps = void 0;
const CurrIndex = __importStar(require("./CurrIndex"));
exports.CurrIndex = CurrIndex;
const TickLabel = __importStar(require("./TickLabel"));
exports.TickLabel = TickLabel;
const Direction = __importStar(require("./Direction"));
exports.Direction = Direction;
const AccountsType = __importStar(require("./AccountsType"));
exports.AccountsType = AccountsType;
var OpenPositionBumps_1 = require("./OpenPositionBumps");
Object.defineProperty(exports, "OpenPositionBumps", { enumerable: true, get: function () { return OpenPositionBumps_1.OpenPositionBumps; } });
var OpenPositionWithMetadataBumps_1 = require("./OpenPositionWithMetadataBumps");
Object.defineProperty(exports, "OpenPositionWithMetadataBumps", { enumerable: true, get: function () { return OpenPositionWithMetadataBumps_1.OpenPositionWithMetadataBumps; } });
var PositionRewardInfo_1 = require("./PositionRewardInfo");
Object.defineProperty(exports, "PositionRewardInfo", { enumerable: true, get: function () { return PositionRewardInfo_1.PositionRewardInfo; } });
var Tick_1 = require("./Tick");
Object.defineProperty(exports, "Tick", { enumerable: true, get: function () { return Tick_1.Tick; } });
var WhirlpoolRewardInfo_1 = require("./WhirlpoolRewardInfo");
Object.defineProperty(exports, "WhirlpoolRewardInfo", { enumerable: true, get: function () { return WhirlpoolRewardInfo_1.WhirlpoolRewardInfo; } });
var WhirlpoolBumps_1 = require("./WhirlpoolBumps");
Object.defineProperty(exports, "WhirlpoolBumps", { enumerable: true, get: function () { return WhirlpoolBumps_1.WhirlpoolBumps; } });
var RemainingAccountsSlice_1 = require("./RemainingAccountsSlice");
Object.defineProperty(exports, "RemainingAccountsSlice", { enumerable: true, get: function () { return RemainingAccountsSlice_1.RemainingAccountsSlice; } });
var RemainingAccountsInfo_1 = require("./RemainingAccountsInfo");
Object.defineProperty(exports, "RemainingAccountsInfo", { enumerable: true, get: function () { return RemainingAccountsInfo_1.RemainingAccountsInfo; } });
