// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/types/RemainingAccountsSlice.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemainingAccountsSlice = void 0;
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class RemainingAccountsSlice {
    accountsType;
    length;
    constructor(fields) {
        this.accountsType = fields.accountsType;
        this.length = fields.length;
    }
    static layout(property) {
        return borsh.struct([types.AccountsType.layout("accountsType"), borsh.u8("length")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new RemainingAccountsSlice({
            accountsType: types.AccountsType.fromDecoded(obj.accountsType),
            length: obj.length,
        });
    }
    static toEncodable(fields) {
        return {
            accountsType: fields.accountsType.toEncodable(),
            length: fields.length,
        };
    }
    toJSON() {
        return {
            accountsType: this.accountsType.toJSON(),
            length: this.length,
        };
    }
    static fromJSON(obj) {
        return new RemainingAccountsSlice({
            accountsType: types.AccountsType.fromJSON(obj.accountsType),
            length: obj.length,
        });
    }
    toEncodable() {
        return RemainingAccountsSlice.toEncodable(this);
    }
}
exports.RemainingAccountsSlice = RemainingAccountsSlice;
