// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/types/Tick.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tick = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class Tick {
    initialized;
    liquidityNet;
    liquidityGross;
    feeGrowthOutsideA;
    feeGrowthOutsideB;
    rewardGrowthsOutside;
    constructor(fields) {
        this.initialized = fields.initialized;
        this.liquidityNet = fields.liquidityNet;
        this.liquidityGross = fields.liquidityGross;
        this.feeGrowthOutsideA = fields.feeGrowthOutsideA;
        this.feeGrowthOutsideB = fields.feeGrowthOutsideB;
        this.rewardGrowthsOutside = fields.rewardGrowthsOutside;
    }
    static layout(property) {
        return borsh.struct([
            borsh.bool("initialized"),
            borsh.i128("liquidityNet"),
            borsh.u128("liquidityGross"),
            borsh.u128("feeGrowthOutsideA"),
            borsh.u128("feeGrowthOutsideB"),
            borsh.array(borsh.u128(), 3, "rewardGrowthsOutside"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new Tick({
            initialized: obj.initialized,
            liquidityNet: obj.liquidityNet,
            liquidityGross: obj.liquidityGross,
            feeGrowthOutsideA: obj.feeGrowthOutsideA,
            feeGrowthOutsideB: obj.feeGrowthOutsideB,
            rewardGrowthsOutside: obj.rewardGrowthsOutside,
        });
    }
    static toEncodable(fields) {
        return {
            initialized: fields.initialized,
            liquidityNet: fields.liquidityNet,
            liquidityGross: fields.liquidityGross,
            feeGrowthOutsideA: fields.feeGrowthOutsideA,
            feeGrowthOutsideB: fields.feeGrowthOutsideB,
            rewardGrowthsOutside: fields.rewardGrowthsOutside,
        };
    }
    toJSON() {
        return {
            initialized: this.initialized,
            liquidityNet: this.liquidityNet.toString(),
            liquidityGross: this.liquidityGross.toString(),
            feeGrowthOutsideA: this.feeGrowthOutsideA.toString(),
            feeGrowthOutsideB: this.feeGrowthOutsideB.toString(),
            rewardGrowthsOutside: this.rewardGrowthsOutside.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new Tick({
            initialized: obj.initialized,
            liquidityNet: new bn_js_1.default(obj.liquidityNet),
            liquidityGross: new bn_js_1.default(obj.liquidityGross),
            feeGrowthOutsideA: new bn_js_1.default(obj.feeGrowthOutsideA),
            feeGrowthOutsideB: new bn_js_1.default(obj.feeGrowthOutsideB),
            rewardGrowthsOutside: obj.rewardGrowthsOutside.map((item) => new bn_js_1.default(item)),
        });
    }
    toEncodable() {
        return Tick.toEncodable(this);
    }
}
exports.Tick = Tick;
