// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/types/WhirlpoolRewardInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhirlpoolRewardInfo = void 0;
const kit_1 = require("@solana/kit"); // eslint-disable-line @typescript-eslint/no-unused-vars
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
const utils_1 = require("../utils");
class WhirlpoolRewardInfo {
    mint;
    vault;
    authority;
    emissionsPerSecondX64;
    growthGlobalX64;
    constructor(fields) {
        this.mint = fields.mint;
        this.vault = fields.vault;
        this.authority = fields.authority;
        this.emissionsPerSecondX64 = fields.emissionsPerSecondX64;
        this.growthGlobalX64 = fields.growthGlobalX64;
    }
    static layout(property) {
        return borsh.struct([
            (0, utils_1.borshAddress)("mint"),
            (0, utils_1.borshAddress)("vault"),
            (0, utils_1.borshAddress)("authority"),
            borsh.u128("emissionsPerSecondX64"),
            borsh.u128("growthGlobalX64"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new WhirlpoolRewardInfo({
            mint: obj.mint,
            vault: obj.vault,
            authority: obj.authority,
            emissionsPerSecondX64: obj.emissionsPerSecondX64,
            growthGlobalX64: obj.growthGlobalX64,
        });
    }
    static toEncodable(fields) {
        return {
            mint: fields.mint,
            vault: fields.vault,
            authority: fields.authority,
            emissionsPerSecondX64: fields.emissionsPerSecondX64,
            growthGlobalX64: fields.growthGlobalX64,
        };
    }
    toJSON() {
        return {
            mint: this.mint,
            vault: this.vault,
            authority: this.authority,
            emissionsPerSecondX64: this.emissionsPerSecondX64.toString(),
            growthGlobalX64: this.growthGlobalX64.toString(),
        };
    }
    static fromJSON(obj) {
        return new WhirlpoolRewardInfo({
            mint: (0, kit_1.address)(obj.mint),
            vault: (0, kit_1.address)(obj.vault),
            authority: (0, kit_1.address)(obj.authority),
            emissionsPerSecondX64: new bn_js_1.default(obj.emissionsPerSecondX64),
            growthGlobalX64: new bn_js_1.default(obj.growthGlobalX64),
        });
    }
    toEncodable() {
        return WhirlpoolRewardInfo.toEncodable(this);
    }
}
exports.WhirlpoolRewardInfo = WhirlpoolRewardInfo;
