// @kamino-finance/kliquidity-sdk/dist/constants/pubkeys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MEMO_PROGRAM_ID = exports.CONSENSUS_ID = exports.STAGING_GLOBAL_CONFIG = exports.STAGING_KAMINO_PROGRAM_ID = exports.STAGING_TOKEN_INFOS = exports.MAINNET_TOKEN_INFOS = exports.LUT_OWNER_KEY = exports.WRAPPED_SOL_MINT = exports.DEFAULT_PUBLIC_KEY = void 0;
const kit_1 = require("@solana/kit");
exports.DEFAULT_PUBLIC_KEY = (0, kit_1.address)('11111111111111111111111111111111');
exports.WRAPPED_SOL_MINT = (0, kit_1.address)('So11111111111111111111111111111111111111112');
exports.LUT_OWNER_KEY = (0, kit_1.address)('LutUqsNaBmLhAT58QVZx17Ku1yFa3des5bQJSWKoeqs');
exports.MAINNET_TOKEN_INFOS = (0, kit_1.address)('3v6ootgJJZbSWEDfZMA1scfh7wcsVVfeocExRxPqCyWH');
exports.STAGING_TOKEN_INFOS = (0, kit_1.address)('78MPQyWGwuERWoFwyryt1yhSanYc4otRGKygU6qnTzDW');
exports.STAGING_KAMINO_PROGRAM_ID = (0, kit_1.address)('SKY3EZaE5p8iXG1ed4kiandK1wnwwqxmWBhjEFykaHB');
exports.STAGING_GLOBAL_CONFIG = (0, kit_1.address)('7D9KE8xxqvsSsPbpTK9DbvkYaodda1wVevPvZJbLGJ71');
exports.CONSENSUS_ID = (0, kit_1.address)('9GJmEHGom9eWo4np4L5vC6b6ri1Df2xN8KFoWixvD1Bs');
exports.MEMO_PROGRAM_ID = (0, kit_1.address)('MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr');
