// @mrgnlabs/marginfi-client-v2/dist/vendor/pyth_crank/idl/pyth_push_oracle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PYTH_PUSH_ORACLE_IDL = void 0;
exports.PYTH_PUSH_ORACLE_IDL = {
    address: "pythWSnswVUd12oZpeFP8e9CVaEqJg25g1Vtc2biRsT",
    metadata: {
        name: "pythPushOracle",
        version: "0.1.0",
        spec: "0.1.0",
    },
    instructions: [
        {
            name: "updatePriceFeed",
            discriminator: [28, 9, 93, 150, 86, 153, 188, 115],
            accounts: [
                {
                    name: "payer",
                    writable: true,
                    signer: true,
                },
                {
                    name: "pythSolanaReceiver",
                },
                {
                    name: "encodedVaa",
                },
                {
                    name: "config",
                },
                {
                    name: "treasury",
                    writable: true,
                },
                {
                    name: "priceFeedAccount",
                    writable: true,
                },
                {
                    name: "systemProgram",
                },
            ],
            args: [
                {
                    name: "params",
                    type: {
                        defined: {
                            name: "postUpdateParams",
                        },
                    },
                },
                {
                    name: "shardId",
                    type: "u16",
                },
                {
                    name: "feedId",
                    type: {
                        array: ["u8", 32],
                    },
                },
            ],
        },
    ],
    errors: [
        {
            code: 6000,
            name: "updatesNotMonotonic",
            msg: "Updates must be monotonically increasing",
        },
        {
            code: 6001,
            name: "priceFeedMessageMismatch",
            msg: "Trying to update price feed with the wrong feed id",
        },
    ],
    types: [
        {
            name: "postUpdateParams",
            type: {
                kind: "struct",
                fields: [
                    {
                        name: "merklePriceUpdate",
                        type: {
                            defined: {
                                name: "merklePriceUpdate",
                            },
                        },
                    },
                    {
                        name: "treasuryId",
                        type: "u8",
                    },
                ],
            },
        },
        {
            name: "merklePriceUpdate",
            type: {
                kind: "struct",
                fields: [
                    {
                        name: "message",
                        type: "bytes",
                    },
                    {
                        name: "proof",
                        type: {
                            vec: {
                                array: ["u8", 20],
                            },
                        },
                    },
                ],
            },
        },
    ],
};
