// @kamino-finance/klend-sdk/dist/@codegen/unstaking_pool/instructions/mint.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.mint = mint;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.u64("stakedSolToDeposit"),
    borsh.u64("minSharesToReceive"),
]);
function mint(args, accounts, programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.user.address, role: 3, signer: accounts.user },
        { address: accounts.unstakeTicket, role: 1 },
        {
            address: accounts.actionAuthority.address,
            role: 2,
            signer: accounts.actionAuthority,
        },
        { address: accounts.userStakedSolToken, role: 1 },
        { address: accounts.userUnstakingSolToken, role: 1 },
        { address: accounts.stakedSolMint, role: 1 },
        { address: accounts.poolState, role: 1 },
        { address: accounts.unstakingSolMint, role: 1 },
        { address: accounts.basePoolAuthority, role: 0 },
        { address: accounts.unstakingTicketAuthority, role: 0 },
        { address: accounts.systemProgram, role: 0 },
        { address: accounts.unstakingSolTokenProgram, role: 0 },
        { address: accounts.stakedSolTokenProgram, role: 0 },
    ];
    const identifier = Buffer.from([51, 57, 225, 47, 182, 146, 137, 166]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        stakedSolToDeposit: args.stakedSolToDeposit,
        minSharesToReceive: args.minSharesToReceive,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
