// @kamino-finance/scope-sdk/dist/@codegen/scope/types/EmaTwap.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmaTwap = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class EmaTwap {
    lastUpdateSlot;
    lastUpdateUnixTimestamp;
    currentEma1h;
    /** The sample tracker is a 64 bit number where each bit represents a point in time. */
    updatesTracker1h;
    padding0;
    padding1;
    constructor(fields) {
        this.lastUpdateSlot = fields.lastUpdateSlot;
        this.lastUpdateUnixTimestamp = fields.lastUpdateUnixTimestamp;
        this.currentEma1h = fields.currentEma1h;
        this.updatesTracker1h = fields.updatesTracker1h;
        this.padding0 = fields.padding0;
        this.padding1 = fields.padding1;
    }
    static layout(property) {
        return borsh.struct([
            borsh.u64("lastUpdateSlot"),
            borsh.u64("lastUpdateUnixTimestamp"),
            borsh.u128("currentEma1h"),
            borsh.u64("updatesTracker1h"),
            borsh.u64("padding0"),
            borsh.array(borsh.u128(), 39, "padding1"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new EmaTwap({
            lastUpdateSlot: obj.lastUpdateSlot,
            lastUpdateUnixTimestamp: obj.lastUpdateUnixTimestamp,
            currentEma1h: obj.currentEma1h,
            updatesTracker1h: obj.updatesTracker1h,
            padding0: obj.padding0,
            padding1: obj.padding1,
        });
    }
    static toEncodable(fields) {
        return {
            lastUpdateSlot: fields.lastUpdateSlot,
            lastUpdateUnixTimestamp: fields.lastUpdateUnixTimestamp,
            currentEma1h: fields.currentEma1h,
            updatesTracker1h: fields.updatesTracker1h,
            padding0: fields.padding0,
            padding1: fields.padding1,
        };
    }
    toJSON() {
        return {
            lastUpdateSlot: this.lastUpdateSlot.toString(),
            lastUpdateUnixTimestamp: this.lastUpdateUnixTimestamp.toString(),
            currentEma1h: this.currentEma1h.toString(),
            updatesTracker1h: this.updatesTracker1h.toString(),
            padding0: this.padding0.toString(),
            padding1: this.padding1.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new EmaTwap({
            lastUpdateSlot: new bn_js_1.default(obj.lastUpdateSlot),
            lastUpdateUnixTimestamp: new bn_js_1.default(obj.lastUpdateUnixTimestamp),
            currentEma1h: new bn_js_1.default(obj.currentEma1h),
            updatesTracker1h: new bn_js_1.default(obj.updatesTracker1h),
            padding0: new bn_js_1.default(obj.padding0),
            padding1: obj.padding1.map((item) => new bn_js_1.default(item)),
        });
    }
    toEncodable() {
        return EmaTwap.toEncodable(this);
    }
}
exports.EmaTwap = EmaTwap;
