// @kamino-finance/scope-sdk/dist/@codegen/scope/types/List.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = void 0;
const kit_1 = require("@solana/kit"); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
const utils_1 = require("../utils");
class List {
    account;
    itemSize;
    count;
    newAccount;
    copiedCount;
    constructor(fields) {
        this.account = fields.account;
        this.itemSize = fields.itemSize;
        this.count = fields.count;
        this.newAccount = fields.newAccount;
        this.copiedCount = fields.copiedCount;
    }
    static layout(property) {
        return borsh.struct([
            (0, utils_1.borshAddress)("account"),
            borsh.u32("itemSize"),
            borsh.u32("count"),
            (0, utils_1.borshAddress)("newAccount"),
            borsh.u32("copiedCount"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new List({
            account: obj.account,
            itemSize: obj.itemSize,
            count: obj.count,
            newAccount: obj.newAccount,
            copiedCount: obj.copiedCount,
        });
    }
    static toEncodable(fields) {
        return {
            account: fields.account,
            itemSize: fields.itemSize,
            count: fields.count,
            newAccount: fields.newAccount,
            copiedCount: fields.copiedCount,
        };
    }
    toJSON() {
        return {
            account: this.account,
            itemSize: this.itemSize,
            count: this.count,
            newAccount: this.newAccount,
            copiedCount: this.copiedCount,
        };
    }
    static fromJSON(obj) {
        return new List({
            account: (0, kit_1.address)(obj.account),
            itemSize: obj.itemSize,
            count: obj.count,
            newAccount: (0, kit_1.address)(obj.newAccount),
            copiedCount: obj.copiedCount,
        });
    }
    toEncodable() {
        return List.toEncodable(this);
    }
}
exports.List = List;
