// @kamino-finance/scope-sdk/dist/@codegen/scope/types/TokenMetadata.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenMetadata = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class TokenMetadata {
    name;
    maxAgePriceSlots;
    groupIdsBitset;
    reserved;
    constructor(fields) {
        this.name = fields.name;
        this.maxAgePriceSlots = fields.maxAgePriceSlots;
        this.groupIdsBitset = fields.groupIdsBitset;
        this.reserved = fields.reserved;
    }
    static layout(property) {
        return borsh.struct([
            borsh.array(borsh.u8(), 32, "name"),
            borsh.u64("maxAgePriceSlots"),
            borsh.u64("groupIdsBitset"),
            borsh.array(borsh.u64(), 15, "reserved"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new TokenMetadata({
            name: obj.name,
            maxAgePriceSlots: obj.maxAgePriceSlots,
            groupIdsBitset: obj.groupIdsBitset,
            reserved: obj.reserved,
        });
    }
    static toEncodable(fields) {
        return {
            name: fields.name,
            maxAgePriceSlots: fields.maxAgePriceSlots,
            groupIdsBitset: fields.groupIdsBitset,
            reserved: fields.reserved,
        };
    }
    toJSON() {
        return {
            name: this.name,
            maxAgePriceSlots: this.maxAgePriceSlots.toString(),
            groupIdsBitset: this.groupIdsBitset.toString(),
            reserved: this.reserved.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new TokenMetadata({
            name: obj.name,
            maxAgePriceSlots: new bn_js_1.default(obj.maxAgePriceSlots),
            groupIdsBitset: new bn_js_1.default(obj.groupIdsBitset),
            reserved: obj.reserved.map((item) => new bn_js_1.default(item)),
        });
    }
    toEncodable() {
        return TokenMetadata.toEncodable(this);
    }
}
exports.TokenMetadata = TokenMetadata;
