// @kamino-finance/scope-sdk/dist/@codegen/scope/types/UpdateTokenMetadataMode.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupIds = exports.MaxPriceAgeSlots = exports.Name = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Name {
    static discriminator = 0;
    static kind = "Name";
    discriminator = 0;
    kind = "Name";
    toJSON() {
        return {
            kind: "Name",
        };
    }
    toEncodable() {
        return {
            Name: {},
        };
    }
}
exports.Name = Name;
class MaxPriceAgeSlots {
    static discriminator = 1;
    static kind = "MaxPriceAgeSlots";
    discriminator = 1;
    kind = "MaxPriceAgeSlots";
    toJSON() {
        return {
            kind: "MaxPriceAgeSlots",
        };
    }
    toEncodable() {
        return {
            MaxPriceAgeSlots: {},
        };
    }
}
exports.MaxPriceAgeSlots = MaxPriceAgeSlots;
class GroupIds {
    static discriminator = 2;
    static kind = "GroupIds";
    discriminator = 2;
    kind = "GroupIds";
    toJSON() {
        return {
            kind: "GroupIds",
        };
    }
    toEncodable() {
        return {
            GroupIds: {},
        };
    }
}
exports.GroupIds = GroupIds;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Name" in obj) {
        return new Name();
    }
    if ("MaxPriceAgeSlots" in obj) {
        return new MaxPriceAgeSlots();
    }
    if ("GroupIds" in obj) {
        return new GroupIds();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Name": {
            return new Name();
        }
        case "MaxPriceAgeSlots": {
            return new MaxPriceAgeSlots();
        }
        case "GroupIds": {
            return new GroupIds();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Name"),
        borsh.struct([], "MaxPriceAgeSlots"),
        borsh.struct([], "GroupIds"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
