// @kamino-finance/scope-sdk/dist/@codegen/scope/accounts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MintsToScopeChains = exports.Configuration = exports.TokenMetadatas = exports.OracleMappings = exports.OraclePrices = exports.OracleTwaps = void 0;
var OracleTwaps_1 = require("./OracleTwaps");
Object.defineProperty(exports, "OracleTwaps", { enumerable: true, get: function () { return OracleTwaps_1.OracleTwaps; } });
var OraclePrices_1 = require("./OraclePrices");
Object.defineProperty(exports, "OraclePrices", { enumerable: true, get: function () { return OraclePrices_1.OraclePrices; } });
var OracleMappings_1 = require("./OracleMappings");
Object.defineProperty(exports, "OracleMappings", { enumerable: true, get: function () { return OracleMappings_1.OracleMappings; } });
var TokenMetadatas_1 = require("./TokenMetadatas");
Object.defineProperty(exports, "TokenMetadatas", { enumerable: true, get: function () { return TokenMetadatas_1.TokenMetadatas; } });
var Configuration_1 = require("./Configuration");
Object.defineProperty(exports, "Configuration", { enumerable: true, get: function () { return Configuration_1.Configuration; } });
var MintsToScopeChains_1 = require("./MintsToScopeChains");
Object.defineProperty(exports, "MintsToScopeChains", { enumerable: true, get: function () { return MintsToScopeChains_1.MintsToScopeChains; } });
