// @kamino-finance/scope-sdk/dist/@codegen/scope/accounts/TokenMetadatas.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenMetadatas = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class TokenMetadatas {
    metadatasArray;
    static discriminator = Buffer.from([
        221, 107, 64, 103, 67, 0, 165, 22,
    ]);
    static layout = borsh.struct([
        borsh.array(types.TokenMetadata.layout(), 512, "metadatasArray"),
    ]);
    constructor(fields) {
        this.metadatasArray = fields.metadatasArray.map((item) => new types.TokenMetadata({ ...item }));
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error("account doesn't belong to this program");
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error("account doesn't belong to this program");
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(TokenMetadatas.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = TokenMetadatas.layout.decode(data.slice(8));
        return new TokenMetadatas({
            metadatasArray: dec.metadatasArray.map((item /* eslint-disable-line @typescript-eslint/no-explicit-any */) => types.TokenMetadata.fromDecoded(item)),
        });
    }
    toJSON() {
        return {
            metadatasArray: this.metadatasArray.map((item) => item.toJSON()),
        };
    }
    static fromJSON(obj) {
        return new TokenMetadatas({
            metadatasArray: obj.metadatasArray.map((item) => types.TokenMetadata.fromJSON(item)),
        });
    }
}
exports.TokenMetadatas = TokenMetadatas;
