// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/BalanceStatus.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Unbalanced = exports.Balanced = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Balanced {
    static discriminator = 0;
    static kind = "Balanced";
    discriminator = 0;
    kind = "Balanced";
    toJSON() {
        return {
            kind: "Balanced",
        };
    }
    toEncodable() {
        return {
            Balanced: {},
        };
    }
}
exports.Balanced = Balanced;
class Unbalanced {
    static discriminator = 1;
    static kind = "Unbalanced";
    discriminator = 1;
    kind = "Unbalanced";
    toJSON() {
        return {
            kind: "Unbalanced",
        };
    }
    toEncodable() {
        return {
            Unbalanced: {},
        };
    }
}
exports.Unbalanced = Unbalanced;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Balanced" in obj) {
        return new Balanced();
    }
    if ("Unbalanced" in obj) {
        return new Unbalanced();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Balanced": {
            return new Balanced();
        }
        case "Unbalanced": {
            return new Unbalanced();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Balanced"),
        borsh.struct([], "Unbalanced"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
