// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/BinAddLiquidityStrategy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrentTick = exports.Uniform = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class Uniform {
    static discriminator = 0;
    static kind = "Uniform";
    discriminator = 0;
    kind = "Uniform";
    value;
    constructor(value) {
        this.value = {
            currentBinIndex: value.currentBinIndex,
            lowerBinIndex: value.lowerBinIndex,
            upperBinIndex: value.upperBinIndex,
            amountXtoDeposit: value.amountXtoDeposit,
            amountYtoDeposit: value.amountYtoDeposit,
            xCurrentBin: value.xCurrentBin,
            yCurrentBin: value.yCurrentBin,
        };
    }
    toJSON() {
        return {
            kind: "Uniform",
            value: {
                currentBinIndex: this.value.currentBinIndex,
                lowerBinIndex: this.value.lowerBinIndex,
                upperBinIndex: this.value.upperBinIndex,
                amountXtoDeposit: this.value.amountXtoDeposit.toString(),
                amountYtoDeposit: this.value.amountYtoDeposit.toString(),
                xCurrentBin: this.value.xCurrentBin.toString(),
                yCurrentBin: this.value.yCurrentBin.toString(),
            },
        };
    }
    toEncodable() {
        return {
            Uniform: {
                currentBinIndex: this.value.currentBinIndex,
                lowerBinIndex: this.value.lowerBinIndex,
                upperBinIndex: this.value.upperBinIndex,
                amountXToDeposit: this.value.amountXtoDeposit,
                amountYToDeposit: this.value.amountYtoDeposit,
                xCurrentBin: this.value.xCurrentBin,
                yCurrentBin: this.value.yCurrentBin,
            },
        };
    }
}
exports.Uniform = Uniform;
class CurrentTick {
    static discriminator = 1;
    static kind = "CurrentTick";
    discriminator = 1;
    kind = "CurrentTick";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "CurrentTick",
            value: [this.value[0]],
        };
    }
    toEncodable() {
        return {
            CurrentTick: {
                _0: this.value[0],
            },
        };
    }
}
exports.CurrentTick = CurrentTick;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Uniform" in obj) {
        const val = obj["Uniform"];
        return new Uniform({
            currentBinIndex: val["currentBinIndex"],
            lowerBinIndex: val["lowerBinIndex"],
            upperBinIndex: val["upperBinIndex"],
            amountXtoDeposit: val["amountXToDeposit"],
            amountYtoDeposit: val["amountYToDeposit"],
            xCurrentBin: val["xCurrentBin"],
            yCurrentBin: val["yCurrentBin"],
        });
    }
    if ("CurrentTick" in obj) {
        const val = obj["CurrentTick"];
        return new CurrentTick([val["_0"]]);
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Uniform": {
            return new Uniform({
                currentBinIndex: obj.value.currentBinIndex,
                lowerBinIndex: obj.value.lowerBinIndex,
                upperBinIndex: obj.value.upperBinIndex,
                amountXtoDeposit: new bn_js_1.default(obj.value.amountXtoDeposit),
                amountYtoDeposit: new bn_js_1.default(obj.value.amountYtoDeposit),
                xCurrentBin: new bn_js_1.default(obj.value.xCurrentBin),
                yCurrentBin: new bn_js_1.default(obj.value.yCurrentBin),
            });
        }
        case "CurrentTick": {
            return new CurrentTick([obj.value[0]]);
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([
            borsh.i32("currentBinIndex"),
            borsh.i32("lowerBinIndex"),
            borsh.i32("upperBinIndex"),
            borsh.u64("amountXToDeposit"),
            borsh.u64("amountYToDeposit"),
            borsh.u64("xCurrentBin"),
            borsh.u64("yCurrentBin"),
        ], "Uniform"),
        borsh.struct([borsh.i32("_0")], "CurrentTick"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
