// @mrgnlabs/marginfi-client-v2/dist/models/emode-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmodeSettings = void 0;
const services_1 = require("../services");
const mrgn_common_1 = require("@mrgnlabs/mrgn-common");
class EmodeSettings {
    constructor(emodeTag, timestamp, flags, emodeEntries) {
        this.emodeTag = emodeTag;
        this.timestamp = timestamp;
        this.flags = flags;
        this.emodeEntries = emodeEntries;
        this.emodeTag = emodeTag;
        this.timestamp = timestamp;
        this.flags = flags;
        this.emodeEntries = emodeEntries;
    }
    static from(emodeSettingsRaw) {
        const emodeTag = (0, services_1.parseEmodeTag)(emodeSettingsRaw.emodeTag);
        const timestamp = emodeSettingsRaw.timestamp.toNumber();
        const flags = (0, services_1.getActiveEmodeFlags)(emodeSettingsRaw.flags);
        const emodeEntries = emodeSettingsRaw.emodeConfig.entries
            .filter((entry) => entry.collateralBankEmodeTag !== 0)
            .map((entry) => {
            return {
                collateralBankEmodeTag: (0, services_1.parseEmodeTag)(entry.collateralBankEmodeTag),
                flags: (0, services_1.getActiveEmodeEntryFlags)(entry.flags),
                assetWeightInit: (0, mrgn_common_1.wrappedI80F48toBigNumber)(entry.assetWeightInit),
                assetWeightMaint: (0, mrgn_common_1.wrappedI80F48toBigNumber)(entry.assetWeightMaint),
            };
        });
        return new EmodeSettings(emodeTag, timestamp, flags, emodeEntries);
    }
}
exports.EmodeSettings = EmodeSettings;
