// @mrgnlabs/marginfi-client-v2/dist/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_ORACLE_MAX_AGE = exports.PYTH_PUSH_ORACLE_ID = exports.TRANSFER_ACCOUNT_AUTHORITY_FLAG = exports.FLASHLOAN_ENABLED_FLAG = exports.DISABLED_FLAG = exports.ADDRESS_LOOKUP_TABLE_FOR_GROUP = exports.USDC_DECIMALS = exports.MAX_CONFIDENCE_INTERVAL_RATIO = exports.SWB_PRICE_CONF_INTERVALS = exports.PYTH_PRICE_CONF_INTERVALS = exports.PDA_BANK_FEE_VAULT_SEED = exports.PDA_BANK_INSURANCE_VAULT_SEED = exports.PDA_BANK_LIQUIDITY_VAULT_SEED = exports.PDA_BANK_FEE_VAULT_AUTH_SEED = exports.PDA_BANK_INSURANCE_VAULT_AUTH_SEED = exports.PDA_BANK_LIQUIDITY_VAULT_AUTH_SEED = void 0;
const web3_js_1 = require("@solana/web3.js");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
exports.PDA_BANK_LIQUIDITY_VAULT_AUTH_SEED = Buffer.from("liquidity_vault_auth");
exports.PDA_BANK_INSURANCE_VAULT_AUTH_SEED = Buffer.from("insurance_vault_auth");
exports.PDA_BANK_FEE_VAULT_AUTH_SEED = Buffer.from("fee_vault_auth");
exports.PDA_BANK_LIQUIDITY_VAULT_SEED = Buffer.from("liquidity_vault");
exports.PDA_BANK_INSURANCE_VAULT_SEED = Buffer.from("insurance_vault");
exports.PDA_BANK_FEE_VAULT_SEED = Buffer.from("fee_vault");
exports.PYTH_PRICE_CONF_INTERVALS = new bignumber_js_1.default(2.12);
exports.SWB_PRICE_CONF_INTERVALS = new bignumber_js_1.default(1.96);
exports.MAX_CONFIDENCE_INTERVAL_RATIO = new bignumber_js_1.default(0.05);
exports.USDC_DECIMALS = 6;
exports.ADDRESS_LOOKUP_TABLE_FOR_GROUP = {
    "4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8": [
        new web3_js_1.PublicKey("BrWF8J3CEuHaXsWk3kqGZ6VHvRp4SJuG9AzvB6ei2kbV"),
        new web3_js_1.PublicKey("8GLUprtyzv6HGrgox7F43EQM5GqE2uKrAHLs69r8DgRj"),
    ], // Main pool
    FCPfpHA69EbS8f9KKSreTRkXbzFpunsKuYf5qNmnJjpo: [new web3_js_1.PublicKey("HxPy7b58KLKSU7w4LUW9xwYQ1NPyRNQkYYk2f7SmYAip")], // staging
};
exports.DISABLED_FLAG = 1 << 0;
exports.FLASHLOAN_ENABLED_FLAG = 1 << 2;
exports.TRANSFER_ACCOUNT_AUTHORITY_FLAG = 1 << 3;
exports.PYTH_PUSH_ORACLE_ID = new web3_js_1.PublicKey("pythWSnswVUd12oZpeFP8e9CVaEqJg25g1Vtc2biRsT");
exports.DEFAULT_ORACLE_MAX_AGE = 60; // seconds
