// @mrgnlabs/mrgn-common/dist/constants/program.consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STAKE_CONFIG_ID = exports.SYSVAR_STAKE_HISTORY_ID = exports.SYSVAR_CLOCK_ID = exports.SYSVAR_RENT_ID = exports.SYSTEM_PROGRAM_ID = exports.MPL_METADATA_PROGRAM_ID = exports.STAKE_PROGRAM_ID = exports.SINGLE_POOL_PROGRAM_ID = exports.MARGINFI_PROGRAM = exports.JUPITER_V6_PROGRAM = exports.LUT_PROGRAM_AUTHORITY_INDEX = void 0;
const web3_js_1 = require("@solana/web3.js");
exports.LUT_PROGRAM_AUTHORITY_INDEX = 5;
exports.JUPITER_V6_PROGRAM = new web3_js_1.PublicKey("JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4");
exports.MARGINFI_PROGRAM = new web3_js_1.PublicKey("MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA");
exports.SINGLE_POOL_PROGRAM_ID = new web3_js_1.PublicKey("SVSPxpvHdN29nkVg9rPapPNDddN5DipNLRUFhyjFThE");
exports.STAKE_PROGRAM_ID = new web3_js_1.PublicKey("Stake11111111111111111111111111111111111111");
exports.MPL_METADATA_PROGRAM_ID = new web3_js_1.PublicKey("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s");
exports.SYSTEM_PROGRAM_ID = new web3_js_1.PublicKey("11111111111111111111111111111111");
exports.SYSVAR_RENT_ID = new web3_js_1.PublicKey("SysvarRent111111111111111111111111111111111");
exports.SYSVAR_CLOCK_ID = new web3_js_1.PublicKey("SysvarC1ock11111111111111111111111111111111");
exports.SYSVAR_STAKE_HISTORY_ID = new web3_js_1.PublicKey("SysvarStakeHistory1111111111111111111111111");
exports.STAKE_CONFIG_ID = new web3_js_1.PublicKey("StakeConfig11111111111111111111111111111111");
