// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/accounts/Position.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Position = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
/* eslint-enable @typescript-eslint/no-unused-vars */
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const utils_1 = require("../utils"); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class Position {
    whirlpool;
    positionMint;
    liquidity;
    tickLowerIndex;
    tickUpperIndex;
    feeGrowthCheckpointA;
    feeOwedA;
    feeGrowthCheckpointB;
    feeOwedB;
    rewardInfos;
    static discriminator = Buffer.from([
        170, 188, 143, 228, 122, 64, 247, 208,
    ]);
    static layout = borsh.struct([
        (0, utils_1.borshAddress)("whirlpool"),
        (0, utils_1.borshAddress)("positionMint"),
        borsh.u128("liquidity"),
        borsh.i32("tickLowerIndex"),
        borsh.i32("tickUpperIndex"),
        borsh.u128("feeGrowthCheckpointA"),
        borsh.u64("feeOwedA"),
        borsh.u128("feeGrowthCheckpointB"),
        borsh.u64("feeOwedB"),
        borsh.array(types.PositionRewardInfo.layout(), 3, "rewardInfos"),
    ]);
    constructor(fields) {
        this.whirlpool = fields.whirlpool;
        this.positionMint = fields.positionMint;
        this.liquidity = fields.liquidity;
        this.tickLowerIndex = fields.tickLowerIndex;
        this.tickUpperIndex = fields.tickUpperIndex;
        this.feeGrowthCheckpointA = fields.feeGrowthCheckpointA;
        this.feeOwedA = fields.feeOwedA;
        this.feeGrowthCheckpointB = fields.feeGrowthCheckpointB;
        this.feeOwedB = fields.feeOwedB;
        this.rewardInfos = fields.rewardInfos.map((item) => new types.PositionRewardInfo({ ...item }));
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error(`PositionFields account ${address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error(`PositionFields account ${info.address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(Position.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = Position.layout.decode(data.slice(8));
        return new Position({
            whirlpool: dec.whirlpool,
            positionMint: dec.positionMint,
            liquidity: dec.liquidity,
            tickLowerIndex: dec.tickLowerIndex,
            tickUpperIndex: dec.tickUpperIndex,
            feeGrowthCheckpointA: dec.feeGrowthCheckpointA,
            feeOwedA: dec.feeOwedA,
            feeGrowthCheckpointB: dec.feeGrowthCheckpointB,
            feeOwedB: dec.feeOwedB,
            rewardInfos: dec.rewardInfos.map((item /* eslint-disable-line @typescript-eslint/no-explicit-any */) => types.PositionRewardInfo.fromDecoded(item)),
        });
    }
    toJSON() {
        return {
            whirlpool: this.whirlpool,
            positionMint: this.positionMint,
            liquidity: this.liquidity.toString(),
            tickLowerIndex: this.tickLowerIndex,
            tickUpperIndex: this.tickUpperIndex,
            feeGrowthCheckpointA: this.feeGrowthCheckpointA.toString(),
            feeOwedA: this.feeOwedA.toString(),
            feeGrowthCheckpointB: this.feeGrowthCheckpointB.toString(),
            feeOwedB: this.feeOwedB.toString(),
            rewardInfos: this.rewardInfos.map((item) => item.toJSON()),
        };
    }
    static fromJSON(obj) {
        return new Position({
            whirlpool: (0, kit_1.address)(obj.whirlpool),
            positionMint: (0, kit_1.address)(obj.positionMint),
            liquidity: new bn_js_1.default(obj.liquidity),
            tickLowerIndex: obj.tickLowerIndex,
            tickUpperIndex: obj.tickUpperIndex,
            feeGrowthCheckpointA: new bn_js_1.default(obj.feeGrowthCheckpointA),
            feeOwedA: new bn_js_1.default(obj.feeOwedA),
            feeGrowthCheckpointB: new bn_js_1.default(obj.feeGrowthCheckpointB),
            feeOwedB: new bn_js_1.default(obj.feeOwedB),
            rewardInfos: obj.rewardInfos.map((item) => types.PositionRewardInfo.fromJSON(item)),
        });
    }
}
exports.Position = Position;
