// @kamino-finance/kliquidity-sdk/dist/@codegen/whirlpools/accounts/TickArray.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TickArray = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const utils_1 = require("../utils"); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
class TickArray {
    startTickIndex;
    ticks;
    whirlpool;
    static discriminator = Buffer.from([
        69, 97, 189, 190, 110, 7, 66, 187,
    ]);
    static layout = borsh.struct([
        borsh.i32("startTickIndex"),
        borsh.array(types.Tick.layout(), 88, "ticks"),
        (0, utils_1.borshAddress)("whirlpool"),
    ]);
    constructor(fields) {
        this.startTickIndex = fields.startTickIndex;
        this.ticks = fields.ticks.map((item) => new types.Tick({ ...item }));
        this.whirlpool = fields.whirlpool;
    }
    static async fetch(rpc, address, programId = programId_1.PROGRAM_ID) {
        const info = await (0, kit_1.fetchEncodedAccount)(rpc, address);
        if (!info.exists) {
            return null;
        }
        if (info.programAddress !== programId) {
            throw new Error(`TickArrayFields account ${address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
        }
        return this.decode(Buffer.from(info.data));
    }
    static async fetchMultiple(rpc, addresses, programId = programId_1.PROGRAM_ID) {
        const infos = await (0, kit_1.fetchEncodedAccounts)(rpc, addresses);
        return infos.map((info) => {
            if (!info.exists) {
                return null;
            }
            if (info.programAddress !== programId) {
                throw new Error(`TickArrayFields account ${info.address} belongs to wrong program ${info.programAddress}, expected ${programId}`);
            }
            return this.decode(Buffer.from(info.data));
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(TickArray.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = TickArray.layout.decode(data.slice(8));
        return new TickArray({
            startTickIndex: dec.startTickIndex,
            ticks: dec.ticks.map((item /* eslint-disable-line @typescript-eslint/no-explicit-any */) => types.Tick.fromDecoded(item)),
            whirlpool: dec.whirlpool,
        });
    }
    toJSON() {
        return {
            startTickIndex: this.startTickIndex,
            ticks: this.ticks.map((item) => item.toJSON()),
            whirlpool: this.whirlpool,
        };
    }
    static fromJSON(obj) {
        return new TickArray({
            startTickIndex: obj.startTickIndex,
            ticks: obj.ticks.map((item) => types.Tick.fromJSON(item)),
            whirlpool: (0, kit_1.address)(obj.whirlpool),
        });
    }
}
exports.TickArray = TickArray;
