// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/DEX.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Meteora = exports.Raydium = exports.Orca = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Orca {
    static discriminator = 0;
    static kind = "Orca";
    discriminator = 0;
    kind = "Orca";
    toJSON() {
        return {
            kind: "Orca",
        };
    }
    toEncodable() {
        return {
            Orca: {},
        };
    }
}
exports.Orca = Orca;
class Raydium {
    static discriminator = 1;
    static kind = "Raydium";
    discriminator = 1;
    kind = "Raydium";
    toJSON() {
        return {
            kind: "Raydium",
        };
    }
    toEncodable() {
        return {
            Raydium: {},
        };
    }
}
exports.Raydium = Raydium;
class Meteora {
    static discriminator = 2;
    static kind = "Meteora";
    discriminator = 2;
    kind = "Meteora";
    toJSON() {
        return {
            kind: "Meteora",
        };
    }
    toEncodable() {
        return {
            Meteora: {},
        };
    }
}
exports.Meteora = Meteora;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Orca" in obj) {
        return new Orca();
    }
    if ("Raydium" in obj) {
        return new Raydium();
    }
    if ("Meteora" in obj) {
        return new Meteora();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Orca": {
            return new Orca();
        }
        case "Raydium": {
            return new Raydium();
        }
        case "Meteora": {
            return new Meteora();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Orca"),
        borsh.struct([], "Raydium"),
        borsh.struct([], "Meteora"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
