// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEX = exports.ScopePriceIdTest = exports.CollateralTestToken = exports.RebalanceType = exports.RebalanceAction = exports.RebalanceTakeProfitStep = exports.RebalanceTakeProfitToken = exports.ExpanderStep = exports.RebalanceDriftStep = exports.DriftDirection = exports.StakingRateSource = exports.RebalanceAutodriftStep = exports.BalanceStatus = exports.UpdateCollateralInfoMode = exports.LiquidityCalculationMode = exports.ReferencePriceType = exports.ExecutiveWithdrawAction = exports.CreationStatus = exports.StrategyType = exports.StrategyStatus = exports.StrategyConfigOption = exports.GlobalConfigOption = exports.MintingMethod = exports.SwapLimit = exports.RebalanceEffects = exports.WithdrawalCapAccumulatorAction = exports.RemoveLiquidityMode = exports.DexSpecificPrice = exports.SimulationPrice = exports.BinAddLiquidityStrategy = exports.RebalanceTakeProfitState = exports.RebalancePricePercentageState = exports.RebalancePricePercentageWithResetState = exports.PeriodicRebalanceState = exports.RebalanceManualState = exports.RebalanceExpanderState = exports.RebalanceDriftState = exports.RebalanceDriftParams = exports.RebalanceAutodriftState = exports.RebalanceAutodriftWindow = exports.RebalanceAutodriftParams = exports.Price = exports.WithdrawalCaps = exports.KaminoRewardInfo = exports.CollateralInfoParams = exports.CollateralInfo = exports.RebalanceRaw = exports.RewardInfo = exports.WhirlpoolRewardInfo = exports.PositionRewardInfo = void 0;
const BinAddLiquidityStrategy = __importStar(require("./BinAddLiquidityStrategy"));
exports.BinAddLiquidityStrategy = BinAddLiquidityStrategy;
const SimulationPrice = __importStar(require("./SimulationPrice"));
exports.SimulationPrice = SimulationPrice;
const DexSpecificPrice = __importStar(require("./DexSpecificPrice"));
exports.DexSpecificPrice = DexSpecificPrice;
const RemoveLiquidityMode = __importStar(require("./RemoveLiquidityMode"));
exports.RemoveLiquidityMode = RemoveLiquidityMode;
const WithdrawalCapAccumulatorAction = __importStar(require("./WithdrawalCapAccumulatorAction"));
exports.WithdrawalCapAccumulatorAction = WithdrawalCapAccumulatorAction;
const RebalanceEffects = __importStar(require("./RebalanceEffects"));
exports.RebalanceEffects = RebalanceEffects;
const SwapLimit = __importStar(require("./SwapLimit"));
exports.SwapLimit = SwapLimit;
const MintingMethod = __importStar(require("./MintingMethod"));
exports.MintingMethod = MintingMethod;
const GlobalConfigOption = __importStar(require("./GlobalConfigOption"));
exports.GlobalConfigOption = GlobalConfigOption;
const StrategyConfigOption = __importStar(require("./StrategyConfigOption"));
exports.StrategyConfigOption = StrategyConfigOption;
const StrategyStatus = __importStar(require("./StrategyStatus"));
exports.StrategyStatus = StrategyStatus;
const StrategyType = __importStar(require("./StrategyType"));
exports.StrategyType = StrategyType;
const CreationStatus = __importStar(require("./CreationStatus"));
exports.CreationStatus = CreationStatus;
const ExecutiveWithdrawAction = __importStar(require("./ExecutiveWithdrawAction"));
exports.ExecutiveWithdrawAction = ExecutiveWithdrawAction;
const ReferencePriceType = __importStar(require("./ReferencePriceType"));
exports.ReferencePriceType = ReferencePriceType;
const LiquidityCalculationMode = __importStar(require("./LiquidityCalculationMode"));
exports.LiquidityCalculationMode = LiquidityCalculationMode;
const UpdateCollateralInfoMode = __importStar(require("./UpdateCollateralInfoMode"));
exports.UpdateCollateralInfoMode = UpdateCollateralInfoMode;
const BalanceStatus = __importStar(require("./BalanceStatus"));
exports.BalanceStatus = BalanceStatus;
const RebalanceAutodriftStep = __importStar(require("./RebalanceAutodriftStep"));
exports.RebalanceAutodriftStep = RebalanceAutodriftStep;
const StakingRateSource = __importStar(require("./StakingRateSource"));
exports.StakingRateSource = StakingRateSource;
const DriftDirection = __importStar(require("./DriftDirection"));
exports.DriftDirection = DriftDirection;
const RebalanceDriftStep = __importStar(require("./RebalanceDriftStep"));
exports.RebalanceDriftStep = RebalanceDriftStep;
const ExpanderStep = __importStar(require("./ExpanderStep"));
exports.ExpanderStep = ExpanderStep;
const RebalanceTakeProfitToken = __importStar(require("./RebalanceTakeProfitToken"));
exports.RebalanceTakeProfitToken = RebalanceTakeProfitToken;
const RebalanceTakeProfitStep = __importStar(require("./RebalanceTakeProfitStep"));
exports.RebalanceTakeProfitStep = RebalanceTakeProfitStep;
const RebalanceAction = __importStar(require("./RebalanceAction"));
exports.RebalanceAction = RebalanceAction;
const RebalanceType = __importStar(require("./RebalanceType"));
exports.RebalanceType = RebalanceType;
const CollateralTestToken = __importStar(require("./CollateralTestToken"));
exports.CollateralTestToken = CollateralTestToken;
const ScopePriceIdTest = __importStar(require("./ScopePriceIdTest"));
exports.ScopePriceIdTest = ScopePriceIdTest;
const DEX = __importStar(require("./DEX"));
exports.DEX = DEX;
var PositionRewardInfo_1 = require("./PositionRewardInfo");
Object.defineProperty(exports, "PositionRewardInfo", { enumerable: true, get: function () { return PositionRewardInfo_1.PositionRewardInfo; } });
var WhirlpoolRewardInfo_1 = require("./WhirlpoolRewardInfo");
Object.defineProperty(exports, "WhirlpoolRewardInfo", { enumerable: true, get: function () { return WhirlpoolRewardInfo_1.WhirlpoolRewardInfo; } });
var RewardInfo_1 = require("./RewardInfo");
Object.defineProperty(exports, "RewardInfo", { enumerable: true, get: function () { return RewardInfo_1.RewardInfo; } });
var RebalanceRaw_1 = require("./RebalanceRaw");
Object.defineProperty(exports, "RebalanceRaw", { enumerable: true, get: function () { return RebalanceRaw_1.RebalanceRaw; } });
var CollateralInfo_1 = require("./CollateralInfo");
Object.defineProperty(exports, "CollateralInfo", { enumerable: true, get: function () { return CollateralInfo_1.CollateralInfo; } });
var CollateralInfoParams_1 = require("./CollateralInfoParams");
Object.defineProperty(exports, "CollateralInfoParams", { enumerable: true, get: function () { return CollateralInfoParams_1.CollateralInfoParams; } });
var KaminoRewardInfo_1 = require("./KaminoRewardInfo");
Object.defineProperty(exports, "KaminoRewardInfo", { enumerable: true, get: function () { return KaminoRewardInfo_1.KaminoRewardInfo; } });
var WithdrawalCaps_1 = require("./WithdrawalCaps");
Object.defineProperty(exports, "WithdrawalCaps", { enumerable: true, get: function () { return WithdrawalCaps_1.WithdrawalCaps; } });
var Price_1 = require("./Price");
Object.defineProperty(exports, "Price", { enumerable: true, get: function () { return Price_1.Price; } });
var RebalanceAutodriftParams_1 = require("./RebalanceAutodriftParams");
Object.defineProperty(exports, "RebalanceAutodriftParams", { enumerable: true, get: function () { return RebalanceAutodriftParams_1.RebalanceAutodriftParams; } });
var RebalanceAutodriftWindow_1 = require("./RebalanceAutodriftWindow");
Object.defineProperty(exports, "RebalanceAutodriftWindow", { enumerable: true, get: function () { return RebalanceAutodriftWindow_1.RebalanceAutodriftWindow; } });
var RebalanceAutodriftState_1 = require("./RebalanceAutodriftState");
Object.defineProperty(exports, "RebalanceAutodriftState", { enumerable: true, get: function () { return RebalanceAutodriftState_1.RebalanceAutodriftState; } });
var RebalanceDriftParams_1 = require("./RebalanceDriftParams");
Object.defineProperty(exports, "RebalanceDriftParams", { enumerable: true, get: function () { return RebalanceDriftParams_1.RebalanceDriftParams; } });
var RebalanceDriftState_1 = require("./RebalanceDriftState");
Object.defineProperty(exports, "RebalanceDriftState", { enumerable: true, get: function () { return RebalanceDriftState_1.RebalanceDriftState; } });
var RebalanceExpanderState_1 = require("./RebalanceExpanderState");
Object.defineProperty(exports, "RebalanceExpanderState", { enumerable: true, get: function () { return RebalanceExpanderState_1.RebalanceExpanderState; } });
var RebalanceManualState_1 = require("./RebalanceManualState");
Object.defineProperty(exports, "RebalanceManualState", { enumerable: true, get: function () { return RebalanceManualState_1.RebalanceManualState; } });
var PeriodicRebalanceState_1 = require("./PeriodicRebalanceState");
Object.defineProperty(exports, "PeriodicRebalanceState", { enumerable: true, get: function () { return PeriodicRebalanceState_1.PeriodicRebalanceState; } });
var RebalancePricePercentageWithResetState_1 = require("./RebalancePricePercentageWithResetState");
Object.defineProperty(exports, "RebalancePricePercentageWithResetState", { enumerable: true, get: function () { return RebalancePricePercentageWithResetState_1.RebalancePricePercentageWithResetState; } });
var RebalancePricePercentageState_1 = require("./RebalancePricePercentageState");
Object.defineProperty(exports, "RebalancePricePercentageState", { enumerable: true, get: function () { return RebalancePricePercentageState_1.RebalancePricePercentageState; } });
var RebalanceTakeProfitState_1 = require("./RebalanceTakeProfitState");
Object.defineProperty(exports, "RebalanceTakeProfitState", { enumerable: true, get: function () { return RebalanceTakeProfitState_1.RebalanceTakeProfitState; } });
