// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/RebalanceAutodriftWindow.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RebalanceAutodriftWindow = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class RebalanceAutodriftWindow {
    stakingRateA;
    stakingRateB;
    epoch;
    theoreticalTick;
    stratMidTick;
    constructor(fields) {
        this.stakingRateA =
            (fields.stakingRateA && new types.Price({ ...fields.stakingRateA })) ||
                null;
        this.stakingRateB =
            (fields.stakingRateB && new types.Price({ ...fields.stakingRateB })) ||
                null;
        this.epoch = fields.epoch;
        this.theoreticalTick = fields.theoreticalTick;
        this.stratMidTick = fields.stratMidTick;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(types.Price.layout(), "stakingRateA"),
            borsh.option(types.Price.layout(), "stakingRateB"),
            borsh.u64("epoch"),
            borsh.i32("theoreticalTick"),
            borsh.i32("stratMidTick"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new RebalanceAutodriftWindow({
            stakingRateA: (obj.stakingRateA && types.Price.fromDecoded(obj.stakingRateA)) || null,
            stakingRateB: (obj.stakingRateB && types.Price.fromDecoded(obj.stakingRateB)) || null,
            epoch: obj.epoch,
            theoreticalTick: obj.theoreticalTick,
            stratMidTick: obj.stratMidTick,
        });
    }
    static toEncodable(fields) {
        return {
            stakingRateA: (fields.stakingRateA && types.Price.toEncodable(fields.stakingRateA)) ||
                null,
            stakingRateB: (fields.stakingRateB && types.Price.toEncodable(fields.stakingRateB)) ||
                null,
            epoch: fields.epoch,
            theoreticalTick: fields.theoreticalTick,
            stratMidTick: fields.stratMidTick,
        };
    }
    toJSON() {
        return {
            stakingRateA: (this.stakingRateA && this.stakingRateA.toJSON()) || null,
            stakingRateB: (this.stakingRateB && this.stakingRateB.toJSON()) || null,
            epoch: this.epoch.toString(),
            theoreticalTick: this.theoreticalTick,
            stratMidTick: this.stratMidTick,
        };
    }
    static fromJSON(obj) {
        return new RebalanceAutodriftWindow({
            stakingRateA: (obj.stakingRateA && types.Price.fromJSON(obj.stakingRateA)) || null,
            stakingRateB: (obj.stakingRateB && types.Price.fromJSON(obj.stakingRateB)) || null,
            epoch: new bn_js_1.default(obj.epoch),
            theoreticalTick: obj.theoreticalTick,
            stratMidTick: obj.stratMidTick,
        });
    }
    toEncodable() {
        return RebalanceAutodriftWindow.toEncodable(this);
    }
}
exports.RebalanceAutodriftWindow = RebalanceAutodriftWindow;
