// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/RebalanceRaw.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RebalanceRaw = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class RebalanceRaw {
    params;
    state;
    referencePriceType;
    constructor(fields) {
        this.params = fields.params;
        this.state = fields.state;
        this.referencePriceType = fields.referencePriceType;
    }
    static layout(property) {
        return borsh.struct([
            borsh.array(borsh.u8(), 128, "params"),
            borsh.array(borsh.u8(), 256, "state"),
            borsh.u8("referencePriceType"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new RebalanceRaw({
            params: obj.params,
            state: obj.state,
            referencePriceType: obj.referencePriceType,
        });
    }
    static toEncodable(fields) {
        return {
            params: fields.params,
            state: fields.state,
            referencePriceType: fields.referencePriceType,
        };
    }
    toJSON() {
        return {
            params: this.params,
            state: this.state,
            referencePriceType: this.referencePriceType,
        };
    }
    static fromJSON(obj) {
        return new RebalanceRaw({
            params: obj.params,
            state: obj.state,
            referencePriceType: obj.referencePriceType,
        });
    }
    toEncodable() {
        return RebalanceRaw.toEncodable(this);
    }
}
exports.RebalanceRaw = RebalanceRaw;
