// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/RebalanceEffects.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawAndFreeze = exports.NewRange = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class NewRange {
    static discriminator = 0;
    static kind = "NewRange";
    discriminator = 0;
    kind = "NewRange";
    value;
    constructor(value) {
        this.value = [value[0], value[1]];
    }
    toJSON() {
        return {
            kind: "NewRange",
            value: [this.value[0], this.value[1]],
        };
    }
    toEncodable() {
        return {
            NewRange: {
                _0: this.value[0],
                _1: this.value[1],
            },
        };
    }
}
exports.NewRange = NewRange;
class WithdrawAndFreeze {
    static discriminator = 1;
    static kind = "WithdrawAndFreeze";
    discriminator = 1;
    kind = "WithdrawAndFreeze";
    toJSON() {
        return {
            kind: "WithdrawAndFreeze",
        };
    }
    toEncodable() {
        return {
            WithdrawAndFreeze: {},
        };
    }
}
exports.WithdrawAndFreeze = WithdrawAndFreeze;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("NewRange" in obj) {
        const val = obj["NewRange"];
        return new NewRange([val["_0"], val["_1"]]);
    }
    if ("WithdrawAndFreeze" in obj) {
        return new WithdrawAndFreeze();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "NewRange": {
            return new NewRange([obj.value[0], obj.value[1]]);
        }
        case "WithdrawAndFreeze": {
            return new WithdrawAndFreeze();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([borsh.i32("_0"), borsh.i32("_1")], "NewRange"),
        borsh.struct([], "WithdrawAndFreeze"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
