// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/types/StakingRateSource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scope = exports.Constant = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Constant {
    static discriminator = 0;
    static kind = "Constant";
    discriminator = 0;
    kind = "Constant";
    toJSON() {
        return {
            kind: "Constant",
        };
    }
    toEncodable() {
        return {
            Constant: {},
        };
    }
}
exports.Constant = Constant;
class Scope {
    static discriminator = 1;
    static kind = "Scope";
    discriminator = 1;
    kind = "Scope";
    toJSON() {
        return {
            kind: "Scope",
        };
    }
    toEncodable() {
        return {
            Scope: {},
        };
    }
}
exports.Scope = Scope;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Constant" in obj) {
        return new Constant();
    }
    if ("Scope" in obj) {
        return new Scope();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Constant": {
            return new Constant();
        }
        case "Scope": {
            return new Scope();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Constant"),
        borsh.struct([], "Scope"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
