// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/instructions/openPosition.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.openPosition = openPosition;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.i32("tickLowerIndex"),
    borsh.i32("tickUpperIndex"),
    borsh.i32("tickArrayLowerStartIndex"),
    borsh.i32("tickArrayUpperStartIndex"),
    borsh.u128("liquidity"),
    borsh.u64("amount0Max"),
    borsh.u64("amount1Max"),
]);
function openPosition(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.payer.address, role: 3, signer: accounts.payer },
        { address: accounts.positionNftOwner, role: 0 },
        {
            address: accounts.positionNftMint.address,
            role: 3,
            signer: accounts.positionNftMint,
        },
        { address: accounts.positionNftAccount, role: 1 },
        { address: accounts.metadataAccount, role: 1 },
        { address: accounts.poolState, role: 1 },
        { address: accounts.protocolPosition, role: 1 },
        { address: accounts.tickArrayLower, role: 1 },
        { address: accounts.tickArrayUpper, role: 1 },
        { address: accounts.personalPosition, role: 1 },
        { address: accounts.tokenAccount0, role: 1 },
        { address: accounts.tokenAccount1, role: 1 },
        { address: accounts.tokenVault0, role: 1 },
        { address: accounts.tokenVault1, role: 1 },
        { address: accounts.rent, role: 0 },
        { address: accounts.systemProgram, role: 0 },
        { address: accounts.tokenProgram, role: 0 },
        { address: accounts.associatedTokenProgram, role: 0 },
        { address: accounts.metadataProgram, role: 0 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([135, 128, 47, 77, 15, 152, 240, 49]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        tickLowerIndex: args.tickLowerIndex,
        tickUpperIndex: args.tickUpperIndex,
        tickArrayLowerStartIndex: args.tickArrayLowerStartIndex,
        tickArrayUpperStartIndex: args.tickArrayUpperStartIndex,
        liquidity: args.liquidity,
        amount0Max: args.amount0Max,
        amount1Max: args.amount1Max,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
