// @kamino-finance/kliquidity-sdk/dist/@codegen/raydium/instructions/swap.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.swap = swap;
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([
    borsh.u64("amount"),
    borsh.u64("otherAmountThreshold"),
    borsh.u128("sqrtPriceLimitX64"),
    borsh.bool("isBaseInput"),
]);
function swap(args, accounts, remainingAccounts = [], programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        { address: accounts.payer.address, role: 2, signer: accounts.payer },
        { address: accounts.ammConfig, role: 0 },
        { address: accounts.poolState, role: 1 },
        { address: accounts.inputTokenAccount, role: 1 },
        { address: accounts.outputTokenAccount, role: 1 },
        { address: accounts.inputVault, role: 1 },
        { address: accounts.outputVault, role: 1 },
        { address: accounts.observationState, role: 1 },
        { address: accounts.tokenProgram, role: 0 },
        { address: accounts.tickArray, role: 1 },
        ...remainingAccounts,
    ];
    const identifier = Buffer.from([248, 198, 158, 145, 225, 117, 135, 200]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        amount: args.amount,
        otherAmountThreshold: args.otherAmountThreshold,
        sqrtPriceLimitX64: args.sqrtPriceLimitX64,
        isBaseInput: args.isBaseInput,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
