// @kamino-finance/scope-sdk/dist/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SCOPE_LOCALNET_CONFIG = exports.SCOPE_DEVNET_CONFIG = exports.SCOPE_MAINNET_CONFIG = exports.U16_MAX = void 0;
const kit_1 = require("@solana/kit");
exports.U16_MAX = 2 ** 16 - 1;
exports.SCOPE_MAINNET_CONFIG = {
    oracleMappings: (0, kit_1.address)('Chpu5ZgfWX5ZzVpUx9Xvv4WPM75Xd7zPJNDPsFnCpLpk'),
    oraclePrices: (0, kit_1.address)('3NJYftD5sjVfxSnUdZ1wVML8f3aC6mp1CXCL6L7TnU8C'),
    programId: (0, kit_1.address)('HFn8GnPADiny6XqUoWE8uRPPxb29ikn4yTuPa9MF2fWJ'),
    configurationAccount: (0, kit_1.address)('AdTiP7QyjUyv6crF4H8z7fxJKU7Z5eCAGvJN1Y55cXxb'),
    kliquidityProgramId: (0, kit_1.address)('6LtLpnUFNByNXLyCoK9wA2MykKAmQNZKBdY8s47dehDc'),
};
exports.SCOPE_DEVNET_CONFIG = {
    ...exports.SCOPE_MAINNET_CONFIG,
    kliquidityProgramId: (0, kit_1.address)('E6qbhrt4pFmCotNUSSEh6E5cRQCEJpMcd79Z56EG9KY'),
};
exports.SCOPE_LOCALNET_CONFIG = {
    ...exports.SCOPE_MAINNET_CONFIG,
    kliquidityProgramId: (0, kit_1.address)('E6qbhrt4pFmCotNUSSEh6E5cRQCEJpMcd79Z56EG9KY'),
};
