// @kamino-finance/scope-sdk/dist/@codegen/kliquidity/instructions/swapUnevenVaults.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = void 0;
exports.swapUnevenVaults = swapUnevenVaults;
/* eslint-disable @typescript-eslint/no-unused-vars */
const kit_1 = require("@solana/kit");
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
const programId_1 = require("../programId");
exports.layout = borsh.struct([borsh.u64("targetLimitBps")]);
function swapUnevenVaults(args, accounts, programAddress = programId_1.PROGRAM_ID) {
    const keys = [
        {
            address: accounts.actionsAuthority.address,
            role: 3,
            signer: accounts.actionsAuthority,
        },
        { address: accounts.strategy, role: 1 },
        { address: accounts.globalConfig, role: 0 },
        { address: accounts.tokenAVault, role: 1 },
        { address: accounts.tokenBVault, role: 1 },
        { address: accounts.tokenAMint, role: 0 },
        { address: accounts.tokenBMint, role: 0 },
        { address: accounts.baseVaultAuthority, role: 1 },
        { address: accounts.pool, role: 1 },
        { address: accounts.position, role: 1 },
        { address: accounts.raydiumPoolConfigOrBaseVaultAuthority, role: 0 },
        { address: accounts.poolTokenVaultA, role: 1 },
        { address: accounts.poolTokenVaultB, role: 1 },
        { address: accounts.tickArray0, role: 1 },
        { address: accounts.tickArray1, role: 1 },
        { address: accounts.tickArray2, role: 1 },
        { address: accounts.oracle, role: 1 },
        { address: accounts.poolProgram, role: 0 },
        { address: accounts.scopePrices, role: 0 },
        { address: accounts.tokenInfos, role: 0 },
        { address: accounts.tickArrayLower, role: 0 },
        { address: accounts.tickArrayUpper, role: 0 },
        { address: accounts.tokenATokenProgram, role: 0 },
        { address: accounts.tokenBTokenProgram, role: 0 },
        { address: accounts.memoProgram, role: 0 },
        { address: accounts.tokenProgram, role: 0 },
        { address: accounts.tokenProgram2022, role: 0 },
        { address: accounts.instructionSysvarAccount, role: 0 },
        { address: accounts.consensusAccount, role: 0 },
        (0, kit_1.isSome)(accounts.eventAuthority)
            ? { address: accounts.eventAuthority.value, role: 0 }
            : { address: programAddress, role: 0 },
    ];
    const identifier = Buffer.from([143, 212, 101, 95, 105, 209, 184, 1]);
    const buffer = Buffer.alloc(1000);
    const len = exports.layout.encode({
        targetLimitBps: args.targetLimitBps,
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = { accounts: keys, programAddress, data };
    return ix;
}
