// @kamino-finance/kliquidity-sdk/dist/utils/jupiter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jupInstructionToSolanaInstruction = jupInstructionToSolanaInstruction;
exports.jupInstructionsToSolanaInstructions = jupInstructionsToSolanaInstructions;
exports.getAccountRole = getAccountRole;
const kit_1 = require("@solana/kit");
function jupInstructionToSolanaInstruction(ix) {
    return {
        data: ix.data ? Buffer.from(ix.data, 'base64') : undefined,
        programAddress: (0, kit_1.address)(ix.programId),
        accounts: ix.accounts.map((k) => ({
            address: (0, kit_1.address)(k.pubkey),
            role: getAccountRole({ isSigner: k.isSigner, isMut: k.isWritable }),
        })),
    };
}
function jupInstructionsToSolanaInstructions(ixs) {
    return ixs.map((ix) => jupInstructionToSolanaInstruction(ix));
}
function getAccountRole({ isSigner, isMut }) {
    if (isSigner && isMut) {
        return kit_1.AccountRole.WRITABLE_SIGNER;
    }
    if (isSigner && !isMut) {
        return kit_1.AccountRole.READONLY_SIGNER;
    }
    if (!isSigner && isMut) {
        return kit_1.AccountRole.WRITABLE;
    }
    return kit_1.AccountRole.READONLY;
}
