// @orca-so/whirlpools-core/dist/nodejs/orca_whirlpools_core_js_bindings.js

let imports = {};
imports['__wbindgen_placeholder__'] = module.exports;
let wasm;
const { TextEncoder, TextDecoder } = require(`util`);

const heap = new Array(128).fill(undefined);

heap.push(undefined, null, true, false);

function getObject(idx) { return heap[idx]; }

let WASM_VECTOR_LEN = 0;

let cachedUint8ArrayMemory0 = null;

function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

let cachedTextEncoder = new TextEncoder('utf-8');

const encodeString = (typeof cachedTextEncoder.encodeInto === 'function'
    ? function (arg, view) {
    return cachedTextEncoder.encodeInto(arg, view);
}
    : function (arg, view) {
    const buf = cachedTextEncoder.encode(arg);
    view.set(buf);
    return {
        read: arg.length,
        written: buf.length
    };
});

function passStringToWasm0(arg, malloc, realloc) {

    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }

    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = encodeString(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

let cachedDataViewMemory0 = null;

function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null || cachedDataViewMemory0.buffer.detached === true || (cachedDataViewMemory0.buffer.detached === undefined && cachedDataViewMemory0.buffer !== wasm.memory.buffer)) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}

let heap_next = heap.length;

function addHeapObject(obj) {
    if (heap_next === heap.length) heap.push(heap.length + 1);
    const idx = heap_next;
    heap_next = heap[idx];

    heap[idx] = obj;
    return idx;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        wasm.__wbindgen_export_2(addHeapObject(e));
    }
}

function dropObject(idx) {
    if (idx < 132) return;
    heap[idx] = heap_next;
    heap_next = idx;
}

function takeObject(idx) {
    const ret = getObject(idx);
    dropObject(idx);
    return ret;
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

function debugString(val) {
    // primitive types
    const type = typeof val;
    if (type == 'number' || type == 'boolean' || val == null) {
        return  `${val}`;
    }
    if (type == 'string') {
        return `"${val}"`;
    }
    if (type == 'symbol') {
        const description = val.description;
        if (description == null) {
            return 'Symbol';
        } else {
            return `Symbol(${description})`;
        }
    }
    if (type == 'function') {
        const name = val.name;
        if (typeof name == 'string' && name.length > 0) {
            return `Function(${name})`;
        } else {
            return 'Function';
        }
    }
    // objects
    if (Array.isArray(val)) {
        const length = val.length;
        let debug = '[';
        if (length > 0) {
            debug += debugString(val[0]);
        }
        for(let i = 1; i < length; i++) {
            debug += ', ' + debugString(val[i]);
        }
        debug += ']';
        return debug;
    }
    // Test for built-in
    const builtInMatches = /\[object ([^\]]+)\]/.exec(toString.call(val));
    let className;
    if (builtInMatches && builtInMatches.length > 1) {
        className = builtInMatches[1];
    } else {
        // Failed to match the standard '[object ClassName]'
        return toString.call(val);
    }
    if (className == 'Object') {
        // we're a user defined class or Object
        // JSON.stringify avoids problems with cycles, and is generally much
        // easier than looping through ownProperties of `val`.
        try {
            return 'Object(' + JSON.stringify(val) + ')';
        } catch (_) {
            return 'Object';
        }
    }
    // errors
    if (val instanceof Error) {
        return `${val.name}: ${val.message}\n${val.stack}`;
    }
    // TODO we could test for more things here, like `Set`s and `Map`s.
    return className;
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}
/**
 * Check if the whirlpool is initialized with adaptive fee
 *
 * # Paramters
 * - `whirlpool`: The whirlpool state
 *
 * # Returns
 * - A boolean value indicating if the whirlpool is initialized with adaptive fee
 */
module.exports.isInitializedWithAdaptiveFee = function(whirlpool) {
    const ret = wasm.isInitializedWithAdaptiveFee(addHeapObject(whirlpool));
    return ret !== 0;
};

/**
 * Calculate the amount A delta between two sqrt_prices
 *
 * # Parameters
 * - `sqrt_price_1`: The first square root price
 * - `sqrt_price_2`: The second square root price
 * - `liquidity`: The liquidity
 * - `round_up`: Whether to round up or not
 *
 * # Returns
 * - `u64`: The amount delta
 */
module.exports.tryGetAmountDeltaA = function(sqrt_price_1, sqrt_price_2, liquidity, round_up) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryGetAmountDeltaA(retptr, addHeapObject(sqrt_price_1), addHeapObject(sqrt_price_2), addHeapObject(liquidity), round_up);
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the amount B delta between two sqrt_prices
 *
 * # Parameters
 * - `sqrt_price_1`: The first square root price
 * - `sqrt_price_2`: The second square root price
 * - `liquidity`: The liquidity
 * - `round_up`: Whether to round up or not
 *
 * # Returns
 * - `u64`: The amount delta
 */
module.exports.tryGetAmountDeltaB = function(sqrt_price_1, sqrt_price_2, liquidity, round_up) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryGetAmountDeltaB(retptr, addHeapObject(sqrt_price_1), addHeapObject(sqrt_price_2), addHeapObject(liquidity), round_up);
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the next square root price
 *
 * # Parameters
 * - `current_sqrt_price`: The current square root price
 * - `current_liquidity`: The current liquidity
 * - `amount`: The amount
 * - `specified_input`: Whether the input is specified
 *
 * # Returns
 * - `u128`: The next square root price
 */
module.exports.tryGetNextSqrtPriceFromA = function(current_sqrt_price, current_liquidity, amount, specified_input) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryGetNextSqrtPriceFromA(retptr, addHeapObject(current_sqrt_price), addHeapObject(current_liquidity), amount, specified_input);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the next square root price
 *
 * # Parameters
 * - `current_sqrt_price`: The current square root price
 * - `current_liquidity`: The current liquidity
 * - `amount`: The amount
 * - `specified_input`: Whether the input is specified
 *
 * # Returns
 * - `u128`: The next square root price
 */
module.exports.tryGetNextSqrtPriceFromB = function(current_sqrt_price, current_liquidity, amount, specified_input) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryGetNextSqrtPriceFromB(retptr, addHeapObject(current_sqrt_price), addHeapObject(current_liquidity), amount, specified_input);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Apply a transfer fee to an amount
 * e.g. You send 10000 amount with 100 fee rate. The fee amount will be 100.
 * So the amount after fee will be 9900.
 *
 * # Parameters
 * - `amount`: The amount to apply the fee to
 * - `transfer_fee`: The transfer fee to apply
 *
 * # Returns
 * - `u64`: The amount after the fee has been applied
 */
module.exports.tryApplyTransferFee = function(amount, transfer_fee) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryApplyTransferFee(retptr, amount, addHeapObject(transfer_fee));
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Reverse the application of a transfer fee to an amount
 * e.g. You received 9900 amount with 100 fee rate. The fee amount will be 100.
 * So the amount before fee will be 10000.
 *
 * # Parameters
 * - `amount`: The amount to reverse the fee from
 * - `transfer_fee`: The transfer fee to reverse
 *
 * # Returns
 * - `u64`: The amount before the fee has been applied
 */
module.exports.tryReverseApplyTransferFee = function(amount, transfer_fee) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryReverseApplyTransferFee(retptr, amount, addHeapObject(transfer_fee));
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Get the maximum amount with a slippage tolerance
 * e.g. Your estimated amount you send is 10000 with 100 slippage tolerance. The max you send will be 10100.
 *
 * # Parameters
 * - `amount`: The amount to apply the fee to
 * - `slippage_tolerance_bps`: The slippage tolerance in bps (should be in range 0..BPS_DENOMINATOR)
 *
 * # Returns
 * - `u64`: The maximum amount
 */
module.exports.tryGetMaxAmountWithSlippageTolerance = function(amount, slippage_tolerance_bps) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryGetMaxAmountWithSlippageTolerance(retptr, amount, slippage_tolerance_bps);
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Get the minimum amount with a slippage tolerance
 * e.g. Your estimated amount you receive is 10000 with 100 slippage tolerance. The min amount you receive will be 9900.
 *
 * # Parameters
 * - `amount`: The amount to apply the fee to
 * - `slippage_tolerance_bps`: The slippage tolerance in bps (should be in range 0..BPS_DENOMINATOR)
 *
 * # Returns
 * - `u64`: The minimum amount
 */
module.exports.tryGetMinAmountWithSlippageTolerance = function(amount, slippage_tolerance_bps) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryGetMinAmountWithSlippageTolerance(retptr, amount, slippage_tolerance_bps);
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Apply a swap fee to an amount
 * e.g. You send 10000 amount with 10000 fee rate. The fee amount will be 100.
 * So the amount after fee will be 9900.
 *
 * # Parameters
 * - `amount`: The amount to apply the fee to
 * - `fee_rate`: The fee rate to apply denominated in 1e6
 *
 * # Returns
 * - `u64`: The amount after the fee has been applied
 */
module.exports.tryApplySwapFee = function(amount, fee_rate) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryApplySwapFee(retptr, amount, fee_rate);
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Reverse the application of a swap fee to an amount
 * e.g. You received 9900 amount with 10000 fee rate. The fee amount will be 100.
 * So the amount before fee will be 10000.
 *
 * # Parameters
 * - `amount`: The amount to reverse the fee from
 * - `fee_rate`: The fee rate to reverse denominated in 1e6
 *
 * # Returns
 * - `u64`: The amount before the fee has been applied
 */
module.exports.tryReverseApplySwapFee = function(amount, fee_rate) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.tryReverseApplySwapFee(retptr, amount, fee_rate);
        var r0 = getDataViewMemory0().getBigInt64(retptr + 8 * 0, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        var r3 = getDataViewMemory0().getInt32(retptr + 4 * 3, true);
        if (r3) {
            throw takeObject(r2);
        }
        return BigInt.asUintN(64, r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate rewards owed for a position
 *
 * # Paramters
 * - `whirlpool`: The whirlpool state
 * - `position`: The position state
 * - `tick_lower`: The lower tick state
 * - `tick_upper`: The upper tick state
 * - `current_timestamp`: The current timestamp
 * - `transfer_fee_1`: The transfer fee for token 1
 * - `transfer_fee_2`: The transfer fee for token 2
 * - `transfer_fee_3`: The transfer fee for token 3
 *
 * # Returns
 * - `CollectRewardsQuote`: The rewards owed for the 3 reward tokens.
 */
module.exports.collectRewardsQuote = function(whirlpool, position, tick_lower, tick_upper, current_timestamp, transfer_fee_1, transfer_fee_2, transfer_fee_3) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.collectRewardsQuote(retptr, addHeapObject(whirlpool), addHeapObject(position), addHeapObject(tick_lower), addHeapObject(tick_upper), current_timestamp, isLikeNone(transfer_fee_1) ? 0 : addHeapObject(transfer_fee_1), isLikeNone(transfer_fee_2) ? 0 : addHeapObject(transfer_fee_2), isLikeNone(transfer_fee_3) ? 0 : addHeapObject(transfer_fee_3));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

function passArray8ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 1, 1) >>> 0;
    getUint8ArrayMemory0().set(arg, ptr / 1);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}
/**
 * Get the first unoccupied position in a bundle
 *
 * # Arguments
 * * `bundle` - The bundle to check
 *
 * # Returns
 * * `u32` - The first unoccupied position (None if full)
 */
module.exports.firstUnoccupiedPositionInBundle = function(bitmap) {
    const ptr0 = passArray8ToWasm0(bitmap, wasm.__wbindgen_export_0);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.firstUnoccupiedPositionInBundle(ptr0, len0);
    return ret === 0x100000001 ? undefined : ret;
};

/**
 * Check whether a position bundle is full
 * A position bundle can contain 256 positions
 *
 * # Arguments
 * * `bundle` - The bundle to check
 *
 * # Returns
 * * `bool` - Whether the bundle is full
 */
module.exports.isPositionBundleFull = function(bitmap) {
    const ptr0 = passArray8ToWasm0(bitmap, wasm.__wbindgen_export_0);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.isPositionBundleFull(ptr0, len0);
    return ret !== 0;
};

/**
 * Check whether a position bundle is empty
 *
 * # Arguments
 * * `bundle` - The bundle to check
 *
 * # Returns
 * * `bool` - Whether the bundle is empty
 */
module.exports.isPositionBundleEmpty = function(bitmap) {
    const ptr0 = passArray8ToWasm0(bitmap, wasm.__wbindgen_export_0);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.isPositionBundleEmpty(ptr0, len0);
    return ret !== 0;
};

/**
 * Get the first tick index in the tick array that contains the specified tick index.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 * - `tick_spacing` - A i32 integer representing the tick spacing
 *
 * # Returns
 * - A i32 integer representing the first tick index in the tick array
 */
module.exports.getTickArrayStartTickIndex = function(tick_index, tick_spacing) {
    const ret = wasm.getTickArrayStartTickIndex(tick_index, tick_spacing);
    return ret;
};

/**
 * Derive the sqrt-price from a tick index. The precision of this method is only guarranted
 * if tick is within the bounds of {max, min} tick-index.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 *
 * # Returns
 * - `Ok`: A u128 Q32.64 representing the sqrt_price
 */
module.exports.tickIndexToSqrtPrice = function(tick_index) {
    const ret = wasm.tickIndexToSqrtPrice(tick_index);
    return takeObject(ret);
};

/**
 * Derive the tick index from a sqrt price. The precision of this method is only guarranted
 * if tick is within the bounds of {max, min} tick-index.
 *
 * # Parameters
 * - `sqrt_price` - A u128 integer representing the sqrt price
 *
 * # Returns
 * - `Ok`: A i32 integer representing the tick integer
 */
module.exports.sqrtPriceToTickIndex = function(sqrt_price) {
    const ret = wasm.sqrtPriceToTickIndex(addHeapObject(sqrt_price));
    return ret;
};

/**
 * Get the initializable tick index.
 * If the tick index is already initializable, it is returned as is.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 * - `tick_spacing` - A i32 integer representing the tick spacing
 * - `round_up` - A boolean value indicating if the supplied tick index should be rounded up. None will round to the nearest.
 *
 * # Returns
 * - A i32 integer representing the previous initializable tick index
 */
module.exports.getInitializableTickIndex = function(tick_index, tick_spacing, round_up) {
    const ret = wasm.getInitializableTickIndex(tick_index, tick_spacing, isLikeNone(round_up) ? 0xFFFFFF : round_up ? 1 : 0);
    return ret;
};

/**
 * Get the previous initializable tick index.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 * - `tick_spacing` - A i32 integer representing the tick spacing
 *
 * # Returns
 * - A i32 integer representing the previous initializable tick index
 */
module.exports.getPrevInitializableTickIndex = function(tick_index, tick_spacing) {
    const ret = wasm.getPrevInitializableTickIndex(tick_index, tick_spacing);
    return ret;
};

/**
 * Get the next initializable tick index.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 * - `tick_spacing` - A i32 integer representing the tick spacing
 *
 * # Returns
 * - A i32 integer representing the next initializable tick index
 */
module.exports.getNextInitializableTickIndex = function(tick_index, tick_spacing) {
    const ret = wasm.getNextInitializableTickIndex(tick_index, tick_spacing);
    return ret;
};

/**
 * Check if a tick is in-bounds.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 *
 * # Returns
 * - A boolean value indicating if the tick is in-bounds
 */
module.exports.isTickIndexInBounds = function(tick_index) {
    const ret = wasm.isTickIndexInBounds(tick_index);
    return ret !== 0;
};

/**
 * Check if a tick is initializable.
 * A tick is initializable if it is divisible by the tick spacing.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 * - `tick_spacing` - A i32 integer representing the tick spacing
 *
 * # Returns
 * - A boolean value indicating if the tick is initializable
 */
module.exports.isTickInitializable = function(tick_index, tick_spacing) {
    const ret = wasm.isTickInitializable(tick_index, tick_spacing);
    return ret !== 0;
};

/**
 * Get the tick index for the inverse of the price that this tick represents.
 * Eg: Consider tick i where Pb/Pa = 1.0001 ^ i
 * inverse of this, i.e. Pa/Pb = 1 / (1.0001 ^ i) = 1.0001^-i
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick integer
 *
 * # Returns
 * - A i32 integer representing the tick index for the inverse of the price
 */
module.exports.invertTickIndex = function(tick_index) {
    const ret = wasm.invertTickIndex(tick_index);
    return ret;
};

/**
 * Get the sqrt price for the inverse of the price that this tick represents.
 * Because converting to a tick index and then back to a sqrt price is lossy,
 * this function is clamped to the nearest tick index.
 *
 * # Parameters
 * - `sqrt_price` - A u128 integer representing the sqrt price
 *
 * # Returns
 * - A u128 integer representing the sqrt price for the inverse of the price
 */
module.exports.invertSqrtPrice = function(sqrt_price) {
    const ret = wasm.invertSqrtPrice(addHeapObject(sqrt_price));
    return takeObject(ret);
};

/**
 * Get the minimum and maximum tick index that can be initialized.
 *
 * # Parameters
 * - `tick_spacing` - A i32 integer representing the tick spacing
 *
 * # Returns
 * - A TickRange struct containing the lower and upper tick index
 */
module.exports.getFullRangeTickIndexes = function(tick_spacing) {
    const ret = wasm.getFullRangeTickIndexes(tick_spacing);
    return takeObject(ret);
};

/**
 * Order tick indexes in ascending order.
 * If the lower tick index is greater than the upper tick index, the indexes are swapped.
 * This is useful for ensuring that the lower tick index is always less than the upper tick index.
 *
 * # Parameters
 * - `tick_index_1` - A i32 integer representing the first tick index
 * - `tick_index_2` - A i32 integer representing the second tick index
 *
 * # Returns
 * - A TickRange struct containing the lower and upper tick index
 */
module.exports.orderTickIndexes = function(tick_index_1, tick_index_2) {
    const ret = wasm.orderTickIndexes(tick_index_1, tick_index_2);
    return takeObject(ret);
};

/**
 * Check if a whirlpool is a full-range only pool.
 *
 * # Parameters
 * - `tick_spacing` - A u16 integer representing the tick spacing
 *
 * # Returns
 * - A boolean value indicating if the whirlpool is a full-range only pool
 */
module.exports.isFullRangeOnly = function(tick_spacing) {
    const ret = wasm.isFullRangeOnly(tick_spacing);
    return ret !== 0;
};

/**
 * Get the index of a tick in a tick array.
 *
 * # Parameters
 * - `tick_index` - A i32 integer representing the tick index
 * - `tick_array_start_index` - A i32 integer representing the start tick index of the tick array
 * - `tick_spacing` - A u16 integer representing the tick spacing
 *
 * # Returns
 * - A u32 integer representing the tick index in the tick array
 */
module.exports.getTickIndexInArray = function(tick_index, tick_array_start_index, tick_spacing) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.getTickIndexInArray(retptr, tick_index, tick_array_start_index, tick_spacing);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return r0 >>> 0;
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._VOLATILITY_ACCUMULATOR_SCALE_FACTOR = function() {
    const ret = wasm._REDUCTION_FACTOR_DENOMINATOR();
    return ret;
};

module.exports._REDUCTION_FACTOR_DENOMINATOR = function() {
    const ret = wasm._REDUCTION_FACTOR_DENOMINATOR();
    return ret;
};

module.exports._ADAPTIVE_FEE_CONTROL_FACTOR_DENOMINATOR = function() {
    const ret = wasm._ADAPTIVE_FEE_CONTROL_FACTOR_DENOMINATOR();
    return ret >>> 0;
};

module.exports._MAX_REFERENCE_AGE = function() {
    const ret = wasm._MAX_REFERENCE_AGE();
    return BigInt.asUintN(64, ret);
};

module.exports._FEE_RATE_HARD_LIMIT = function() {
    const ret = wasm._ADAPTIVE_FEE_CONTROL_FACTOR_DENOMINATOR();
    return ret >>> 0;
};

module.exports._POSITION_BUNDLE_SIZE = function() {
    const ret = wasm._POSITION_BUNDLE_SIZE();
    return ret >>> 0;
};

module.exports._TICK_ARRAY_NOT_EVENLY_SPACED = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._TICK_ARRAY_NOT_EVENLY_SPACED(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._TICK_INDEX_OUT_OF_BOUNDS = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._TICK_INDEX_OUT_OF_BOUNDS(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_TICK_INDEX = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_TICK_INDEX(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._ARITHMETIC_OVERFLOW = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._ARITHMETIC_OVERFLOW(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._AMOUNT_EXCEEDS_MAX_U64 = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._AMOUNT_EXCEEDS_MAX_U64(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._SQRT_PRICE_OUT_OF_BOUNDS = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._SQRT_PRICE_OUT_OF_BOUNDS(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._TICK_SEQUENCE_EMPTY = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._TICK_SEQUENCE_EMPTY(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._SQRT_PRICE_LIMIT_OUT_OF_BOUNDS = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._SQRT_PRICE_LIMIT_OUT_OF_BOUNDS(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_SQRT_PRICE_LIMIT_DIRECTION = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_SQRT_PRICE_LIMIT_DIRECTION(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._ZERO_TRADABLE_AMOUNT = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._ZERO_TRADABLE_AMOUNT(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_TIMESTAMP = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_TIMESTAMP(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_TRANSFER_FEE = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_TRANSFER_FEE(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_SLIPPAGE_TOLERANCE = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_SLIPPAGE_TOLERANCE(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._TICK_INDEX_NOT_IN_ARRAY = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._TICK_INDEX_NOT_IN_ARRAY(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_TICK_ARRAY_SEQUENCE = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_TICK_ARRAY_SEQUENCE(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._INVALID_ADAPTIVE_FEE_INFO = function() {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm._INVALID_ADAPTIVE_FEE_INFO(retptr);
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        return getStringFromWasm0(r0, r1);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports._NUM_REWARDS = function() {
    const ret = wasm._NUM_REWARDS();
    return ret >>> 0;
};

module.exports._FEE_RATE_DENOMINATOR = function() {
    const ret = wasm._FEE_RATE_DENOMINATOR();
    return ret >>> 0;
};

module.exports._TICK_ARRAY_SIZE = function() {
    const ret = wasm._TICK_ARRAY_SIZE();
    return ret >>> 0;
};

module.exports._FULL_RANGE_ONLY_TICK_SPACING_THRESHOLD = function() {
    const ret = wasm._FULL_RANGE_ONLY_TICK_SPACING_THRESHOLD();
    return ret;
};

module.exports._MIN_TICK_INDEX = function() {
    const ret = wasm._MIN_TICK_INDEX();
    return ret;
};

module.exports._MAX_TICK_INDEX = function() {
    const ret = wasm._MAX_TICK_INDEX();
    return ret;
};

/**
 * Check if a position is in range.
 * When a position is in range it is earning fees and rewards
 *
 * # Parameters
 * - `sqrt_price` - A u128 integer representing the sqrt price of the pool
 * - `tick_index_1` - A i32 integer representing the first tick index of the position
 * - `tick_index_2` - A i32 integer representing the second tick index of the position
 *
 * # Returns
 * - A boolean value indicating if the position is in range
 */
module.exports.isPositionInRange = function(current_sqrt_price, tick_index_1, tick_index_2) {
    const ret = wasm.isPositionInRange(addHeapObject(current_sqrt_price), tick_index_1, tick_index_2);
    return ret !== 0;
};

/**
 * Calculate the status of a position
 * The status can be one of three values:
 * - InRange: The position is in range
 * - BelowRange: The position is below the range
 * - AboveRange: The position is above the range
 *
 * # Parameters
 * - `sqrt_price` - A u128 integer representing the sqrt price of the pool
 * - `tick_index_1` - A i32 integer representing the first tick index of the position
 * - `tick_index_2` - A i32 integer representing the second tick index of the position
 *
 * # Returns
 * - A PositionStatus enum value indicating the status of the position
 */
module.exports.positionStatus = function(current_sqrt_price, tick_index_1, tick_index_2) {
    const ret = wasm.positionStatus(addHeapObject(current_sqrt_price), tick_index_1, tick_index_2);
    return takeObject(ret);
};

/**
 * Calculate the token_a / token_b ratio of a (ficticious) position
 *
 * # Parameters
 * - `sqrt_price` - A u128 integer representing the sqrt price of the pool
 * - `tick_index_1` - A i32 integer representing the first tick index of the position
 * - `tick_index_2` - A i32 integer representing the second tick index of the position
 *
 * # Returns
 * - A PositionRatio struct containing the ratio of token_a and token_b
 */
module.exports.positionRatio = function(current_sqrt_price, tick_index_1, tick_index_2) {
    const ret = wasm.positionRatio(addHeapObject(current_sqrt_price), tick_index_1, tick_index_2);
    return takeObject(ret);
};

/**
 * Convert a price into a sqrt priceX64
 * IMPORTANT: floating point operations can reduce the precision of the result.
 * Make sure to do these operations last and not to use the result for further calculations.
 *
 * # Parameters
 * * `price` - The price to convert
 * * `decimals_a` - The number of decimals of the base token
 * * `decimals_b` - The number of decimals of the quote token
 *
 * # Returns
 * * `u128` - The sqrt priceX64
 */
module.exports.priceToSqrtPrice = function(price, decimals_a, decimals_b) {
    const ret = wasm.priceToSqrtPrice(price, decimals_a, decimals_b);
    return takeObject(ret);
};

/**
 * Convert a sqrt priceX64 into a tick index
 * IMPORTANT: floating point operations can reduce the precision of the result.
 * Make sure to do these operations last and not to use the result for further calculations.
 *
 * # Parameters
 * * `sqrt_price` - The sqrt priceX64 to convert
 * * `decimals_a` - The number of decimals of the base token
 * * `decimals_b` - The number of decimals of the quote token
 *
 * # Returns
 * * `f64` - The decimal price
 */
module.exports.sqrtPriceToPrice = function(sqrt_price, decimals_a, decimals_b) {
    const ret = wasm.sqrtPriceToPrice(addHeapObject(sqrt_price), decimals_a, decimals_b);
    return ret;
};

/**
 * Invert a price
 * IMPORTANT: floating point operations can reduce the precision of the result.
 * Make sure to do these operations last and not to use the result for further calculations.
 *
 * # Parameters
 * * `price` - The price to invert
 * * `decimals_a` - The number of decimals of the base token
 * * `decimals_b` - The number of decimals of the quote token
 *
 * # Returns
 * * `f64` - The inverted price
 */
module.exports.invertPrice = function(price, decimals_a, decimals_b) {
    const ret = wasm.invertPrice(price, decimals_a, decimals_b);
    return ret;
};

/**
 * Convert a tick index into a price
 * IMPORTANT: floating point operations can reduce the precision of the result.
 * Make sure to do these operations last and not to use the result for further calculations.
 *
 * # Parameters
 * * `tick_index` - The tick index to convert
 * * `decimals_a` - The number of decimals of the base token
 * * `decimals_b` - The number of decimals of the quote token
 *
 * # Returns
 * * `f64` - The decimal price
 */
module.exports.tickIndexToPrice = function(tick_index, decimals_a, decimals_b) {
    const ret = wasm.tickIndexToPrice(tick_index, decimals_a, decimals_b);
    return ret;
};

/**
 * Convert a price into a tick index
 * IMPORTANT: floating point operations can reduce the precision of the result.
 * Make sure to do these operations last and not to use the result for further calculations.
 *
 * # Parameters
 * * `price` - The price to convert
 * * `decimals_a` - The number of decimals of the base token
 * * `decimals_b` - The number of decimals of the quote token
 *
 * # Returns
 * * `i32` - The tick index
 */
module.exports.priceToTickIndex = function(price, decimals_a, decimals_b) {
    const ret = wasm.priceToTickIndex(price, decimals_a, decimals_b);
    return ret;
};

/**
 * Computes the exact input or output amount for a swap transaction.
 *
 * # Arguments
 * - `token_in`: The input token amount.
 * - `specified_token_a`: If `true`, the input token is token A. Otherwise, it is token B.
 * - `slippage_tolerance`: The slippage tolerance in basis points.
 * - `whirlpool`: The whirlpool state.
 * - `oracle`: The oracle data for the whirlpool.
 * - `tick_arrays`: The tick arrays needed for the swap.
 * - `timestamp`: The timestamp for the swap.
 * - `transfer_fee_a`: The transfer fee for token A.
 * - `transfer_fee_b`: The transfer fee for token B.
 *
 * # Returns
 * The exact input or output amount for the swap transaction.
 */
module.exports.swapQuoteByInputToken = function(token_in, specified_token_a, slippage_tolerance_bps, whirlpool, oracle, tick_arrays, timestamp, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.swapQuoteByInputToken(retptr, token_in, specified_token_a, slippage_tolerance_bps, addHeapObject(whirlpool), isLikeNone(oracle) ? 0 : addHeapObject(oracle), addHeapObject(tick_arrays), timestamp, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Computes the exact input or output amount for a swap transaction.
 *
 * # Arguments
 * - `token_out`: The output token amount.
 * - `specified_token_a`: If `true`, the output token is token A. Otherwise, it is token B.
 * - `slippage_tolerance`: The slippage tolerance in basis points.
 * - `whirlpool`: The whirlpool state.
 * - `oracle`: The oracle data for the whirlpool.
 * - `tick_arrays`: The tick arrays needed for the swap.
 * - `timestamp`: The timestamp for the swap.
 * - `transfer_fee_a`: The transfer fee for token A.
 * - `transfer_fee_b`: The transfer fee for token B.
 *
 * # Returns
 * The exact input or output amount for the swap transaction.
 */
module.exports.swapQuoteByOutputToken = function(token_out, specified_token_a, slippage_tolerance_bps, whirlpool, oracle, tick_arrays, timestamp, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.swapQuoteByOutputToken(retptr, token_out, specified_token_a, slippage_tolerance_bps, addHeapObject(whirlpool), isLikeNone(oracle) ? 0 : addHeapObject(oracle), addHeapObject(tick_arrays), timestamp, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate fees owed for a position
 *
 * # Paramters
 * - `whirlpool`: The whirlpool state
 * - `position`: The position state
 * - `tick_lower`: The lower tick state
 * - `tick_upper`: The upper tick state
 * - `transfer_fee_a`: The transfer fee for token A
 * - `transfer_fee_b`: The transfer fee for token B
 *
 * # Returns
 * - `CollectFeesQuote`: The fees owed for token A and token B
 */
module.exports.collectFeesQuote = function(whirlpool, position, tick_lower, tick_upper, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.collectFeesQuote(retptr, addHeapObject(whirlpool), addHeapObject(position), addHeapObject(tick_lower), addHeapObject(tick_upper), isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the quote for decreasing liquidity
 *
 * # Parameters
 * - `liquidity_delta` - The amount of liquidity to decrease
 * - `slippage_tolerance` - The slippage tolerance in bps
 * - `current_sqrt_price` - The current sqrt price of the pool
 * - `tick_index_1` - The first tick index of the position
 * - `tick_index_2` - The second tick index of the position
 * - `transfer_fee_a` - The transfer fee for token A in bps
 * - `transfer_fee_b` - The transfer fee for token B in bps
 *
 * # Returns
 * - A DecreaseLiquidityQuote struct containing the estimated token amounts
 */
module.exports.decreaseLiquidityQuote = function(liquidity_delta, slippage_tolerance_bps, current_sqrt_price, tick_index_1, tick_index_2, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.decreaseLiquidityQuote(retptr, addHeapObject(liquidity_delta), slippage_tolerance_bps, addHeapObject(current_sqrt_price), tick_index_1, tick_index_2, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the quote for decreasing liquidity given a token a amount
 *
 * # Parameters
 * - `token_amount_a` - The amount of token a to decrease
 * - `slippage_tolerance` - The slippage tolerance in bps
 * - `current_sqrt_price` - The current sqrt price of the pool
 * - `tick_index_1` - The first tick index of the position
 * - `tick_index_2` - The second tick index of the position
 * - `transfer_fee_a` - The transfer fee for token A in bps
 * - `transfer_fee_b` - The transfer fee for token B in bps
 *
 * # Returns
 * - A DecreaseLiquidityQuote struct containing the estimated token amounts
 */
module.exports.decreaseLiquidityQuoteA = function(token_amount_a, slippage_tolerance_bps, current_sqrt_price, tick_index_1, tick_index_2, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.decreaseLiquidityQuoteA(retptr, token_amount_a, slippage_tolerance_bps, addHeapObject(current_sqrt_price), tick_index_1, tick_index_2, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the quote for decreasing liquidity given a token b amount
 *
 * # Parameters
 * - `token_amount_b` - The amount of token b to decrease
 * - `slippage_tolerance` - The slippage tolerance in bps
 * - `current_sqrt_price` - The current sqrt price of the pool
 * - `tick_index_1` - The first tick index of the position
 * - `tick_index_2` - The second tick index of the position
 * - `transfer_fee_a` - The transfer fee for token A in bps
 * - `transfer_fee_b` - The transfer fee for token B in bps
 *
 * # Returns
 * - A DecreaseLiquidityQuote struct containing the estimated token amounts
 */
module.exports.decreaseLiquidityQuoteB = function(token_amount_b, slippage_tolerance_bps, current_sqrt_price, tick_index_1, tick_index_2, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.decreaseLiquidityQuoteB(retptr, token_amount_b, slippage_tolerance_bps, addHeapObject(current_sqrt_price), tick_index_1, tick_index_2, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the quote for increasing liquidity
 *
 * # Parameters
 * - `liquidity_delta` - The amount of liquidity to increase
 * - `slippage_tolerance` - The slippage tolerance in bps
 * - `current_sqrt_price` - The current sqrt price of the pool
 * - `tick_index_1` - The first tick index of the position
 * - `tick_index_2` - The second tick index of the position
 * - `transfer_fee_a` - The transfer fee for token A in bps
 * - `transfer_fee_b` - The transfer fee for token B in bps
 *
 * # Returns
 * - An IncreaseLiquidityQuote struct containing the estimated token amounts
 */
module.exports.increaseLiquidityQuote = function(liquidity_delta, slippage_tolerance_bps, current_sqrt_price, tick_index_1, tick_index_2, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.increaseLiquidityQuote(retptr, addHeapObject(liquidity_delta), slippage_tolerance_bps, addHeapObject(current_sqrt_price), tick_index_1, tick_index_2, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the quote for increasing liquidity given a token a amount
 *
 * # Parameters
 * - `token_amount_a` - The amount of token a to increase
 * - `slippage_tolerance` - The slippage tolerance in bps
 * - `current_sqrt_price` - The current sqrt price of the pool
 * - `tick_index_1` - The first tick index of the position
 * - `tick_index_2` - The second tick index of the position
 * - `transfer_fee_a` - The transfer fee for token A in bps
 * - `transfer_fee_b` - The transfer fee for token B in bps
 *
 * # Returns
 * - An IncreaseLiquidityQuote struct containing the estimated token amounts
 */
module.exports.increaseLiquidityQuoteA = function(token_amount_a, slippage_tolerance_bps, current_sqrt_price, tick_index_1, tick_index_2, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.increaseLiquidityQuoteA(retptr, token_amount_a, slippage_tolerance_bps, addHeapObject(current_sqrt_price), tick_index_1, tick_index_2, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

/**
 * Calculate the quote for increasing liquidity given a token b amount
 *
 * # Parameters
 * - `token_amount_b` - The amount of token b to increase
 * - `slippage_tolerance` - The slippage tolerance in bps
 * - `current_sqrt_price` - The current sqrt price of the pool
 * - `tick_index_1` - The first tick index of the position
 * - `tick_index_2` - The second tick index of the position
 * - `transfer_fee_a` - The transfer fee for token A in bps
 * - `transfer_fee_b` - The transfer fee for token B in bps
 *
 * # Returns
 * - An IncreaseLiquidityQuote struct containing the estimated token amounts
 */
module.exports.increaseLiquidityQuoteB = function(token_amount_b, slippage_tolerance_bps, current_sqrt_price, tick_index_1, tick_index_2, transfer_fee_a, transfer_fee_b) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        wasm.increaseLiquidityQuoteB(retptr, token_amount_b, slippage_tolerance_bps, addHeapObject(current_sqrt_price), tick_index_1, tick_index_2, isLikeNone(transfer_fee_a) ? 0 : addHeapObject(transfer_fee_a), isLikeNone(transfer_fee_b) ? 0 : addHeapObject(transfer_fee_b));
        var r0 = getDataViewMemory0().getInt32(retptr + 4 * 0, true);
        var r1 = getDataViewMemory0().getInt32(retptr + 4 * 1, true);
        var r2 = getDataViewMemory0().getInt32(retptr + 4 * 2, true);
        if (r2) {
            throw takeObject(r1);
        }
        return takeObject(r0);
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
};

module.exports.__wbg_String_eecc4a11987127d6 = function(arg0, arg1) {
    const ret = String(getObject(arg1));
    const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_export_0, wasm.__wbindgen_export_1);
    const len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
};

module.exports.__wbg_buffer_609cc3eee51ed158 = function(arg0) {
    const ret = getObject(arg0).buffer;
    return addHeapObject(ret);
};

module.exports.__wbg_call_672a4d21634d4a24 = function() { return handleError(function (arg0, arg1) {
    const ret = getObject(arg0).call(getObject(arg1));
    return addHeapObject(ret);
}, arguments) };

module.exports.__wbg_done_769e5ede4b31c67b = function(arg0) {
    const ret = getObject(arg0).done;
    return ret;
};

module.exports.__wbg_get_67b2ba62fc30de12 = function() { return handleError(function (arg0, arg1) {
    const ret = Reflect.get(getObject(arg0), getObject(arg1));
    return addHeapObject(ret);
}, arguments) };

module.exports.__wbg_get_b9b93047fe3cf45b = function(arg0, arg1) {
    const ret = getObject(arg0)[arg1 >>> 0];
    return addHeapObject(ret);
};

module.exports.__wbg_getwithrefkey_1dc361bd10053bfe = function(arg0, arg1) {
    const ret = getObject(arg0)[getObject(arg1)];
    return addHeapObject(ret);
};

module.exports.__wbg_getwithrefkey_6550b2c093d2eb18 = function(arg0, arg1) {
    const ret = getObject(arg0)[getObject(arg1)];
    return addHeapObject(ret);
};

module.exports.__wbg_instanceof_ArrayBuffer_e14585432e3737fc = function(arg0) {
    let result;
    try {
        result = getObject(arg0) instanceof ArrayBuffer;
    } catch (_) {
        result = false;
    }
    const ret = result;
    return ret;
};

module.exports.__wbg_instanceof_Uint8Array_17156bcf118086a9 = function(arg0) {
    let result;
    try {
        result = getObject(arg0) instanceof Uint8Array;
    } catch (_) {
        result = false;
    }
    const ret = result;
    return ret;
};

module.exports.__wbg_isArray_a1eab7e0d067391b = function(arg0) {
    const ret = Array.isArray(getObject(arg0));
    return ret;
};

module.exports.__wbg_isSafeInteger_343e2beeeece1bb0 = function(arg0) {
    const ret = Number.isSafeInteger(getObject(arg0));
    return ret;
};

module.exports.__wbg_iterator_9a24c88df860dc65 = function() {
    const ret = Symbol.iterator;
    return addHeapObject(ret);
};

module.exports.__wbg_length_a446193dc22c12f8 = function(arg0) {
    const ret = getObject(arg0).length;
    return ret;
};

module.exports.__wbg_length_e2d2a49132c1b256 = function(arg0) {
    const ret = getObject(arg0).length;
    return ret;
};

module.exports.__wbg_new_405e22f390576ce2 = function() {
    const ret = new Object();
    return addHeapObject(ret);
};

module.exports.__wbg_new_78feb108b6472713 = function() {
    const ret = new Array();
    return addHeapObject(ret);
};

module.exports.__wbg_new_a12002a7f91c75be = function(arg0) {
    const ret = new Uint8Array(getObject(arg0));
    return addHeapObject(ret);
};

module.exports.__wbg_next_25feadfc0913fea9 = function(arg0) {
    const ret = getObject(arg0).next;
    return addHeapObject(ret);
};

module.exports.__wbg_next_6574e1a8a62d1055 = function() { return handleError(function (arg0) {
    const ret = getObject(arg0).next();
    return addHeapObject(ret);
}, arguments) };

module.exports.__wbg_set_37837023f3d740e8 = function(arg0, arg1, arg2) {
    getObject(arg0)[arg1 >>> 0] = takeObject(arg2);
};

module.exports.__wbg_set_3807d5f0bfc24aa7 = function(arg0, arg1, arg2) {
    getObject(arg0)[takeObject(arg1)] = takeObject(arg2);
};

module.exports.__wbg_set_65595bdd868b3009 = function(arg0, arg1, arg2) {
    getObject(arg0).set(getObject(arg1), arg2 >>> 0);
};

module.exports.__wbg_value_cd1ffa7b1ab794f1 = function(arg0) {
    const ret = getObject(arg0).value;
    return addHeapObject(ret);
};

module.exports.__wbindgen_as_number = function(arg0) {
    const ret = +getObject(arg0);
    return ret;
};

module.exports.__wbindgen_bigint_from_i64 = function(arg0) {
    const ret = arg0;
    return addHeapObject(ret);
};

module.exports.__wbindgen_bigint_from_u128 = function(arg0, arg1) {
    const ret = BigInt.asUintN(64, arg0) << BigInt(64) | BigInt.asUintN(64, arg1);
    return addHeapObject(ret);
};

module.exports.__wbindgen_bigint_from_u64 = function(arg0) {
    const ret = BigInt.asUintN(64, arg0);
    return addHeapObject(ret);
};

module.exports.__wbindgen_bigint_get_as_i64 = function(arg0, arg1) {
    const v = getObject(arg1);
    const ret = typeof(v) === 'bigint' ? v : undefined;
    getDataViewMemory0().setBigInt64(arg0 + 8 * 1, isLikeNone(ret) ? BigInt(0) : ret, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
};

module.exports.__wbindgen_boolean_get = function(arg0) {
    const v = getObject(arg0);
    const ret = typeof(v) === 'boolean' ? (v ? 1 : 0) : 2;
    return ret;
};

module.exports.__wbindgen_debug_string = function(arg0, arg1) {
    const ret = debugString(getObject(arg1));
    const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_export_0, wasm.__wbindgen_export_1);
    const len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
};

module.exports.__wbindgen_error_new = function(arg0, arg1) {
    const ret = new Error(getStringFromWasm0(arg0, arg1));
    return addHeapObject(ret);
};

module.exports.__wbindgen_in = function(arg0, arg1) {
    const ret = getObject(arg0) in getObject(arg1);
    return ret;
};

module.exports.__wbindgen_is_array = function(arg0) {
    const ret = Array.isArray(getObject(arg0));
    return ret;
};

module.exports.__wbindgen_is_bigint = function(arg0) {
    const ret = typeof(getObject(arg0)) === 'bigint';
    return ret;
};

module.exports.__wbindgen_is_function = function(arg0) {
    const ret = typeof(getObject(arg0)) === 'function';
    return ret;
};

module.exports.__wbindgen_is_object = function(arg0) {
    const val = getObject(arg0);
    const ret = typeof(val) === 'object' && val !== null;
    return ret;
};

module.exports.__wbindgen_is_undefined = function(arg0) {
    const ret = getObject(arg0) === undefined;
    return ret;
};

module.exports.__wbindgen_jsval_eq = function(arg0, arg1) {
    const ret = getObject(arg0) === getObject(arg1);
    return ret;
};

module.exports.__wbindgen_jsval_loose_eq = function(arg0, arg1) {
    const ret = getObject(arg0) == getObject(arg1);
    return ret;
};

module.exports.__wbindgen_memory = function() {
    const ret = wasm.memory;
    return addHeapObject(ret);
};

module.exports.__wbindgen_number_get = function(arg0, arg1) {
    const obj = getObject(arg1);
    const ret = typeof(obj) === 'number' ? obj : undefined;
    getDataViewMemory0().setFloat64(arg0 + 8 * 1, isLikeNone(ret) ? 0 : ret, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
};

module.exports.__wbindgen_number_new = function(arg0) {
    const ret = arg0;
    return addHeapObject(ret);
};

module.exports.__wbindgen_object_clone_ref = function(arg0) {
    const ret = getObject(arg0);
    return addHeapObject(ret);
};

module.exports.__wbindgen_object_drop_ref = function(arg0) {
    takeObject(arg0);
};

module.exports.__wbindgen_shr = function(arg0, arg1) {
    const ret = getObject(arg0) >> getObject(arg1);
    return addHeapObject(ret);
};

module.exports.__wbindgen_string_get = function(arg0, arg1) {
    const obj = getObject(arg1);
    const ret = typeof(obj) === 'string' ? obj : undefined;
    var ptr1 = isLikeNone(ret) ? 0 : passStringToWasm0(ret, wasm.__wbindgen_export_0, wasm.__wbindgen_export_1);
    var len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
};

module.exports.__wbindgen_string_new = function(arg0, arg1) {
    const ret = getStringFromWasm0(arg0, arg1);
    return addHeapObject(ret);
};

module.exports.__wbindgen_throw = function(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
};

const path = require('path').join(__dirname, 'orca_whirlpools_core_js_bindings_bg.wasm');
const bytes = require('fs').readFileSync(path);

const wasmModule = new WebAssembly.Module(bytes);
const wasmInstance = new WebAssembly.Instance(wasmModule, imports);
wasm = wasmInstance.exports;
module.exports.__wasm = wasm;
