// @mrgnlabs/marginfi-client-v2/dist/services/price/utils/deserialize.utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dtoToOraclePrice = dtoToOraclePrice;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
function dtoToOraclePrice(dto) {
    return {
        priceRealtime: {
            price: new bignumber_js_1.default(dto.priceRealtime.price),
            confidence: new bignumber_js_1.default(dto.priceRealtime.confidence),
            lowestPrice: new bignumber_js_1.default(dto.priceRealtime.lowestPrice),
            highestPrice: new bignumber_js_1.default(dto.priceRealtime.highestPrice),
        },
        priceWeighted: {
            price: new bignumber_js_1.default(dto.priceWeighted.price),
            confidence: new bignumber_js_1.default(dto.priceWeighted.confidence),
            lowestPrice: new bignumber_js_1.default(dto.priceWeighted.lowestPrice),
            highestPrice: new bignumber_js_1.default(dto.priceWeighted.highestPrice),
        },
        timestamp: new bignumber_js_1.default(dto.timestamp),
        pythShardId: dto.pythShardId,
    };
}
