// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/CreationStatus.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.STAGING = exports.DEPRECATED = exports.LIVE = exports.SHADOW = exports.IGNORED = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class IGNORED {
    static discriminator = 0;
    static kind = "IGNORED";
    discriminator = 0;
    kind = "IGNORED";
    toJSON() {
        return {
            kind: "IGNORED",
        };
    }
    toEncodable() {
        return {
            IGNORED: {},
        };
    }
}
exports.IGNORED = IGNORED;
class SHADOW {
    static discriminator = 1;
    static kind = "SHADOW";
    discriminator = 1;
    kind = "SHADOW";
    toJSON() {
        return {
            kind: "SHADOW",
        };
    }
    toEncodable() {
        return {
            SHADOW: {},
        };
    }
}
exports.SHADOW = SHADOW;
class LIVE {
    static discriminator = 2;
    static kind = "LIVE";
    discriminator = 2;
    kind = "LIVE";
    toJSON() {
        return {
            kind: "LIVE",
        };
    }
    toEncodable() {
        return {
            LIVE: {},
        };
    }
}
exports.LIVE = LIVE;
class DEPRECATED {
    static discriminator = 3;
    static kind = "DEPRECATED";
    discriminator = 3;
    kind = "DEPRECATED";
    toJSON() {
        return {
            kind: "DEPRECATED",
        };
    }
    toEncodable() {
        return {
            DEPRECATED: {},
        };
    }
}
exports.DEPRECATED = DEPRECATED;
class STAGING {
    static discriminator = 4;
    static kind = "STAGING";
    discriminator = 4;
    kind = "STAGING";
    toJSON() {
        return {
            kind: "STAGING",
        };
    }
    toEncodable() {
        return {
            STAGING: {},
        };
    }
}
exports.STAGING = STAGING;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("IGNORED" in obj) {
        return new IGNORED();
    }
    if ("SHADOW" in obj) {
        return new SHADOW();
    }
    if ("LIVE" in obj) {
        return new LIVE();
    }
    if ("DEPRECATED" in obj) {
        return new DEPRECATED();
    }
    if ("STAGING" in obj) {
        return new STAGING();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "IGNORED": {
            return new IGNORED();
        }
        case "SHADOW": {
            return new SHADOW();
        }
        case "LIVE": {
            return new LIVE();
        }
        case "DEPRECATED": {
            return new DEPRECATED();
        }
        case "STAGING": {
            return new STAGING();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "IGNORED"),
        borsh.struct([], "SHADOW"),
        borsh.struct([], "LIVE"),
        borsh.struct([], "DEPRECATED"),
        borsh.struct([], "STAGING"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
