// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/ExecutiveWithdrawAction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rebalance = exports.Unfreeze = exports.Freeze = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Freeze {
    static discriminator = 0;
    static kind = "Freeze";
    discriminator = 0;
    kind = "Freeze";
    toJSON() {
        return {
            kind: "Freeze",
        };
    }
    toEncodable() {
        return {
            Freeze: {},
        };
    }
}
exports.Freeze = Freeze;
class Unfreeze {
    static discriminator = 1;
    static kind = "Unfreeze";
    discriminator = 1;
    kind = "Unfreeze";
    toJSON() {
        return {
            kind: "Unfreeze",
        };
    }
    toEncodable() {
        return {
            Unfreeze: {},
        };
    }
}
exports.Unfreeze = Unfreeze;
class Rebalance {
    static discriminator = 2;
    static kind = "Rebalance";
    discriminator = 2;
    kind = "Rebalance";
    toJSON() {
        return {
            kind: "Rebalance",
        };
    }
    toEncodable() {
        return {
            Rebalance: {},
        };
    }
}
exports.Rebalance = Rebalance;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Freeze" in obj) {
        return new Freeze();
    }
    if ("Unfreeze" in obj) {
        return new Unfreeze();
    }
    if ("Rebalance" in obj) {
        return new Rebalance();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Freeze": {
            return new Freeze();
        }
        case "Unfreeze": {
            return new Unfreeze();
        }
        case "Rebalance": {
            return new Rebalance();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Freeze"),
        borsh.struct([], "Unfreeze"),
        borsh.struct([], "Rebalance"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
