// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/ExpanderStep.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Recenter = exports.ExpandOrContract = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class ExpandOrContract {
    static discriminator = 0;
    static kind = "ExpandOrContract";
    discriminator = 0;
    kind = "ExpandOrContract";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "ExpandOrContract",
            value: [this.value[0]],
        };
    }
    toEncodable() {
        return {
            ExpandOrContract: {
                _0: this.value[0],
            },
        };
    }
}
exports.ExpandOrContract = ExpandOrContract;
class Recenter {
    static discriminator = 1;
    static kind = "Recenter";
    discriminator = 1;
    kind = "Recenter";
    toJSON() {
        return {
            kind: "Recenter",
        };
    }
    toEncodable() {
        return {
            Recenter: {},
        };
    }
}
exports.Recenter = Recenter;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("ExpandOrContract" in obj) {
        const val = obj["ExpandOrContract"];
        return new ExpandOrContract([val["_0"]]);
    }
    if ("Recenter" in obj) {
        return new Recenter();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "ExpandOrContract": {
            return new ExpandOrContract([obj.value[0]]);
        }
        case "Recenter": {
            return new Recenter();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([borsh.u16("_0")], "ExpandOrContract"),
        borsh.struct([], "Recenter"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
