// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/RebalanceDriftParams.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RebalanceDriftParams = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class RebalanceDriftParams {
    startMidTick;
    ticksBelowMid;
    ticksAboveMid;
    secondsPerTick;
    direction;
    constructor(fields) {
        this.startMidTick = fields.startMidTick;
        this.ticksBelowMid = fields.ticksBelowMid;
        this.ticksAboveMid = fields.ticksAboveMid;
        this.secondsPerTick = fields.secondsPerTick;
        this.direction = fields.direction;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i32("startMidTick"),
            borsh.i32("ticksBelowMid"),
            borsh.i32("ticksAboveMid"),
            borsh.u64("secondsPerTick"),
            types.DriftDirection.layout("direction"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new RebalanceDriftParams({
            startMidTick: obj.startMidTick,
            ticksBelowMid: obj.ticksBelowMid,
            ticksAboveMid: obj.ticksAboveMid,
            secondsPerTick: obj.secondsPerTick,
            direction: types.DriftDirection.fromDecoded(obj.direction),
        });
    }
    static toEncodable(fields) {
        return {
            startMidTick: fields.startMidTick,
            ticksBelowMid: fields.ticksBelowMid,
            ticksAboveMid: fields.ticksAboveMid,
            secondsPerTick: fields.secondsPerTick,
            direction: fields.direction.toEncodable(),
        };
    }
    toJSON() {
        return {
            startMidTick: this.startMidTick,
            ticksBelowMid: this.ticksBelowMid,
            ticksAboveMid: this.ticksAboveMid,
            secondsPerTick: this.secondsPerTick.toString(),
            direction: this.direction.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new RebalanceDriftParams({
            startMidTick: obj.startMidTick,
            ticksBelowMid: obj.ticksBelowMid,
            ticksAboveMid: obj.ticksAboveMid,
            secondsPerTick: new bn_js_1.default(obj.secondsPerTick),
            direction: types.DriftDirection.fromJSON(obj.direction),
        });
    }
    toEncodable() {
        return RebalanceDriftParams.toEncodable(this);
    }
}
exports.RebalanceDriftParams = RebalanceDriftParams;
