// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/RebalanceDriftState.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RebalanceDriftState = void 0;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const types = __importStar(require("../types")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class RebalanceDriftState {
    step;
    lastDriftTimestamp;
    lastMidTick;
    constructor(fields) {
        this.step = fields.step;
        this.lastDriftTimestamp = fields.lastDriftTimestamp;
        this.lastMidTick = fields.lastMidTick;
    }
    static layout(property) {
        return borsh.struct([
            types.RebalanceDriftStep.layout("step"),
            borsh.u64("lastDriftTimestamp"),
            borsh.i32("lastMidTick"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new RebalanceDriftState({
            step: types.RebalanceDriftStep.fromDecoded(obj.step),
            lastDriftTimestamp: obj.lastDriftTimestamp,
            lastMidTick: obj.lastMidTick,
        });
    }
    static toEncodable(fields) {
        return {
            step: fields.step.toEncodable(),
            lastDriftTimestamp: fields.lastDriftTimestamp,
            lastMidTick: fields.lastMidTick,
        };
    }
    toJSON() {
        return {
            step: this.step.toJSON(),
            lastDriftTimestamp: this.lastDriftTimestamp.toString(),
            lastMidTick: this.lastMidTick,
        };
    }
    static fromJSON(obj) {
        return new RebalanceDriftState({
            step: types.RebalanceDriftStep.fromJSON(obj.step),
            lastDriftTimestamp: new bn_js_1.default(obj.lastDriftTimestamp),
            lastMidTick: obj.lastMidTick,
        });
    }
    toEncodable() {
        return RebalanceDriftState.toEncodable(this);
    }
}
exports.RebalanceDriftState = RebalanceDriftState;
