// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/RebalanceType.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Autodrift = exports.Expander = exports.PeriodicRebalance = exports.TakeProfit = exports.Drift = exports.PricePercentageWithReset = exports.PricePercentage = exports.Manual = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const borsh = __importStar(require("@coral-xyz/borsh"));
class Manual {
    static discriminator = 0;
    static kind = "Manual";
    discriminator = 0;
    kind = "Manual";
    toJSON() {
        return {
            kind: "Manual",
        };
    }
    toEncodable() {
        return {
            Manual: {},
        };
    }
}
exports.Manual = Manual;
class PricePercentage {
    static discriminator = 1;
    static kind = "PricePercentage";
    discriminator = 1;
    kind = "PricePercentage";
    toJSON() {
        return {
            kind: "PricePercentage",
        };
    }
    toEncodable() {
        return {
            PricePercentage: {},
        };
    }
}
exports.PricePercentage = PricePercentage;
class PricePercentageWithReset {
    static discriminator = 2;
    static kind = "PricePercentageWithReset";
    discriminator = 2;
    kind = "PricePercentageWithReset";
    toJSON() {
        return {
            kind: "PricePercentageWithReset",
        };
    }
    toEncodable() {
        return {
            PricePercentageWithReset: {},
        };
    }
}
exports.PricePercentageWithReset = PricePercentageWithReset;
class Drift {
    static discriminator = 3;
    static kind = "Drift";
    discriminator = 3;
    kind = "Drift";
    toJSON() {
        return {
            kind: "Drift",
        };
    }
    toEncodable() {
        return {
            Drift: {},
        };
    }
}
exports.Drift = Drift;
class TakeProfit {
    static discriminator = 4;
    static kind = "TakeProfit";
    discriminator = 4;
    kind = "TakeProfit";
    toJSON() {
        return {
            kind: "TakeProfit",
        };
    }
    toEncodable() {
        return {
            TakeProfit: {},
        };
    }
}
exports.TakeProfit = TakeProfit;
class PeriodicRebalance {
    static discriminator = 5;
    static kind = "PeriodicRebalance";
    discriminator = 5;
    kind = "PeriodicRebalance";
    toJSON() {
        return {
            kind: "PeriodicRebalance",
        };
    }
    toEncodable() {
        return {
            PeriodicRebalance: {},
        };
    }
}
exports.PeriodicRebalance = PeriodicRebalance;
class Expander {
    static discriminator = 6;
    static kind = "Expander";
    discriminator = 6;
    kind = "Expander";
    toJSON() {
        return {
            kind: "Expander",
        };
    }
    toEncodable() {
        return {
            Expander: {},
        };
    }
}
exports.Expander = Expander;
class Autodrift {
    static discriminator = 7;
    static kind = "Autodrift";
    discriminator = 7;
    kind = "Autodrift";
    toJSON() {
        return {
            kind: "Autodrift",
        };
    }
    toEncodable() {
        return {
            Autodrift: {},
        };
    }
}
exports.Autodrift = Autodrift;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Manual" in obj) {
        return new Manual();
    }
    if ("PricePercentage" in obj) {
        return new PricePercentage();
    }
    if ("PricePercentageWithReset" in obj) {
        return new PricePercentageWithReset();
    }
    if ("Drift" in obj) {
        return new Drift();
    }
    if ("TakeProfit" in obj) {
        return new TakeProfit();
    }
    if ("PeriodicRebalance" in obj) {
        return new PeriodicRebalance();
    }
    if ("Expander" in obj) {
        return new Expander();
    }
    if ("Autodrift" in obj) {
        return new Autodrift();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Manual": {
            return new Manual();
        }
        case "PricePercentage": {
            return new PricePercentage();
        }
        case "PricePercentageWithReset": {
            return new PricePercentageWithReset();
        }
        case "Drift": {
            return new Drift();
        }
        case "TakeProfit": {
            return new TakeProfit();
        }
        case "PeriodicRebalance": {
            return new PeriodicRebalance();
        }
        case "Expander": {
            return new Expander();
        }
        case "Autodrift": {
            return new Autodrift();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "Manual"),
        borsh.struct([], "PricePercentage"),
        borsh.struct([], "PricePercentageWithReset"),
        borsh.struct([], "Drift"),
        borsh.struct([], "TakeProfit"),
        borsh.struct([], "PeriodicRebalance"),
        borsh.struct([], "Expander"),
        borsh.struct([], "Autodrift"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
