// @kamino-finance/kliquidity-sdk/dist/@codegen/kliquidity/types/RemoveLiquidityMode.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.All = exports.Bps = exports.Liquidity = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class Liquidity {
    static discriminator = 0;
    static kind = "Liquidity";
    discriminator = 0;
    kind = "Liquidity";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "Liquidity",
            value: [this.value[0].toString()],
        };
    }
    toEncodable() {
        return {
            Liquidity: {
                _0: this.value[0],
            },
        };
    }
}
exports.Liquidity = Liquidity;
class Bps {
    static discriminator = 1;
    static kind = "Bps";
    discriminator = 1;
    kind = "Bps";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "Bps",
            value: [this.value[0]],
        };
    }
    toEncodable() {
        return {
            Bps: {
                _0: this.value[0],
            },
        };
    }
}
exports.Bps = Bps;
class All {
    static discriminator = 2;
    static kind = "All";
    discriminator = 2;
    kind = "All";
    toJSON() {
        return {
            kind: "All",
        };
    }
    toEncodable() {
        return {
            All: {},
        };
    }
}
exports.All = All;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Liquidity" in obj) {
        const val = obj["Liquidity"];
        return new Liquidity([val["_0"]]);
    }
    if ("Bps" in obj) {
        const val = obj["Bps"];
        return new Bps([val["_0"]]);
    }
    if ("All" in obj) {
        return new All();
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Liquidity": {
            return new Liquidity([new bn_js_1.default(obj.value[0])]);
        }
        case "Bps": {
            return new Bps([obj.value[0]]);
        }
        case "All": {
            return new All();
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([borsh.u128("_0")], "Liquidity"),
        borsh.struct([borsh.u16("_0")], "Bps"),
        borsh.struct([], "All"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
