// @switchboard-xyz/common/dist/esm/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bs58 = exports.BN = exports.Big = exports.serializeOracleJob = exports.deserializeOracleJob = exports.sleep = exports.Task = exports.OracleJob = exports.OracleFeed = exports.isSupportedEvmChain = exports.isSupportedChainId = exports.isSupportedChain = exports.getSupportedEvmChainId = exports.getSupportedEvmChain = exports.getSupportedChain = exports.networks = exports.BigUtils = void 0;
exports.BigUtils = __importStar(require("./big.cjs"));
__exportStar(require("./CrossbarClient.cjs"), exports);
__exportStar(require("./FeedHash.cjs"), exports);
var index_js_1 = require("./networks/index.cjs");
Object.defineProperty(exports, "networks", { enumerable: true, get: function () { return __importDefault(index_js_1).default; } });
var index_js_2 = require("./networks/index.cjs");
Object.defineProperty(exports, "getSupportedChain", { enumerable: true, get: function () { return index_js_2.getSupportedChain; } });
Object.defineProperty(exports, "getSupportedEvmChain", { enumerable: true, get: function () { return index_js_2.getSupportedEvmChain; } });
Object.defineProperty(exports, "getSupportedEvmChainId", { enumerable: true, get: function () { return index_js_2.getSupportedEvmChainId; } });
Object.defineProperty(exports, "isSupportedChain", { enumerable: true, get: function () { return index_js_2.isSupportedChain; } });
Object.defineProperty(exports, "isSupportedChainId", { enumerable: true, get: function () { return index_js_2.isSupportedChainId; } });
Object.defineProperty(exports, "isSupportedEvmChain", { enumerable: true, get: function () { return index_js_2.isSupportedEvmChain; } });
__exportStar(require("./networks/types.cjs"), exports);
var protos_js_1 = require("./protos.cjs");
Object.defineProperty(exports, "OracleFeed", { enumerable: true, get: function () { return protos_js_1.OracleFeed; } });
Object.defineProperty(exports, "OracleJob", { enumerable: true, get: function () { return protos_js_1.OracleJob; } });
Object.defineProperty(exports, "Task", { enumerable: true, get: function () { return protos_js_1.Task; } });
__exportStar(require("./secrets.cjs"), exports);
__exportStar(require("./SwitchboardDecimal.cjs"), exports);
__exportStar(require("./types.cjs"), exports);
var async_js_1 = require("./utils/async.cjs");
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return async_js_1.sleep; } });
__exportStar(require("./utils/index.cjs"), exports);
var oracle_job_js_1 = require("./utils/oracle-job.cjs"); // NOTE: Exported here for compatibility with old code
Object.defineProperty(exports, "deserializeOracleJob", { enumerable: true, get: function () { return oracle_job_js_1.deserializeOracleJob; } });
Object.defineProperty(exports, "serializeOracleJob", { enumerable: true, get: function () { return oracle_job_js_1.serializeOracleJob; } });
var big_js_1 = require("big.js");
Object.defineProperty(exports, "Big", { enumerable: true, get: function () { return big_js_1.Big; } });
var bn_js_1 = require("bn.js");
Object.defineProperty(exports, "BN", { enumerable: true, get: function () { return __importDefault(bn_js_1).default; } });
var bs58_1 = require("bs58");
Object.defineProperty(exports, "bs58", { enumerable: true, get: function () { return __importDefault(bs58_1).default; } });
// import { OracleJob } from "./protos/index.js";
// export import ITask = OracleJob.ITask;
// export import Task = OracleJob.Task;
const minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
minimal_js_1.default.util.toJSONOptions = {
    longs: String,
    enums: String,
    bytes: String,
    json: true,
    // oneofs: true,
};
